<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Performance Benchmarks • BPCells</title>
<script src="../../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../../deps/headroom-0.11.0/headroom.min.js"></script><script src="../../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../../deps/search-1.0.0/fuse.min.js"></script><script src="../../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../../pkgdown.js"></script><link href="../../extra.css" rel="stylesheet">
<meta property="og:title" content="Performance Benchmarks">
<script defer data-domain="benparks.net" src="https://plausible.benparks.net/js/visit-counts.js"></script><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.16.15/dist/katex.min.css" integrity="sha384-Htz9HMhiwV8GuQ28Xr9pEs1B4qJiYu/nYLLwlDklR53QibDfmQzi7rYxXhMH/5/u" crossorigin="anonymous">
<!-- The loading of KaTeX is deferred to speed up page rendering --><script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.15/dist/katex.min.js" integrity="sha384-bxmi2jLGCvnsEqMuYLKE/KsVCxV3PqmKeK6Y6+lmNXBry6+luFkEOsmp5vD9I/7+" crossorigin="anonymous"></script><!-- To automatically render math in text elements, include the auto-render extension: --><script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.15/dist/contrib/auto-render.min.js" integrity="sha384-hCXGrW6PitJEwbkoStFjeJxv+fSOOQKOPbJxSfM6G5sWZjAyWhXiTIIAmQqnlLlh" crossorigin="anonymous" onload="renderMathInElement(document.body);"></script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../../index.html">BPCells</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.3.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../../articles/pbmc3k.html">Get Started</a></li>
<li class="nav-item"><a class="nav-link" href="../../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../../articles/pbmc3k.html">Basic tutorial</a></li>
    <li><a class="dropdown-item" href="../../articles/web-only/benchmarks.html">Performance Benchmarks</a></li>
    <li><a class="dropdown-item" href="../../articles/web-only/how-it-works.html">How BPCells works</a></li>
    <li><a class="dropdown-item" href="../../articles/web-only/programming-efficiency.html">Efficiency tips</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../../articles/index.html">More articles...</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../../news/index.html">News</a></li>
<li class="nav-item"><a class="nav-link" href="../../python/index.html"><span class="fa fa-arrow-up-right-from-square"></span> Python Docs</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/bnprks/BPCells" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Performance Benchmarks</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/bnprks/BPCells/tree/main/r/vignettes/web-only/benchmarks.Rmd" class="external-link"><code>vignettes/web-only/benchmarks.Rmd</code></a></small>
      <div class="d-none name"><code>benchmarks.Rmd</code></div>
    </div>

    
    
<!-- This is to make sure the logo file gets copied to output on a full website rebuild <img src="/home/bparks/dev/github/bnprks/BPCells/r/vignettes/web-only/logo.png" style="visibility: hidden; width: 0px" /> -->
<p>These performance benchmarks are based on extensive benchmarks from
the <a href="https://www.biorxiv.org/content/10.1101/2025.03.27.645853v1" class="external-link">BPCells
manuscript</a>. Full benchmarking/plotting code and data TSVs are
available <a href="https://github.com/GreenleafLab/BPCells_paper" class="external-link">here</a>. These
numbers are just a snapshot of tool performance as of December 2024, so
we encourage you to check performance on your own datasets with the
latest versions of all software packages to get the most up-to-date
results.</p>
<div class="section level3">
<h3 id="rna-seq-normalization-pca">RNA-seq normalization + PCA<a class="anchor" aria-label="anchor" href="#rna-seq-normalization-pca"></a>
</h3>
<p>Because BPCells can perform all operations streaming from disk, it is
able to use dramatically less memory than traditional in-memory
workflows. We have also extensively optimized the underlying C++ code to
make BPCells much faster than traditional disk-backed tools like
DelayedArray.</p>
<p>In the benchmark below, we show the time and memory usage to perform
a standardized workflow of data import, normalization, variable gene
selection, and PCA. Note that we show Seurat’s in-memory workflow,
though Seurat v5 also offers a BPCells integration for disk-backed
operations. All tools were given a 3 hour time limit and up to 256GB of
RAM, but only BPCells was able to process the largest datasets within
these resource limits.</p>
<p><img src="benchmark-plots%2Fpca_comparison.png" class="mx-auto d-block" style="max-height: 20rem"></p>
<p>Notice that execution speed tends to scale with the number of
non-zero entries in a matrix, whereas memory usage for BPCells scales
with the number of cells (i.e. space required to store the output PCA
embeddings).</p>
<div class="section level4">
<h4 id="multi-threading">Multi-threading<a class="anchor" aria-label="anchor" href="#multi-threading"></a>
</h4>
<p>The above benchmark was run single-threaded since not all tools
support multi-threading. However, BPCells can offer 5-10x speedups with
multiple threads:</p>
<p><img src="benchmark-plots%2Fpca_multithread.png" class="mx-auto d-block" style="max-height: 20rem"></p>
</div>
</div>
<div class="section level3">
<h3 id="bitpacking-compression">Bitpacking compression<a class="anchor" aria-label="anchor" href="#bitpacking-compression"></a>
</h3>
<div class="section level4">
<h4 id="counts-matrices-rna-or-atac">Counts matrices (RNA or ATAC)<a class="anchor" aria-label="anchor" href="#counts-matrices-rna-or-atac"></a>
</h4>
<p>BPCells uses bitpacking compression to help speed up disk-backed
workflows. While general purpose compression algorithms shrink file
sizes and reduce disk read bandwidth, they typically come at a high
compute cost. BPCells is able to provide similar space savings to
general-purpose compression algorithms like gzip (10x HDF5) and
LZ4-Blosc (zarr), but with much lower compute cost. AnnData h5ad files
default to not using compression due to the speed costs for read/write,
but with BPCells bitpacking compression we can get faster read/write
with 4-7x space savings.</p>
<p>Here, we benchmark storing + loading a 1.3M cell RNA-seq experiment
from 10x Genomics, using default compression settings for each storage
format<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content="&lt;p&gt;10x HDF5: gzip; h5ad: none, AnnData zarr: LZ4-Blosc,
BPCells: bitpacking&lt;/p&gt;"><sup>1</sup></a>.
Notice that only BPCells is the only compressed format fast enough to
provide faster read and write speeds than an uncompressed h5ad<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content="&lt;p&gt;These benchmarks were run single-threaded on local SSD
storage, which is the hardest scenario to outperform uncompressed
formats.&lt;/p&gt;"><sup>2</sup></a>.</p>
<p><img src="benchmark-plots%2Frna_storage.png" class="mx-auto d-block"></p>
<p>Note that we don’t benchmark 10x HDF5 write, since 10x does not
directly provide software to perform arbitrary matrix writes. It would
likely be even slower than the 10x read speed.</p>
</div>
<div class="section level4">
<h4 id="fragment-alignments-atac">Fragment alignments (ATAC)<a class="anchor" aria-label="anchor" href="#fragment-alignments-atac"></a>
</h4>
<p>In addition to RNA/ATAC counts matrices, BPCells also introduces
compressed file formats for scATAC-seq fragments. We compared to 10x
fragment files, ArchR, and SnapATAC2 for storage of a 1M cell dataset,
and found that BPCells gave the smallest file sizes and fastest
read/write speeds.</p>
<p>By using bitpacking compression, BPCells can afford the storage space
to keep fragments in fully genome-sorted order. This makes import from
10x fragment files dramatically faster compared to ArchR and SnapATAC2
which must re-sort fragments to be grouped by cell. We’ll see later that
it also speeds up genomic overlap calculations.</p>
<p><img src="benchmark-plots%2Fatac_storage.png" class="mx-auto d-block" style="max-height: 20rem"></p>
</div>
</div>
<div class="section level3">
<h3 id="atac-seq-overlap-calculations">ATAC-seq overlap calculations<a class="anchor" aria-label="anchor" href="#atac-seq-overlap-calculations"></a>
</h3>
<p>When working with large reference datasets, it’s often helpful to be
able to quickly re-quantify cell x peak overlap matrices directly from
fragments, as datasets all usually use different sets of peak
coordinates depending on the biology of interest. BPCells stores
fragments in genome-sorted order, so it’s able to perform peak
calculations much faster than ArchR or SnapATAC2.</p>
<p>BPCells takes only seconds to find all overlaps with a small set of
10 peaks, which also means that it is very fast at calculating genomic
coverage tracks for visualization of large datasets.</p>
<p><img src="benchmark-plots%2Fpeaks_comparison.png" class="mx-auto d-block" style="max-height: 20rem"></p>
</div>
<div class="section level3">
<h3 id="m-cell-analysis-of-cellxgene-census">44M cell analysis of CELLxGENE census<a class="anchor" aria-label="anchor" href="#m-cell-analysis-of-cellxgene-census"></a>
</h3>
<p>With the combination of fast storage and disk-backed compute, BPCells
is able to handle all unique human cells from the CELLxGENE census on a
laptop with 16 threads and 32GB of RAM. Compared to the TileDB matrix
storage format, we found that the BPCells file formats offer much faster
read/write times at similar space usage<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content="&lt;p&gt;These plots show the BPCells feature-major layout
compared to TileDB with raw counts. See the BPCells manuscript for
timing information on servers and for additional matrix variants. Write
time is time to subset the CELLxGENE census to remove duplicate cells,
including the time to swap storage order from cell-major to
feature-major for BPCells.&lt;/p&gt;"><sup>3</sup></a>. Note that this
dataset would take over 750GB to store the counts matrix without
compression.</p>
<p><img src="benchmark-plots%2Fcellxgene_file_handling.png" class="mx-auto d-block" style="max-height: 20rem"></p>
<p>Because of these much faster file read speeds, BPCells is able to do
light compute tasks like computing per-gene mean and variance across 44M
cells in &lt;3 minutes. PCA is the most expensive operation BPCells
performs, using 167 passes over the input matrix to calculate 32 PCs.
Still, it completes in less than 1 hour on a server and 6.2 hours on a
laptop. This makes atlas-scale analysis possible on a laptop while
leaving headroom on a server for datasets up to an order of magnitude
larger.</p>
<p><img src="benchmark-plots%2Fcellxgene_compute_tasks.png" class="mx-auto d-block" style="max-height: 20rem"></p>
<p>Benchmark details: Mean and variance computed on a log-normalized
matrix, unlike the plots in the BPCells manuscript which were just
scaled by read counts to match the default normalization of the
CELLxGENE census. Laptop has 32GB of RAM and 16 threads; server has
256GB of RAM and 32 threads. See <a href="https://www.biorxiv.org/content/10.1101/2025.03.27.645853v1" class="external-link">BPCells
manuscript</a> for further details.</p>
</div>
<div class="section level2">
<h2 id="update-log">Update log<a class="anchor" aria-label="anchor" href="#update-log"></a>
</h2>
<ul>
<li>March 29, 2025: Update benchmarks with numbers + performance from
BPCells manuscript.
<ul>
<li>Normalization &amp; PCA: Add more comparison datasets, add
DelayedArray results, and add multithreading plot</li>
<li>RNA storage: Add comparisons with zarr and include write times.</li>
<li>ATAC storage: Add 10x and SnapATAC2 results and include read times,
and switch comparison to be on a larger dataset.</li>
<li>Peak matrix: Update benchmarks to show cross-dataset results at 100k
peaks, and peak subset results on 1M cells rather than 30k cells. Add in
comparisons to SnapATAC2.</li>
<li>Add results for 44M cell analysis.</li>
</ul>
</li>
<li>March 30, 2023: Added clarification that AnnData benchmarks referred
to h5ad with default compression settings (i.e. none).</li>
<li>March 29, 2023: Created benchmark page.</li>
</ul>
</div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Created by Benjamin Parks.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>



  <img src="https://plausible.benparks.net/flask-plausible/bpcells-docs.png" style="position:absolute;">
</body>
</html>
