<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Efficiency tips • BPCells</title>
<script src="../../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../../deps/headroom-0.11.0/headroom.min.js"></script><script src="../../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../../deps/search-1.0.0/fuse.min.js"></script><script src="../../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../../pkgdown.js"></script><link href="../../extra.css" rel="stylesheet">
<meta property="og:title" content="Efficiency tips">
<script defer data-domain="benparks.net" src="https://plausible.benparks.net/js/visit-counts.js"></script><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.16.15/dist/katex.min.css" integrity="sha384-Htz9HMhiwV8GuQ28Xr9pEs1B4qJiYu/nYLLwlDklR53QibDfmQzi7rYxXhMH/5/u" crossorigin="anonymous">
<!-- The loading of KaTeX is deferred to speed up page rendering --><script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.15/dist/katex.min.js" integrity="sha384-bxmi2jLGCvnsEqMuYLKE/KsVCxV3PqmKeK6Y6+lmNXBry6+luFkEOsmp5vD9I/7+" crossorigin="anonymous"></script><!-- To automatically render math in text elements, include the auto-render extension: --><script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.15/dist/contrib/auto-render.min.js" integrity="sha384-hCXGrW6PitJEwbkoStFjeJxv+fSOOQKOPbJxSfM6G5sWZjAyWhXiTIIAmQqnlLlh" crossorigin="anonymous" onload="renderMathInElement(document.body);"></script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../../index.html">BPCells</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.3.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../../articles/pbmc3k.html">Get Started</a></li>
<li class="nav-item"><a class="nav-link" href="../../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../../articles/pbmc3k.html">Basic tutorial</a></li>
    <li><a class="dropdown-item" href="../../articles/web-only/benchmarks.html">Performance Benchmarks</a></li>
    <li><a class="dropdown-item" href="../../articles/web-only/how-it-works.html">How BPCells works</a></li>
    <li><a class="dropdown-item" href="../../articles/web-only/programming-efficiency.html">Efficiency tips</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../../articles/index.html">More articles...</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../../news/index.html">News</a></li>
<li class="nav-item"><a class="nav-link" href="../../python/index.html"><span class="fa fa-arrow-up-right-from-square"></span> Python Docs</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/bnprks/BPCells" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Efficiency tips</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/bnprks/BPCells/tree/main/r/vignettes/web-only/programming-efficiency.Rmd" class="external-link"><code>vignettes/web-only/programming-efficiency.Rmd</code></a></small>
      <div class="d-none name"><code>programming-efficiency.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="normalizations-and-pca">Normalizations and PCA<a class="anchor" aria-label="anchor" href="#normalizations-and-pca"></a>
</h2>
<ul>
<li><p>Avoid dense matrices whenever possible. Put normalizations that
preserve sparsity (0 values stay 0) before normalizations that break
sparsity (e.g. adding values to each row/column). A typical RNA-seq
matrix has &lt;5% non-zero entries, so your code will operate on 20x
more entries with a dense matrix.</p></li>
<li>
<p>For most operations, we recommend using lazy evaluation to avoid
creating intermediate matrices. The one common exception to this rule is
when running PCA. Because PCA requires looping through the matrix
several hundred times, it is often faster to write the matrix to disk
once just before PCA rather than recalculating the entries on each PCA
iteration.</p>
<ul>
<li>For storage efficiency, keep any sparsity-breaking normalizations
delayed, but store all the sparse normalizations in a temporary location
with <code><a href="../../reference/matrix_io.html">write_matrix_dir()</a></code> then apply the sparsity-breaking
normalizations</li>
</ul>
</li>
<li><p>Adding values to the rows/columns of a matrix has very little
overhead for PCA because it translates into a pre or post processing
step before each mat-vec multiply iteration. As a sparsity-breaking
operation, adding a vector to the matrix causes most other operations to
become more expensive, however.</p></li>
</ul>
</div>
<div class="section level2">
<h2 id="storage-order">Storage order<a class="anchor" aria-label="anchor" href="#storage-order"></a>
</h2>
<ul>
<li>Sparse matrices can be stored in a row-major or column-major
orientation with BPCells. Along the indexed dimension (e.g. rows for
row-major), BPCells can efficiently seek to a selected column without
reading the whole matrix. This has performance implications for certain
operations:
<ul>
<li>Marker features can only be computed on a matrix indexed by
gene/feature.</li>
<li>Sparse matrix multiplication can only be performed between matrices
with the same storage order</li>
<li>Sparse matrix multiplication performance can change dramatically
depending on the storage order and relative matrix size/sparsity. For
column-major matrices, the left matrix should be fast to load and
contain few delayed operations, while the right matrix can be slow to
load and contain many delayed operations. For row-major matrices the
left/right preferences are reversed.</li>
</ul>
</li>
<li>You can check the storage order for a matrix by printing it out in
the R terminal</li>
<li>When calling the <code><a href="https://rdrr.io/pkg/BiocGenerics/man/t.html" class="external-link">t()</a></code> function, BPCells just flips a
boolean flag for whether the matrix is row-major or column-major. This
does not affect the underlying storage order.</li>
<li>To adjust the underlying storage order, call
<code><a href="../../reference/transpose_storage_order.html">transpose_storage_order()</a></code>. This is a slower operation, that
requires writing a new copy of the data to disk.</li>
</ul>
</div>
<div class="section level2">
<h2 id="other-tips">Other tips<a class="anchor" aria-label="anchor" href="#other-tips"></a>
</h2>
<ul>
<li>When running disk-backed analysis, always try to store your working
copy of the data on fast local SSDs. This is the default for laptops,
but on servers you may want to copy data files from a networked file
system to a physically attached SSD for best performance.</li>
<li>Use a single call to <code><a href="../../reference/matrix_stats.html">matrix_stats()</a></code> to calculate mean +
variance in a single pass through the matrix when possible. See the
function reference for details.</li>
<li>For ATAC-seq data, you can calculate variable features on the tile
matrix without ever saving it to disk. This allows you to subset to
variable tiles and create a peak matrix with just your variable tiles
for some space savings.</li>
</ul>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Created by Benjamin Parks.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>



  <img src="https://plausible.benparks.net/flask-plausible/bpcells-docs.png" style="position:absolute;">
</body>
</html>
