<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Single Cell Counts Matrices to PCA • BPCells</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><link href="extra.css" rel="stylesheet">
<meta property="og:title" content="Single Cell Counts Matrices to PCA">
<meta name="description" content="&gt; Efficient operations for single cell ATAC-seq fragments and RNA counts matrices. Interoperable with standard file formats, and introduces efficient bit-packed formats that allow large storage savings and increased read speeds.">
<meta property="og:description" content="&gt; Efficient operations for single cell ATAC-seq fragments and RNA counts matrices. Interoperable with standard file formats, and introduces efficient bit-packed formats that allow large storage savings and increased read speeds.">
<script defer data-domain="benparks.net" src="https://plausible.benparks.net/js/visit-counts.js"></script><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.16.15/dist/katex.min.css" integrity="sha384-Htz9HMhiwV8GuQ28Xr9pEs1B4qJiYu/nYLLwlDklR53QibDfmQzi7rYxXhMH/5/u" crossorigin="anonymous">
<!-- The loading of KaTeX is deferred to speed up page rendering --><script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.15/dist/katex.min.js" integrity="sha384-bxmi2jLGCvnsEqMuYLKE/KsVCxV3PqmKeK6Y6+lmNXBry6+luFkEOsmp5vD9I/7+" crossorigin="anonymous"></script><!-- To automatically render math in text elements, include the auto-render extension: --><script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.15/dist/contrib/auto-render.min.js" integrity="sha384-hCXGrW6PitJEwbkoStFjeJxv+fSOOQKOPbJxSfM6G5sWZjAyWhXiTIIAmQqnlLlh" crossorigin="anonymous" onload="renderMathInElement(document.body);"></script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">BPCells</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.3.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="articles/pbmc3k.html">Get Started</a></li>
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="articles/pbmc3k.html">Basic tutorial</a></li>
    <li><a class="dropdown-item" href="articles/web-only/benchmarks.html">Performance Benchmarks</a></li>
    <li><a class="dropdown-item" href="articles/web-only/how-it-works.html">How BPCells works</a></li>
    <li><a class="dropdown-item" href="articles/web-only/programming-efficiency.html">Efficiency tips</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="articles/index.html">More articles...</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="news/index.html">News</a></li>
<li class="nav-item"><a class="nav-link" href="python/index.html"><span class="fa fa-arrow-up-right-from-square"></span> Python Docs</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/bnprks/BPCells" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="bpcells">BPCells<a class="anchor" aria-label="anchor" href="#bpcells"></a>
</h1></div>
<p><em>This site is for the R package. Python site <a href="python/index.html">here</a> (experimental)</em></p>
<p>BPCells is a package for high performance single cell analysis of large RNA-seq and ATAC-seq datasets. It can run normalization and PCA of a 1.3M cell dataset in 4 minutes with 2GB of RAM, or create scATAC-seq peak matrices from fragment coordinates with 50x less CPU time than ArchR or SnapATAC2. BPCells can even handle the full CELLxGENE census human dataset, running full precision PCA on a 44M cell x 60k gene matrix in 6 hours on a laptop or &lt;1 hour on a server. See our <a href="articles/web-only/benchmarks.html">benchmarks page</a> for details.</p>
<p>BPCells provides:</p>
<ul>
<li>Efficient storage of single cell datasets via bitpacking compression</li>
<li>Fast, disk-backed RNA-seq and ATAC-seq data processing powered by C++</li>
<li>Downstream analysis such as marker genes, and clustering</li>
<li>Interoperability with AnnData, 10x datasets, R sparse matrices, and GRanges</li>
<li>Demonstrated scalability to 44M cells on a laptop</li>
</ul>
<p>Additionally, BPCells exposes its optimized data processing infrastructure for use in scaling 3rd party single cell tools (e.g. Seurat)</p>
<div class="section level2">
<h2 id="learn-more">Learn more<a class="anchor" aria-label="anchor" href="#learn-more"></a>
</h2>
<ul>
<li><a href="https://www.biorxiv.org/content/10.1101/2025.03.27.645853v1" class="external-link">BioRxiv preprint</a></li>
<li><a href="articles/web-only/benchmarks.html">Benchmarks</a></li>
<li><a href="articles/pbmc3k.html">Multiomic analysis example</a></li>
<li><a href="articles/web-only/how-it-works.html">How BPCells works</a></li>
<li><a href="articles/index.html">Additional articles</a></li>
<li><a href="reference/index.html">Function documentation</a></li>
<li><a href="news/index.html">News</a></li>
</ul>
</div>
<div class="section level2">
<h2 id="r-installation">R Installation<a class="anchor" aria-label="anchor" href="#r-installation"></a>
</h2>
<p>We recommend installing BPCells directly from github:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"bnprks/BPCells/r"</span><span class="op">)</span></span></code></pre></div>
<p>Before installing, you must have the HDF5 library installed and accessible on your system. HDF5 can be installed from your choice of package manager.</p>
<p>For Mac and Windows users having trouble installing from github, check our <a href="https://bnprks.r-universe.dev/BPCells" class="external-link">R-universe</a> page for instructions to install pre-built binary packages. These binary packages automatically track the latest github main branch.</p>
<details><summary>
Click here for operating system specific installation information for github-based installs
</summary><div class="section level3">
<h3 id="linux">Linux<a class="anchor" aria-label="anchor" href="#linux"></a>
</h3>
<p>Obtaining the HDF5 dependency is usually pretty straightforward on Linux</p>
<ul>
<li>apt: <code>sudo apt-get install libhdf5-dev</code>
</li>
<li>yum: <code>sudo yum install hdf5-devel</code>
</li>
<li>conda: <code>conda install -c conda-forge hdf5</code>
<ul>
<li>Note: Linux users should prefer their distro’s package manager (e.g. <code>apt</code> or <code>yum</code>) when possible, as it appears to give a slightly more reliable installation experience.</li>
</ul>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="windows">Windows<a class="anchor" aria-label="anchor" href="#windows"></a>
</h3>
<p>Compiling R packages from source on Windows requires installing <a href="https://cran.r-project.org/bin/windows/Rtools/" class="external-link">R tools for Windows</a>. See <a href="https://github.com/bnprks/BPCells/issues/9" class="external-link">Issue #9</a> for more discussion.</p>
</div>
<div class="section level3">
<h3 id="macos">MacOS<a class="anchor" aria-label="anchor" href="#macos"></a>
</h3>
<p>For MacOS, installing HDF5 through homebrew seems to be most reliable: <code>brew install hdf5</code>.</p>
<p><strong>Mac-specific troubleshooting</strong>:</p>
<ul>
<li>
<strong>Macs with ARM CPUs</strong>: a common error is to have an ARM-based HDF5 install but an x86-based R install. This will cause errors when BPCells tries to access HDF5 during installation.
<ul>
<li>Check your R installation by running <code><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo()</a></code>, and seeing if it lists ARM or x86 under “Platform”.</li>
<li>The easiest option is to use ARM R because homebrew will default to an ARM hdf5 installation</li>
<li>It is <a href="https://codetinkering.com/switch-homebrew-arm-x86/" class="external-link">possible</a> (though tricky) to install an x86 copy of homebrew in order to access an x86 version of hdf5</li>
</ul>
</li>
<li>
<strong>Older Macs (10.14 Mojave or older)</strong>: The default compiler on old Macs does not support needed C++17 filesystem features. See <a href="https://github.com/bnprks/BPCells/issues/3#issuecomment-1375238635" class="external-link">issue #3</a> for tips getting a newer compiler set up via homebrew.</li>
</ul>
</div>
<div class="section level3">
<h3 id="supported-compilers">Supported compilers<a class="anchor" aria-label="anchor" href="#supported-compilers"></a>
</h3>
<p>In most cases, you will already have an appropriate compiler. BPCells recommends gcc &gt;=9.1, or clang &gt;= 9.0. This corresponds to versions from late-2018 and newer. Older versions may work in some cases so long as they have basic C++17 support, but they are not officially supported.</p>
</div>
<div class="section level3">
<h3 id="general-installation-troubleshooting">General Installation troubleshooting<a class="anchor" aria-label="anchor" href="#general-installation-troubleshooting"></a>
</h3>
<p>BPCells tries to print informative error messages during compilation to help diagnose the problem. For a more verbose set of information, run <code>Sys.setenv(BPCELLS_DEBUG_INSTALL="true")</code> prior to <code>remotes::install_github("bnprks/BPCells/r")</code>. If you still can’t solve the issue with that additional information, feel free to file a Github issue, being sure to use a <a href="https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/organizing-information-with-collapsed-sections" class="external-link">collapsible section</a> for the verbose installation log.</p>

</div>
</details>
</div>
<div class="section level2">
<h2 id="contributing">Contributing<a class="anchor" aria-label="anchor" href="#contributing"></a>
</h2>
<p>BPCells is an open source project, and we welcome quality contributions. If you are interested in contributing and have experience with C++, along with Python or R, feel free to reach out with ideas you would like to implement yourself. We’re happy to provide pointers for how to get started, time permitting.</p>
<p>If you are unfamiliar with C++ it will be difficult for you to contribute code, but detailed bug reports with <a href="https://reprex.tidyverse.org/articles/reprex-dos-and-donts.html" class="external-link">reproducible examples</a> are still a great way to help out. Github issues are the best forum for this.</p>
<p>If you maintain a single cell analysis package and want to use BPCells to improve your scalability, we’re happy to provide advice. We have had a couple of labs try this so far, with promising success. Email is the best way to get in touch for this (look in the <code>DESCRIPTION</code> file on github for contact info). Python developers welcome, though the current python package is still in experimental status.</p>
</div>
</div>
  </main><aside class="col-md-3"><div class="">
<h2 data-toc-skip></h2>
<ul class="list-unstyled">
<li><p><img src="articles/web-only/logo.png" width="150px"></p></li>
</ul>
</div>

<div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/bnprks/BPCells" class="external-link">Browse source code</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li>Apache-2.0 or <a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a>
</li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing BPCells</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Benjamin Parks <br><small class="roles"> Author, maintainer, copyright holder </small> <a href="https://orcid.org/0000-0002-0261-7472" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Immanuel Abdi <br><small class="roles"> Author </small>  </li>
<li>Stanford University <br><small class="roles"> Copyright holder, funder </small>  </li>
<li>Genentech, Inc. <br><small class="roles"> Copyright holder, funder </small>  </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Created by Benjamin Parks.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>



  <img src="https://plausible.benparks.net/flask-plausible/bpcells-docs.png" style="position:absolute;">
</body>
</html>
