% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/data.R
\name{get_demo_mat}
\alias{get_demo_mat}
\alias{get_demo_frags}
\alias{remove_demo_data}
\title{Retrieve BPCells demo data}
\usage{
get_demo_mat(filter_qc = TRUE, subset = TRUE)

get_demo_frags(filter_qc = TRUE, subset = TRUE)

remove_demo_data()
}
\arguments{
\item{filter_qc}{(bool) Whether to filter both the RNA and ATAC data using qc metrics (described in \code{details}).}

\item{subset}{(bool) Whether to subset to only genes/insertions on chromosome 4 and 11.}
}
\value{
\itemize{
\item \code{get_demo_mat()}: (IterableMatrix) A \verb{(features x cells)} matrix.
}

\itemize{
\item \code{get_demo_frags()}: (IterableFragments) A Fragments object.
}

\itemize{
\item \code{remove_demo_data()}: \code{NULL}
}
}
\description{
\ifelse{html}{\href{https://lifecycle.r-lib.org/articles/stages.html#experimental}{\figure{lifecycle-experimental.svg}{options: alt='[Experimental]'}}}{\strong{[Experimental]}} \cr Functions to download matrices and fragments derived from the 10X Genomics PBMC 3k dataset,
with options to filter with common qc metrics, and to subset genes and fragments to only chromosome 4 and 11.
}
\details{
These data functions are experimental.
The interface, as well as the demo dataset itself will likely undergo changes in the near future.

\strong{Data Processing}:

The first time either \code{get_demo_mat()} are ran \code{get_demo_frags()},
demo data is downloaded and stored in the BPCells data directory  (under \code{file.path(tools::R_user_dir("BPcells", which="data"), "demo_data")}).
Subsequent calls to this function will use the previously downloaded matrix/fragments, given that the same combination of filtering and
subsetting has been performed previously.
The preparation of this matrix can be reproduced by running the internal function \code{prepare_demo_data()} with \code{directory} set to the BPCells data directory.

In the case that demo data is not pre-downloaded and demo data download fails, \code{prepare_demo_data()} will act
as a fallback.

Both the matrix from \code{get_demo_mat()} and the fragments from \code{get_demo_frags()}
may be removed by running \code{remove_demo_data()}.

Filtering using QC information on the fragments and matrix object chooses cells with at least 1000 reads, 1000 frags, and a minimum tss enrichment of 10.
Subsetting provides only genes and insertions on chromosomes 4 and 11.

\strong{Dimensions}:
\tabular{lll}{
\strong{Condition} \tab \strong{RNA matrix (features x cells)} \tab \strong{Fragments (chromosomes x cells)} \cr
Raw                \tab (36601 x 650165)                       \tab (39 x 462264)                            \cr
Filter             \tab (36601 x 2600)                         \tab (39 x 2600)                              \cr
Subset             \tab (3582 x 650165)                        \tab (2 x 462264)                             \cr
Filter + Subset    \tab (3582 x 2600)                          \tab (2 x 2600)                               \cr
}
\strong{Data size}:
\tabular{lll}{
\strong{Condition}          \tab \strong{RNA matrix (MB)} \tab \strong{Fragments (MB)} \cr
Raw                         \tab 31.9                     \tab 200                     \cr
Filter                      \tab 9.4                      \tab 137                     \cr
Subset                      \tab 18.3                     \tab 25.6                    \cr
Filter + Subset             \tab 1.2                      \tab 12.3                    \cr
}

\strong{Function Description}:
\itemize{
\item \code{get_demo_mat()}: Retrieve a demo \code{IterableMatrix} object representing the 10X Genomics PBMC 3k dataset.
}

\itemize{
\item \code{get_demo_frags()}: Retrieve a demo \code{IterableFragments} object representing the 10X Genomics PBMC 3k dataset.
}

\itemize{
\item \code{remove_demo_data()}: Remove the demo data from the BPCells data directory.
}
}
\examples{
#######################################################################
## get_demo_mat() example
get_demo_mat()
#######################################################################
#######################################################################
## get_demo_frags() example
get_demo_frags()
#######################################################################
#######################################################################
## remove_demo_data() example
remove_demo_data()


## Demo data folder is now empty
data_dir <- file.path(tools::R_user_dir("BPCells", which = "data"), "demo_data")
list.files(data_dir)
#######################################################################
}
