<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Changelog â€¢ BPCells</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Changelog"><script defer data-domain="benparks.net" src="https://plausible.benparks.net/js/visit-counts.js"></script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">BPCells</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.2.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/pbmc3k.html">Get Started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/pbmc3k.html">Basic tutorial</a></li>
    <li><a class="dropdown-item" href="../articles/web-only/benchmarks.html">Performance Benchmarks</a></li>
    <li><a class="dropdown-item" href="../articles/web-only/how-it-works.html">How BPCells works</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/index.html">More articles...</a></li>
  </ul></li>
<li class="active nav-item"><a class="nav-link" href="../news/index.html">News</a></li>
<li class="nav-item"><a class="nav-link" href="../python/index.html"><span class="fa fa-arrow-up-right-from-square"></span> Python Docs</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/bnprks/BPCells" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-news">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Changelog</h1>
      <small>Source: <a href="https://github.com/bnprks/BPCells/tree/main/r/NEWS.md" class="external-link"><code>NEWS.md</code></a></small>
    </div>

    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.0" id="bpcells-10-roadmap">BPCells 1.0 Roadmap<a class="anchor" aria-label="anchor" href="#bpcells-10-roadmap"></a></h2>
<ul><li>
<del>Parallelization</del> (basic support complete. See below)</li>
<li>Native python library (re-using C++ backend)</li>
<li>Peak-gene correlations</li>
<li>MACS peak calling</li>
</ul><p>Contributions welcome :)</p>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.2.1" id="bpcells-021-main-branch---in-progress">BPCells 0.2.1 (main branch - in progress)<a class="anchor" aria-label="anchor" href="#bpcells-021-main-branch---in-progress"></a></h2>
<div class="section level3">
<h3 id="features-0-2-1">Features<a class="anchor" aria-label="anchor" href="#features-0-2-1"></a></h3>
<ul><li>
<code><a href="../reference/apply_by_row.html">apply_by_col()</a></code> and <code><a href="../reference/apply_by_row.html">apply_by_row()</a></code> allow providing custom R functions to compute per row/col summaries. In initial tests calculating row/col means using R functions is ~2x slower than the C++-based implementation but memory usage remains low.</li>
<li>Add <code><a href="../reference/IterableMatrix-methods.html">rowMaxs()</a></code> and <code><a href="../reference/IterableMatrix-methods.html">colMaxs()</a></code> functions, which return the maximum value in each row or column of a matrix. If <code>matrixStats</code> or <code>MatrixGenerics</code> packages are installed, <code><a href="../reference/IterableMatrix-methods.html">BPCells::rowMaxs()</a></code> will fall back to their implementations for non-BPCells objects. Thanks to <a href="https://github.com/immanuelazn" class="external-link">@immanuelazn</a> for their first contribution as a new lab hire!</li>
<li>Add <code><a href="../reference/regress_out.html">regress_out()</a></code> to allow removing unwanted sources of variation via least squares linear regression models. Thanks to <a href="https://github.com/ycli1995" class="external-link">@ycli1995</a> for pull request <a href="https://github.com/bnprks/BPCells/issues/110" class="external-link">#110</a></li>
<li>Add <code><a href="../reference/trackplot_genome_annotation.html">trackplot_genome_annotation()</a></code> for plotting peaks, with options for directional arrows, colors, labels, and peak widths. (pull request <a href="https://github.com/bnprks/BPCells/issues/113" class="external-link">#113</a>)</li>
<li>Add MACS2/3 input creation and peak calling through <code><a href="../reference/call_peaks_macs.html">call_peaks_macs()</a></code>(pull request <a href="https://github.com/bnprks/BPCells/issues/118" class="external-link">#118</a>). Note, renamed from <code>call_macs_peaks()</code> in pull request <a href="https://github.com/bnprks/BPCells/issues/143" class="external-link">#143</a></li>
<li>Add <code><a href="../reference/IterableMatrix-methods.html">rowQuantiles()</a></code> and <code><a href="../reference/IterableMatrix-methods.html">colQuantiles()</a></code> functions, which return the quantiles of each row/column of a matrix. Currently <code><a href="../reference/IterableMatrix-methods.html">rowQuantiles()</a></code> only works on row-major matrices and <code><a href="../reference/IterableMatrix-methods.html">colQuantiles()</a></code> only works on col-major matrices. If <code>matrixStats</code> or <code>MatrixGenerics</code> packages are installed, <code><a href="../reference/IterableMatrix-methods.html">BPCells::colQuantiles()</a></code> will fall back to their implementations for non-BPCells objects. (pull request <a href="https://github.com/bnprks/BPCells/issues/128" class="external-link">#128</a>)</li>
<li>Add <code><a href="../reference/pseudobulk_matrix.html">pseudobulk_matrix()</a></code> which allows pseudobulk aggregation by <code>sum</code> or <code>mean</code> and calculation of per-pseudobulk <code>variance</code> and <code>nonzero</code> statistics for each gene (pull request <a href="https://github.com/bnprks/BPCells/issues/128" class="external-link">#128</a>)</li>
</ul></div>
<div class="section level3">
<h3 id="improvements-0-2-1">Improvements<a class="anchor" aria-label="anchor" href="#improvements-0-2-1"></a></h3>
<ul><li>
<code><a href="../reference/trackplot_loop.html">trackplot_loop()</a></code> now accepts discrete color scales</li>
<li>
<code><a href="../reference/trackplot_combine.html">trackplot_combine()</a></code> now has smarter layout logic for margins, as well as detecting when plots are being combined that cover different genomic regions. (pull request <a href="https://github.com/bnprks/BPCells/issues/116" class="external-link">#116</a>)</li>
<li>
<code><a href="../reference/select_cells.html">select_cells()</a></code> and <code><a href="../reference/select_chromosomes.html">select_chromosomes()</a></code> now also allow using a logical mask for selection. (pull request <a href="https://github.com/bnprks/BPCells/issues/117" class="external-link">#117</a>)</li>
<li>BPCells installation can now also be configured by setting the <code>LDFLAGS</code> or <code>CFLAGS</code> as environment variables in addition to setting them in <code>~/.R/Makevars</code> (pull request <a href="https://github.com/bnprks/BPCells/issues/124" class="external-link">#124</a>)</li>
<li>
<code><a href="../reference/open_matrix_anndata_hdf5.html">open_matrix_anndata_hdf5()</a></code> now supports reading AnnData matrices in the dense format. (pull request <a href="https://github.com/bnprks/BPCells/issues/146" class="external-link">#146</a>)</li>
<li>
<code><a href="../reference/cluster.html">cluster_graph_leiden()</a></code> now has better defaults that produce reasonable cluster counts regardless of dataset size. (pull request <a href="https://github.com/bnprks/BPCells/issues/147" class="external-link">#147</a>)</li>
</ul></div>
<div class="section level3">
<h3 id="bug-fixes-0-2-1">Bug-fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-0-2-1"></a></h3>
<ul><li>Fixed error message when a matrix is too large to be converted to dgCMatrix. (Thanks to <a href="https://github.com/RookieA1" class="external-link">@RookieA1</a> for reporting issue <a href="https://github.com/bnprks/BPCells/issues/95" class="external-link">#95</a>)</li>
<li>Fixed forgetting dimnames when subsetting after certain sets of operations. (Thanks to <a href="https://github.com/Yunuuuu" class="external-link">@Yunuuuu</a> for reporting issues <a href="https://github.com/bnprks/BPCells/issues/97" class="external-link">#97</a> and <a href="https://github.com/bnprks/BPCells/issues/100" class="external-link">#100</a>)</li>
<li>Fixed plotting crashes when running <code><a href="../reference/trackplot_coverage.html">trackplot_coverage()</a></code> with fragments from a single cluster. (Thanks to <a href="https://github.com/sjessa" class="external-link">@sjessa</a> for directly reporting this bug and coming up with a fix)</li>
<li>Fixed issues with <code><a href="../reference/trackplot_coverage.html">trackplot_coverage()</a></code> when called with ranges less than 500 bp in length (Thanks to <a href="https://github.com/bettybliu" class="external-link">@bettybliu</a> for directly reporting this bug.)</li>
<li>Fix Rcpp warning created when handling compressed matrices with only one non-zero entry (pull request <a href="https://github.com/bnprks/BPCells/issues/123" class="external-link">#123</a>)</li>
<li>Fixed discrepancy between default ArchR and BPCells peak calling insertion method, where BPCells defaulted to only using the start of each fragment as opposed to ArchRâ€™s method of using both start and end sites of fragments (pull request <a href="https://github.com/bnprks/BPCells/issues/143" class="external-link">#143</a>)</li>
<li>Fix error in <code><a href="../reference/tile_matrix.html">tile_matrix()</a></code> with fragment mode (pull request <a href="https://github.com/bnprks/BPCells/issues/141" class="external-link">#141</a>)</li>
<li>Fix precision bug in <code><a href="../reference/sctransform_pearson.html">sctransform_pearson()</a></code> on ARM architecture (pull request <a href="https://github.com/bnprks/BPCells/issues/141" class="external-link">#141</a>)</li>
<li>Fix type-confusion error when <code><a href="../reference/pseudobulk_matrix.html">pseudobulk_matrix()</a></code> gets an integer matrix (pull request <a href="https://github.com/bnprks/BPCells/issues/174" class="external-link">#174</a>)</li>
</ul></div>
<div class="section level3">
<h3 id="deprecations-0-2-1">Deprecations<a class="anchor" aria-label="anchor" href="#deprecations-0-2-1"></a></h3>
<ul><li>
<code><a href="../reference/trackplot_coverage.html">trackplot_coverage()</a></code> <code>legend_label</code> argument is now ignored, as the color legend is no longer shown by default for coverage plots.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.2.0" id="bpcells-020-6142024">BPCells 0.2.0 (6/14/2024)<a class="anchor" aria-label="anchor" href="#bpcells-020-6142024"></a></h2>
<p>We are finally declaring a new release version, covering a large amount of changes and improvements over the past year. Among the major features here are parallelization options for <code><a href="../reference/svds.html">svds()</a></code> and <code><a href="../reference/matrix_stats.html">matrix_stats()</a></code>, improved genomic track plots, and runtime CPU feature detection for SIMD code (enables higher performance, more portable builds). Full details of changes below.</p>
<p>This version also comes with a new installation path, which is done in preparation for a future Python package release. (So we can have one folder for R and one for Python, rather than having all the R files sit in the root folder). This is a breaking change and requires a slightly modified installation command.</p>
<p>Thanks to <a href="https://github.com/brgew" class="external-link">@brgew</a>, <a href="https://github.com/ycli1995" class="external-link">@ycli1995</a>, and <a href="https://github.com/Yunuuuu" class="external-link">@Yunuuuu</a> for pull requests that contributed to this release, as well as all users who submitted github issues to help identify and fix bugs.</p>
<div class="section level3">
<h3 id="breaking-changes-0-2-0">Breaking changes<a class="anchor" aria-label="anchor" href="#breaking-changes-0-2-0"></a></h3>
<ul><li>Installation location has changed, to make room for a future python package release. New installs will have to use <code>remotes::install_github("bnprks/BPCells/r")</code> (note the additional <code>/r</code>)
<ul><li>r-universe mirrors will have to add <code>"subdir": "r"</code> to their <code>packages.json</code> config.</li>
</ul></li>
<li>New slots have been added to 10x matrix objects, so any saved RDS files may need to have their 10x matrix inputs re-opened and replaced by calling <code><a href="../reference/all_matrix_inputs.html">all_matrix_inputs()</a></code>. Outside of loading old RDS files no changes should be needed.</li>
<li>
<code><a href="../reference/trackplot_gene.html">trackplot_gene()</a></code> now returns a plot with a facet label to match the new trackplot system. This label can be removed by by calling <code>trackplot_gene(...) + ggplot2::facet_null()</code> to be equivalent to the old functionâ€™s output.</li>
</ul></div>
<div class="section level3">
<h3 id="deprecations-0-2-0">Deprecations<a class="anchor" aria-label="anchor" href="#deprecations-0-2-0"></a></h3>
<ul><li>
<code><a href="../reference/draw_trackplot_grid.html">draw_trackplot_grid()</a></code> deprecated, replaced by <code><a href="../reference/trackplot_combine.html">trackplot_combine()</a></code> with simplified arguments</li>
<li>
<code><a href="../reference/trackplot_bulk.html">trackplot_bulk()</a></code> has been deprecated, replaced by <code><a href="../reference/trackplot_coverage.html">trackplot_coverage()</a></code> with equivalent functionality</li>
<li>The old function names will output deprecation warnings, but otherwise work as before.</li>
</ul></div>
<div class="section level3">
<h3 id="features-0-2-0">Features<a class="anchor" aria-label="anchor" href="#features-0-2-0"></a></h3>
<ul><li>New <code><a href="../reference/svds.html">svds()</a></code> function, based on the excellent Spectra C++ library (used in RSpectra) by Yixuan Qiu. This should ensure lower memory usage compared to <code>irlba</code>, while achieving similar speed + accuracy.</li>
<li>Limited parallelization is now supported. This is easiest to use via the <code>threads</code> argument to <code><a href="../reference/matrix_stats.html">matrix_stats()</a></code> and <code><a href="../reference/svds.html">svds()</a></code>.
<ul><li>All normalizations are supported, but a few operations like <code><a href="../reference/marker_features.html">marker_features()</a></code> and writing a matrix to disk remain single-threaded.</li>
<li>Running <code><a href="../reference/svds.html">svds()</a></code> with many threads on gene-major matrices can result in high memory usage for now. This problem is not present for cell-major matrices.</li>
</ul></li>
<li>Reading text-based MatrixMarket inputs (e.g.Â from 10x or Parse) is now supported via <code><a href="../reference/import_matrix_market.html">import_matrix_market()</a></code> and the convenience function <code><a href="../reference/import_matrix_market.html">import_matrix_market_10x()</a></code>. Our implementation uses disk-backed sorting to allow importing large files with low memory usage.</li>
<li>Added <code><a href="../reference/binarize.html">binarize()</a></code> function and associated generics <code>&lt;</code>, <code>&lt;=</code>, <code>&gt;</code>, and <code>&gt;=</code>. This only supports comparison with non-negative numbers currently. (Thanks to contribution from <a href="https://github.com/brgew" class="external-link">@brgew</a>)</li>
<li>Added <code><a href="https://rdrr.io/r/base/Round.html" class="external-link">round()</a></code> matrix transformation (Thanks to contributions from <a href="https://github.com/brgew" class="external-link">@brgew</a>)</li>
<li>Add getter/setter function <code><a href="../reference/all_matrix_inputs.html">all_matrix_inputs()</a></code> to help enable relocating the underlying storage for BPCells matrix transform objects.</li>
<li>All hdf5-writing functions now support a <code>gzip_level</code> parameter, which will enable a shuffle + gzip filter for compression. This is generally much slower than bitpacking compression, but it adds improved storage options for files that must be read by outside programs. Thanks to <a href="https://github.com/ycli1995" class="external-link">@ycli1995</a> for submitting this improvement in pull <a href="https://github.com/bnprks/BPCells/issues/42" class="external-link">#42</a>.</li>
<li>AnnData export now supported via <code><a href="../reference/open_matrix_anndata_hdf5.html">write_matrix_anndata_hdf5()</a></code> (issue <a href="https://github.com/bnprks/BPCells/issues/49" class="external-link">#49</a>)</li>
<li>Re-licensed code base to use dual-licensed Apache V2 or MIT instead of GPLv3</li>
<li>Assigning to a subset is now supported (e.g.Â <code>m1[i,j] &lt;- m2</code>). Note that this does not modify data on disk. Instead, it uses a series of subsetting and concatenation operations to provide the <em>appearance</em> of overwriting the appropriate entries.</li>
<li>Added <code><a href="../reference/knn_graph.html">knn_to_geodesic_graph()</a></code>, which matches the Scanpy default construction for graph-based clustering</li>
<li>Add <code><a href="../reference/checksum.html">checksum()</a></code>, which allows for calculating an MD5 checksum of a matrix contents. Thanks to <a href="https://github.com/brgrew" class="external-link">@brgrew</a> for submitting this improvement in pull request <a href="https://github.com/bnprks/BPCells/issues/83" class="external-link">#83</a></li>
<li>
<code><a href="../reference/write_insertion_bedgraph.html">write_insertion_bedgraph()</a></code> allows exporting pseudobulk insertion data to bedgraph format</li>
</ul></div>
<div class="section level3">
<h3 id="improvements-0-2-0">Improvements<a class="anchor" aria-label="anchor" href="#improvements-0-2-0"></a></h3>
<ul><li>Merging fragments with <code><a href="https://rdrr.io/r/base/c.html" class="external-link">c()</a></code> now handles inputs with mismatched chromosome names.</li>
<li>Merging fragments is now 2-3.5x faster</li>
<li>SNN graph construction in <code><a href="../reference/knn_graph.html">knn_to_snn_graph()</a></code> should work more smoothly on large datasets due to C++ implementation</li>
<li>Reduced memory usage in <code><a href="../reference/marker_features.html">marker_features()</a></code> for samples with millions of cells and a large number of clusters to compare.</li>
<li>On Windows, increased the maximum number of files that can be simultaneously open. Previously, opening &gt;63 compressed counts matrices simultaneously would hit the limit. Now at least 1,000 simultaneous matrices should be possible.</li>
<li>Subsetting peak or tile matrices with <code>[</code> now propagates through so we always avoid computing parts of the peak/tile matrix that have been discarded by our subset. Subsetting a tile matrix will automatically convert into a peak matrix when possible for improved efficiency.</li>
<li>Subsetting RowBindMatrices and ColBindMatrices now propagates through so we avoid touching matrices with no selected indices</li>
<li>Added logic to help reduce cases where subsetting causes BPCells to fall back to a less efficient matrix-vector multiply algorithm. This affects most math transforms. As part of this, the filtering part of a subset will propagate to earlier transformation steps, while the reordering will not. Thanks to <a href="https://github.com/nimanouri-nm" class="external-link">@nimanouri-nm</a> for raising issue <a href="https://github.com/bnprks/BPCells/issues/65" class="external-link">#65</a> to fix a bug in the initial implementation.</li>
<li>Additional C++17 filesystem backwards compatibility that should allow slightly older compilers such as GCC 7.5 to build BPCells.</li>
<li>
<code><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix()</a></code> will produce integer matrices when appropriate (Thanks to <a href="https://github.com/Yunuuuu" class="external-link">@Yunuuuu</a> in pull <a href="https://github.com/bnprks/BPCells/issues/77" class="external-link">#77</a>)</li>
<li>10x HDF5 matrices can now read and write non-integer types when requested (Thanks to <a href="https://github.com/ycli1995" class="external-link">@ycli1995</a> in pull <a href="https://github.com/bnprks/BPCells/issues/75" class="external-link">#75</a>)</li>
<li>Old-style 10x files from cellranger v2 can now read multi-genome files, which are returned as a list (Thanks to <a href="https://github.com/ycli1995" class="external-link">@ycli1995</a> in pull <a href="https://github.com/bnprks/BPCells/issues/75" class="external-link">#75</a>)</li>
<li>Trackplots have received several improvements
<ul><li>Trackplots now use faceting to provide per-plot labels, leading to an easier-to-use <code><a href="../reference/trackplot_combine.html">trackplot_combine()</a></code>
</li>
<li>
<code><a href="../reference/trackplot_gene.html">trackplot_gene()</a></code> now draws arrows for the direction of transcription</li>
<li>
<code><a href="../reference/trackplot_loop.html">trackplot_loop()</a></code> is a new track type allows plotting interactions between genomic regions, for instance peak-gene correlations or loop calls from Hi-C</li>
<li>
<code><a href="../reference/trackplot_scalebar.html">trackplot_scalebar()</a></code> is added to show genomic scale</li>
<li>All trackplot functions now return ggplot objects with additional metadata stored for the plotting height of each track</li>
<li>Labels and heights for trackplots can be adjusted using <code><a href="../reference/trackplot_utils.html">set_trackplot_label()</a></code> and <code><a href="../reference/trackplot_utils.html">set_trackplot_height()</a></code>
</li>
<li>The getting started pbmc 3k vignette now includes the updated trackplot APIs in its final example</li>
</ul></li>
<li>Add <code><a href="../reference/IterableMatrix-methods.html">rowVars()</a></code> and <code><a href="../reference/IterableMatrix-methods.html">colVars()</a></code> functions, as convenience wrappers around <code><a href="../reference/matrix_stats.html">matrix_stats()</a></code>. If <code>matrixStats</code> or <code>MatrixGenerics</code> packages are installed, <code><a href="../reference/IterableMatrix-methods.html">BPCells::rowVars()</a></code> will fall back to their implementations for non-BPCells objects. Unfortunately, <code><a href="https://rdrr.io/pkg/matrixStats/man/rowVars.html" class="external-link">matrixStats::rowVars()</a></code> is not generic, so either <code><a href="../reference/IterableMatrix-methods.html">BPCells::rowVars()</a></code> or <code><a href="../reference/IterableMatrix-methods.html">BPCells::colVars()</a></code>
</li>
<li>Optimize mean and variance calculations for matrices added to a per-row or per-column constant.</li>
<li>Migrate SIMD code to use <a href="https://github.com/google/highway" class="external-link"><code>highway</code></a>.
<ul><li>Adds run-time detection of CPU features to eliminate architecture-specific compilation</li>
<li>For now, the <code>Pow</code> SIMD implementation is removed, but <code>Square</code> gets a new SIMD implementation</li>
<li>Empirically, most operations using SIMD math instructions are about 2x faster. This includes <code><a href="https://rdrr.io/r/base/Log.html" class="external-link">log1p()</a></code>, and <code><a href="../reference/sctransform_pearson.html">sctransform_pearson()</a></code>
</li>
<li>Minor speedups on dense-sparse matrix multiply functions (1.1-1.5x faster)</li>
</ul></li>
</ul></div>
<div class="section level3">
<h3 id="bug-fixes-0-2-0">Bug-fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-0-2-0"></a></h3>
<ul><li>Fixed a few fragment transforms where using <code>chrNames(frags) &lt;- val</code> or <code>cellNames(frags) &lt;- val</code> could cause downstream errors.</li>
<li>Fixed errors in <code><a href="../reference/transpose_storage_order.html">transpose_storage_order()</a></code> for matrices with &gt;4 billion non-zero entries.</li>
<li>Fixed error in <code><a href="../reference/transpose_storage_order.html">transpose_storage_order()</a></code> for matrices with no non-zero entries.</li>
<li>Fixed bug writing fragment files with &gt;512 chromosomes.</li>
<li>Fixed bug when reading fragment files with &gt;4 billion fragments.</li>
<li>Fixed file permissions errors when using read-only hdf5 files (Issue <a href="https://github.com/bnprks/BPCells/issues/26" class="external-link">#26</a> reported thanks to <a href="https://github.com/ttumkaya" class="external-link">@ttumkaya</a>)</li>
<li>Renaming <code><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames()</a></code> or <code><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames()</a></code> is now propagated when saving matrices (Issue <a href="https://github.com/bnprks/BPCells/issues/29" class="external-link">#29</a> reported thanks to <a href="https://github.com/realzehuali" class="external-link">@realzehuali</a>, with an additional fix after report thanks to <a href="https://github.com/Dario-Rocha" class="external-link">@Dario-Rocha</a>)</li>
<li>Fixed 64-bit integer overflow (!) that could cause incorrect p-value calculations in <code><a href="../reference/marker_features.html">marker_features()</a></code> for features with more than 2.6 million zeros.</li>
<li>Improved robustness of the Windows installation process for setups that do not need the -lsz linker flag to compile hdf5</li>
<li>Fixed possible memory safety bug where wrapped R objects (such as dgCMatrix) could be potentially garbage collected while C++ was still trying to access the data in rare circumstances.</li>
<li>Fixed case when dimnames were not preserved when calling <code><a href="../reference/convert_matrix_type.html">convert_matrix_type()</a></code> twice in a row such that it cancels out (e.g.Â double -&gt; uint32_t -&gt; double). Thanks to <a href="https://github.com/brgrew" class="external-link">@brgrew</a> reporting issue <a href="https://github.com/bnprks/BPCells/issues/43" class="external-link">#43</a></li>
<li>Caused and fixed issue resulting in unusably slow performance reading matrices from HDF5 files. Broken versions range from commit 21f8dcf until the fix in 3711a40 (October 18-November 3, 2023). Thanks to <a href="https://github.com/abhiachoudhary" class="external-link">@abhiachoudhary</a> for reporting this in issue <a href="https://github.com/bnprks/BPCells/issues/53" class="external-link">#53</a></li>
<li>Fixed error with <code><a href="../reference/svds.html">svds()</a></code> not handling row-major matrices correctly. Thanks to <a href="https://github.com/ycli1995" class="external-link">@ycli1995</a> for reporting this in issue <a href="https://github.com/bnprks/BPCells/issues/55" class="external-link">#55</a></li>
<li>Fixed error with row/col name handling for AnnData matrices. Thanks to <a href="https://github.com/lisch7" class="external-link">@lisch7</a> for reporting this in issue <a href="https://github.com/bnprks/BPCells/issues/57" class="external-link">#57</a></li>
<li>Fixed error with merging matrices of different data types. Thanks to <a href="https://github.com/Yunuuuu" class="external-link">@Yunuuuu</a> for identifying the issue and providing a fix (<a href="https://github.com/bnprks/BPCells/issues/68" class="external-link">#68</a> and <a href="https://github.com/bnprks/BPCells/issues/70" class="external-link">#70</a>)</li>
<li>Fixed issue with losing dimnames on subset assignment <code>[&lt;-</code>. Thanks to <a href="https://github.com/Yunuuuu" class="external-link">@Yunuuuu</a> for identifying the issue <a href="https://github.com/bnprks/BPCells/issues/67" class="external-link">#67</a></li>
<li>Fixed incorrect results with some cases of scaling matrix after shifting. Thanks to <a href="https://github.com/Yunuuuu" class="external-link">@Yunuuuu</a> for identifying the issue <a href="https://github.com/bnprks/BPCells/issues/72" class="external-link">#72</a></li>
<li>Fixed infinite loop bug when calling <code><a href="../reference/transpose_storage_order.html">transpose_storage_order()</a></code> on a densely-transformed matrix. Thanks to <a href="https://github.com/Yunuuuu" class="external-link">@Yunuuuu</a> for reporting this in issue <a href="https://github.com/bnprks/BPCells/issues/71" class="external-link">#71</a></li>
<li>h5ad outputs will now subset properly when loaded by the Python anndata package (Thanks to issue described by <a href="https://github.com/ggruenhagen3" class="external-link">@ggruenhagen3</a> in issue <a href="https://github.com/bnprks/BPCells/issues/49" class="external-link">#49</a> and fixed by <a href="https://github.com/ycli1995" class="external-link">@ycli1995</a> in pull <a href="https://github.com/bnprks/BPCells/issues/81" class="external-link">#81</a>)</li>
<li>Disk-backed fragment objects now load via absolute path, matching the behavior of matrices and making it so objects loaded via <code><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS()</a></code> can be used from different working directories.</li>
<li>
<code>footprints()</code> now respects user interrupts via Ctrl-C</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.1.0" id="bpcells-010-472023">BPCells 0.1.0 (4/7/2023)<a class="anchor" aria-label="anchor" href="#bpcells-010-472023"></a></h2>
<div class="section level3">
<h3 id="features-0-1-0">Features<a class="anchor" aria-label="anchor" href="#features-0-1-0"></a></h3>
<ul><li>ATAC-seq Analysis
<ul><li>Reading/writing 10x fragment files on disk</li>
<li>Reading/writing compressed fragments on disk (in folder or hdf5 group)</li>
<li>Interconversion of fragments objects with GRanges / data.frame</li>
<li>Merging of multiple source fragment files transparently at run time</li>
<li>Calculation of Cell x Peak matrices, and Cell x Tile matrices</li>
<li>ArchR-compatible QC calculations</li>
<li>ArchR-compatible gene activity score calculations</li>
<li>Filtering fragments by chromosmes, cells, lengths, or genomic region</li>
<li>Fast peak calling approximation via overlapping tiles</li>
</ul></li>
<li>Single cell matrices
<ul><li>Conversion to/from R sparse matrices</li>
<li>Read-write access to 10x hdf5 feature matrices, and read-only access to AnnData files</li>
<li>Reading/writing of compressed matrices on disk (in folder or hdf5 group)</li>
<li>Support for integer or single/double-precision floating point matrices on disk</li>
<li>Fast transposition of storage order, to switch between indexing by cell or by gene/feature.</li>
<li>Concatenation of multiple source matrix files transparently at run time</li>
<li>Single-pass calculation of row/column mean and variance</li>
<li>Wilcoxon marker feature calculation</li>
<li>Transparent handling of vector <code>+</code>, <code>-</code>, <code>*</code>, <code>/</code>, and <code>log1p</code> for streaming normalization, along with other less common operations. This allows implementation of ATAC-seq LSI and Seurat default normalization, along with most published log-based normalizations.</li>
<li>SCTransform pearson residual calculation</li>
<li>Multiplication of sparse matrices</li>
</ul></li>
<li>Single cell plotting utilities
<ul><li>Read count knee cutoffs</li>
<li>UMAP embeddings</li>
<li>Dot plots</li>
<li>Transcription factor footprinting / TSS profile plotting</li>
<li>Fragments vs.Â TSS Enrichment ATAC-seq QC plot</li>
<li>Pseudobulk genome track plots, with gene annotation plots</li>
</ul></li>
<li>Additional utility functions
<ul><li>Matching gene symbols/IDs to canonical symbols</li>
<li>Download transcript annotations from Gencode or GTF files</li>
<li>Download + parse UCSC chromosome sizes</li>
<li>Parse peak files BED format; Download ENCODE blacklist region</li>
<li>Wrappers for knn graph calculation + clustering</li>
</ul></li>
</ul><p>Note: All operations interoperate with all storage formats. For example, all matrix operations can be applied directly to an AnnData or 10x matrix file. In many cases the bitpacking-compressed formats will provide performance/space advantages, but are not required to use the computations.</p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Created by Benjamin Parks.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>



  <img src="https://plausible.benparks.net/flask-plausible/bpcells-docs.png" style="position:absolute;"></body></html>

