<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Changelog • BPCells</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Changelog"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><script defer data-domain="benparks.net" src="https://plausible.benparks.net/js/visit-counts.js"></script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">BPCells</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/pbmc3k.html">Get Started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/pbmc3k.html">Basic tutorial</a>
    <a class="dropdown-item" href="../articles/web-only/benchmarks.html">Performance Benchmarks</a>
    <a class="dropdown-item" href="../articles/web-only/how-it-works.html">How BPCells works</a>
    <div class="dropdown-divider"></div>
    <a class="dropdown-item" href="../articles/index.html">More articles...</a>
  </div>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../news/index.html">News</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/bnprks/BPCells/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-news">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Changelog</h1>
      <small>Source: <a href="https://github.com/bnprks/BPCells/blob/HEAD/NEWS.md" class="external-link"><code>NEWS.md</code></a></small>
    </div>

    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.0" id="bpcells-10-roadmap">BPCells 1.0 Roadmap<a class="anchor" aria-label="anchor" href="#bpcells-10-roadmap"></a></h2>
<ul><li>Parallelization</li>
<li>Native python library (re-using C++ backend)</li>
<li>Peak-gene correlations</li>
<li>MACS peak calling</li>
</ul><p>Contributions welcome :)</p>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.2.0" id="bpcells-020-unreleased---github-main-branch">BPCells 0.2.0 (unreleased - github main branch)<a class="anchor" aria-label="anchor" href="#bpcells-020-unreleased---github-main-branch"></a></h2>
<div class="section level3">
<h3 id="features-0-2-0">Features<a class="anchor" aria-label="anchor" href="#features-0-2-0"></a></h3>
<ul><li>Reading text-based MatrixMarket inputs (e.g. from 10x or Parse) is now supported via <code><a href="../reference/import_matrix_market.html">import_matrix_market()</a></code> and the convenience function <code><a href="../reference/import_matrix_market.html">import_matrix_market_10x()</a></code>. Our implementation uses disk-backed sorting to allow importing large files with low memory usage.</li>
<li>Merging fragments with <code><a href="https://rdrr.io/r/base/c.html" class="external-link">c()</a></code> now handles inputs with mismatched chromosome names.</li>
</ul></div>
<div class="section level3">
<h3 id="bug-fixes-0-2-0">Bug-fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-0-2-0"></a></h3>
<ul><li>Fixed a few fragment transforms where using <code>chrNames(frags) &lt;- val</code> or <code>cellNames(frags) &lt;- val</code> could cause downstream errors.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.1.0" id="bpcells-010">BPCells 0.1.0<a class="anchor" aria-label="anchor" href="#bpcells-010"></a></h2>
<div class="section level3">
<h3 id="features-0-1-0">Features<a class="anchor" aria-label="anchor" href="#features-0-1-0"></a></h3>
<ul><li>ATAC-seq Analysis
<ul><li>Reading/writing 10x fragment files on disk</li>
<li>Reading/writing compressed fragments on disk (in folder or hdf5 group)</li>
<li>Interconversion of fragments objects with GRanges / data.frame</li>
<li>Merging of multiple source fragment files transparently at run time</li>
<li>Calculation of Cell x Peak matrices, and Cell x Tile matrices</li>
<li>ArchR-compatible QC calculations</li>
<li>ArchR-compatible gene activity score calculations</li>
<li>Filtering fragments by chromosmes, cells, lengths, or genomic region</li>
<li>Fast peak calling approximation via overlapping tiles</li>
</ul></li>
<li>Single cell matrices
<ul><li>Conversion to/from R sparse matrices</li>
<li>Read-write access to 10x hdf5 feature matrices, and read-only access to AnnData files</li>
<li>Reading/writing of compressed matrices on disk (in folder or hdf5 group)</li>
<li>Support for integer or single/double-precision floating point matrices on disk</li>
<li>Fast transposition of storage order, to switch between indexing by cell or by gene/feature.</li>
<li>Concatenation of multiple source matrix files transparently at run time</li>
<li>Single-pass calculation of row/column mean and variance</li>
<li>Wilcoxon marker feature calculation</li>
<li>Transparent handling of vector <code>+</code>, <code>-</code>, <code>*</code>, <code>/</code>, and <code>log1p</code> for streaming normalization, along with other less common operations. This allows implementation of ATAC-seq LSI and Seurat default normalization, along with most published log-based normalizations.</li>
<li>SCTransform pearson residual calculation</li>
<li>Multiplication of sparse matrices</li>
</ul></li>
<li>Single cell plotting utilities
<ul><li>Read count knee cutoffs</li>
<li>UMAP embeddings</li>
<li>Dot plots</li>
<li>Transcription factor footprinting / TSS profile plotting</li>
<li>Fragments vs. TSS Enrichment ATAC-seq QC plot</li>
<li>Pseudobulk genome track plots, with gene annotation plots</li>
</ul></li>
<li>Additional utility functions
<ul><li>Matching gene symbols/IDs to canonical symbols</li>
<li>Download transcript annotations from Gencode or GTF files</li>
<li>Download + parse UCSC chromosome sizes</li>
<li>Parse peak files BED format; Download ENCODE blacklist region</li>
<li>Wrappers for knn graph calculation + clustering</li>
</ul></li>
</ul><p>Note: All operations interoperate with all storage formats. For example, all matrix operations can be applied directly to an AnnData or 10x matrix file. In many cases the bitpacking-compressed formats will provide performance/space advantages, but are not required to use the computations.</p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Benjamin Parks.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  <img src="https://plausible.benparks.net/flask-plausible/bpcells-docs.png" style="position:absolute;"></body></html>

