<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Calculate GeneActivityScores — gene_score_weights_archr • BPCells</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Calculate GeneActivityScores — gene_score_weights_archr"><meta name="description" content="Gene activity scores can be calculated as a distance-weighted sum of per-tile accessibility.
The tile weights for each gene can be represented as a sparse matrix of dimension genes x tiles.
If we multiply this weight matrix by a corresponding tile matrix (tiles x cells), then we can
get a gene activity score matrix of genes x cells. gene_score_weights_archr() calculates the
weight matrix (best if you have a pre-computed tile matrix), while gene_score_archr() provides
a easy-to-use wrapper."><meta property="og:description" content="Gene activity scores can be calculated as a distance-weighted sum of per-tile accessibility.
The tile weights for each gene can be represented as a sparse matrix of dimension genes x tiles.
If we multiply this weight matrix by a corresponding tile matrix (tiles x cells), then we can
get a gene activity score matrix of genes x cells. gene_score_weights_archr() calculates the
weight matrix (best if you have a pre-computed tile matrix), while gene_score_archr() provides
a easy-to-use wrapper."><script defer data-domain="benparks.net" src="https://plausible.benparks.net/js/visit-counts.js"></script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">BPCells</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.2.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/pbmc3k.html">Get Started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/pbmc3k.html">Basic tutorial</a></li>
    <li><a class="dropdown-item" href="../articles/web-only/benchmarks.html">Performance Benchmarks</a></li>
    <li><a class="dropdown-item" href="../articles/web-only/how-it-works.html">How BPCells works</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/index.html">More articles...</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">News</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/bnprks/BPCells" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Calculate GeneActivityScores</h1>
      <small class="dont-index">Source: <a href="https://github.com/bnprks/BPCells/tree/main/r/R/geneScores.R" class="external-link"><code>R/geneScores.R</code></a></small>
      <div class="d-none name"><code>gene_scores.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Gene activity scores can be calculated as a distance-weighted sum of per-tile accessibility.
The tile weights for each gene can be represented as a sparse matrix of dimension genes x tiles.
If we multiply this weight matrix by a corresponding tile matrix (tiles x cells), then we can
get a gene activity score matrix of genes x cells. <code>gene_score_weights_archr()</code> calculates the
weight matrix (best if you have a pre-computed tile matrix), while <code>gene_score_archr()</code> provides
a easy-to-use wrapper.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">gene_score_weights_archr</span><span class="op">(</span></span>
<span>  <span class="va">genes</span>,</span>
<span>  <span class="va">chromosome_sizes</span>,</span>
<span>  blacklist <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  tile_width <span class="op">=</span> <span class="fl">500</span>,</span>
<span>  gene_name_column <span class="op">=</span> <span class="st">"gene_id"</span>,</span>
<span>  addArchRBug <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">gene_score_archr</span><span class="op">(</span></span>
<span>  <span class="va">fragments</span>,</span>
<span>  <span class="va">genes</span>,</span>
<span>  <span class="va">chromosome_sizes</span>,</span>
<span>  blacklist <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  tile_width <span class="op">=</span> <span class="fl">500</span>,</span>
<span>  gene_name_column <span class="op">=</span> <span class="st">"gene_id"</span>,</span>
<span>  addArchRBug <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  tile_max_count <span class="op">=</span> <span class="fl">4</span>,</span>
<span>  scale_factor <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>  tile_matrix_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">"gene_score_tile_mat"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-genes">genes<a class="anchor" aria-label="anchor" href="#arg-genes"></a></dt>
<dd><p>Gene coordinates given as GRanges, data.frame, or list. See <code><a href="genomic-ranges-like.html">help("genomic-ranges-like")</a></code> for details on format and coordinate systems. Required attributes:</p><ul><li><p><code>chr</code>, <code>start</code>, <code>end</code>: genomic position</p></li>
<li><p><code>strand</code>: +/- or TRUE/FALSE for positive or negative strand</p></li>
</ul></dd>


<dt id="arg-chromosome-sizes">chromosome_sizes<a class="anchor" aria-label="anchor" href="#arg-chromosome-sizes"></a></dt>
<dd><p>Chromosome start and end coordinates given as GRanges, data.frame, or list. See <code><a href="genomic-ranges-like.html">help("genomic-ranges-like")</a></code> for details on format and coordinate systems. Required attributes:</p><ul><li><p><code>chr</code>, <code>start</code>, <code>end</code>: genomic position</p></li>
</ul><p>See <code><a href="read_ucsc_chrom_sizes.html">read_ucsc_chrom_sizes()</a></code>.</p></dd>


<dt id="arg-blacklist">blacklist<a class="anchor" aria-label="anchor" href="#arg-blacklist"></a></dt>
<dd><p>Regions to exclude from calculations, given as GRanges, data.frame, or list. See <code><a href="genomic-ranges-like.html">help("genomic-ranges-like")</a></code> for details on format and coordinate systems. Required attributes:</p><ul><li><p><code>chr</code>, <code>start</code>, <code>end</code>: genomic position</p></li>
</ul></dd>


<dt id="arg-tile-width">tile_width<a class="anchor" aria-label="anchor" href="#arg-tile-width"></a></dt>
<dd><p>Size of tiles to consider</p></dd>


<dt id="arg-gene-name-column">gene_name_column<a class="anchor" aria-label="anchor" href="#arg-gene-name-column"></a></dt>
<dd><p>If not NULL, a column name of <code>genes</code> to use as row names</p></dd>


<dt id="arg-addarchrbug">addArchRBug<a class="anchor" aria-label="anchor" href="#arg-addarchrbug"></a></dt>
<dd><p>Replicate ArchR bug in handling nested genes</p></dd>


<dt id="arg-fragments">fragments<a class="anchor" aria-label="anchor" href="#arg-fragments"></a></dt>
<dd><p>Input fragments object</p></dd>


<dt id="arg-tile-max-count">tile_max_count<a class="anchor" aria-label="anchor" href="#arg-tile-max-count"></a></dt>
<dd><p>Maximum value in the tile counts matrix. If not null, tile counts higher than this will be clipped to <code>tile_max_count</code>.
Equivalent to <code>ceiling</code> argument of <code>ArchR::addGeneScoreMatrix()</code></p></dd>


<dt id="arg-scale-factor">scale_factor<a class="anchor" aria-label="anchor" href="#arg-scale-factor"></a></dt>
<dd><p>If not null, counts for each cell will be scaled to sum to <code>scale_factor</code>. Equivalent to <code>scaleTo</code> argument of <code>ArchR::addGeneScoreMatrix()</code></p></dd>


<dt id="arg-tile-matrix-path">tile_matrix_path<a class="anchor" aria-label="anchor" href="#arg-tile-matrix-path"></a></dt>
<dd><p>Path of a directory where the intermediate tile matrix will be saved</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p><strong>gene_score_weights_archr</strong></p>
<p>Weight matrix of dimension genes x tiles</p>
<p><strong>gene_score_archr</strong></p>
<p>Gene score matrix of dimension genes x cells.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p><strong>gene_score_weights_archr</strong>:</p>
<p>Given a set of tile coordinates and distances returned by <code><a href="gene_score_tiles_archr.html">gene_score_tiles_archr()</a></code>,
calculate a weight matrix of dimensions genes x tiles. This matrix can be
multiplied with a tile matrix to obtain ArchR-compatible gene activity scores.</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Created by Benjamin Parks.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer></div>



  <img src="https://plausible.benparks.net/flask-plausible/bpcells-docs.png" style="position:absolute;"></body></html>

