<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Genomic range formats — genomic-ranges-like • BPCells</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Genomic range formats — genomic-ranges-like"><meta name="description" content='BPCells accepts a flexible set of genomic ranges-like objects as input, either
GRanges, data.frame, lists, or character vectors. These objects must specify chromosome, start,
and end coordinates along with optional metadata about each range.
With the exception of GenomicRanges::GRanges objects, BPCells assumes all
objects use a zero-based, end-exclusive coordinate system (see below for details).
Valid Range-like objects


BPCells can interpret the following types as ranges:
list(), data.frame(), with columns:
chr: Character or factor of chromosome names
start: Start coordinates (0-based)
end: End coordinates (exclusive)
(optional) strand: "+"/"-" or TRUE/FALSE for pos/neg strand
(optional) Additional metadata as named list entries or data.frame columns

GenomicRanges::GRanges
start(x) is interpreted as a 1-based start coordinate
end(x) is interpreted as an inclusive end coordinate
strand(x): "*" entries are interpeted as postive strand
(optional) mcols(x) holds additional metadata

character
Given in format "chr1:1000-2000" or "chr1:1,000-2,000"
Uses 0-based, end-exclusive coordinate system
Cannot be used for ranges where additional metadata is required






Range coordinate systems


There are two main conventions for the coordinate systems:
One-based, end-inclusive ranges
The first base of a chromosome is numbered 1
The last base in a range is equal to the end coordinate
e.g. 1-5 describes the first 5 bases of the chromosome
Used in formats such as SAM, GTF
In BPCells, used when reading or writing GenomicRanges::GRanges objects


Zero-based, end-exclusive ranges
The first base of a chromosome is numbered 0
The last base in a range is one less than the end coordinate
e.g. 0-5 describes the first 5 bases of the chromosome
Used in formats such as BAM, BED
In BPCells, used for all other range objects



'><meta property="og:description" content='BPCells accepts a flexible set of genomic ranges-like objects as input, either
GRanges, data.frame, lists, or character vectors. These objects must specify chromosome, start,
and end coordinates along with optional metadata about each range.
With the exception of GenomicRanges::GRanges objects, BPCells assumes all
objects use a zero-based, end-exclusive coordinate system (see below for details).
Valid Range-like objects


BPCells can interpret the following types as ranges:
list(), data.frame(), with columns:
chr: Character or factor of chromosome names
start: Start coordinates (0-based)
end: End coordinates (exclusive)
(optional) strand: "+"/"-" or TRUE/FALSE for pos/neg strand
(optional) Additional metadata as named list entries or data.frame columns

GenomicRanges::GRanges
start(x) is interpreted as a 1-based start coordinate
end(x) is interpreted as an inclusive end coordinate
strand(x): "*" entries are interpeted as postive strand
(optional) mcols(x) holds additional metadata

character
Given in format "chr1:1000-2000" or "chr1:1,000-2,000"
Uses 0-based, end-exclusive coordinate system
Cannot be used for ranges where additional metadata is required






Range coordinate systems


There are two main conventions for the coordinate systems:
One-based, end-inclusive ranges
The first base of a chromosome is numbered 1
The last base in a range is equal to the end coordinate
e.g. 1-5 describes the first 5 bases of the chromosome
Used in formats such as SAM, GTF
In BPCells, used when reading or writing GenomicRanges::GRanges objects


Zero-based, end-exclusive ranges
The first base of a chromosome is numbered 0
The last base in a range is one less than the end coordinate
e.g. 0-5 describes the first 5 bases of the chromosome
Used in formats such as BAM, BED
In BPCells, used for all other range objects



'><script defer data-domain="benparks.net" src="https://plausible.benparks.net/js/visit-counts.js"></script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">BPCells</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.2.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/pbmc3k.html">Get Started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/pbmc3k.html">Basic tutorial</a></li>
    <li><a class="dropdown-item" href="../articles/web-only/benchmarks.html">Performance Benchmarks</a></li>
    <li><a class="dropdown-item" href="../articles/web-only/how-it-works.html">How BPCells works</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/index.html">More articles...</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">News</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/bnprks/BPCells" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Genomic range formats</h1>
      <small class="dont-index">Source: <a href="https://github.com/bnprks/BPCells/tree/main/r/R/BPCells-package.R" class="external-link"><code>R/BPCells-package.R</code></a></small>
      <div class="d-none name"><code>genomic-ranges-like.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>BPCells accepts a flexible set of genomic ranges-like objects as input, either
GRanges, data.frame, lists, or character vectors. These objects must specify chromosome, start,
and end coordinates along with optional metadata about each range.
With the exception of <code><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GenomicRanges::GRanges</a></code> objects, BPCells assumes all
objects use a zero-based, end-exclusive coordinate system (see below for details).</p><div class="section">
<h3 id="valid-range-like-objects">Valid Range-like objects<a class="anchor" aria-label="anchor" href="#valid-range-like-objects"></a></h3>


<p>BPCells can interpret the following types as ranges:</p><ul><li><p><code><a href="https://rdrr.io/r/base/list.html" class="external-link">list()</a></code>, <code><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame()</a></code>, with columns:</p><ul><li><p><code>chr</code>: Character or factor of chromosome names</p></li>
<li><p><code>start</code>: Start coordinates (0-based)</p></li>
<li><p><code>end</code>: End coordinates (exclusive)</p></li>
<li><p>(optional) <code>strand</code>: <code>"+"</code>/<code>"-"</code> or <code>TRUE</code>/<code>FALSE</code> for pos/neg strand</p></li>
<li><p>(optional) Additional metadata as named <code>list</code> entries or <code>data.frame</code> columns</p></li>
</ul></li>
<li><p><code><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GenomicRanges::GRanges</a></code></p><ul><li><p><code>start(x)</code> is interpreted as a 1-based start coordinate</p></li>
<li><p><code>end(x)</code> is interpreted as an inclusive end coordinate</p></li>
<li><p><code>strand(x)</code>: <code>"*"</code> entries are interpeted as postive strand</p></li>
<li><p>(optional) <code>mcols(x)</code> holds additional metadata</p></li>
</ul></li>
<li><p><code>character</code></p><ul><li><p>Given in format <code>"chr1:1000-2000"</code> or <code>"chr1:1,000-2,000"</code></p></li>
<li><p>Uses 0-based, end-exclusive coordinate system</p></li>
<li><p>Cannot be used for ranges where additional metadata is required</p></li>
</ul></li>
</ul></div>

<div class="section">
<h3 id="range-coordinate-systems">Range coordinate systems<a class="anchor" aria-label="anchor" href="#range-coordinate-systems"></a></h3>


<p>There are two main conventions for the coordinate systems:</p>
<p><em><strong>One-based, end-inclusive ranges</strong></em></p><ul><li><p>The first base of a chromosome is numbered 1</p></li>
<li><p>The last base in a range is equal to the end coordinate</p></li>
<li><p>e.g. 1-5 describes the first 5 bases of the chromosome</p></li>
<li><p>Used in formats such as SAM, GTF</p></li>
<li><p>In BPCells, used when reading or writing <code><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GenomicRanges::GRanges</a></code> objects</p></li>
</ul><p><em><strong>Zero-based, end-exclusive ranges</strong></em></p><ul><li><p>The first base of a chromosome is numbered 0</p></li>
<li><p>The last base in a range is one less than the end coordinate</p></li>
<li><p>e.g. 0-5 describes the first 5 bases of the chromosome</p></li>
<li><p>Used in formats such as BAM, BED</p></li>
<li><p>In BPCells, used for all other range objects</p></li>
</ul></div>

    </div>



  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Created by Benjamin Parks.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer></div>



  <img src="https://plausible.benparks.net/flask-plausible/bpcells-docs.png" style="position:absolute;"></body></html>

