<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Package index â€¢ BPCells</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Package index"><script defer data-domain="benparks.net" src="https://plausible.benparks.net/js/visit-counts.js"></script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">BPCells</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.2.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/pbmc3k.html">Get Started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/pbmc3k.html">Basic tutorial</a></li>
    <li><a class="dropdown-item" href="../articles/web-only/benchmarks.html">Performance Benchmarks</a></li>
    <li><a class="dropdown-item" href="../articles/web-only/how-it-works.html">How BPCells works</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/index.html">More articles...</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">News</a></li>
<li class="nav-item"><a class="nav-link" href="../python/index.html"><span class="fa fa-arrow-up-right-from-square"></span> Python Docs</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/bnprks/BPCells" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Package index</h1>
    </div>

    <div class="section level2">
      <h2 id="atac-seq-fragments">ATAC-seq Fragments<a class="anchor" aria-label="anchor" href="#atac-seq-fragments"></a></h2>




    </div><div class="section level2">

      <h3 id="fragment-i-o">Fragment I/O<a class="anchor" aria-label="anchor" href="#fragment-i-o"></a></h3>



    </div><div class="section level2">




      <dl><dt>

          <code><a href="open_fragments_10x.html">open_fragments_10x()</a></code> <code><a href="open_fragments_10x.html">write_fragments_10x()</a></code>

        </dt>
        <dd>Read/write a 10x fragments file</dd>
      </dl><dl><dt>

          <code><a href="fragment_io.html">write_fragments_memory()</a></code> <code><a href="fragment_io.html">write_fragments_dir()</a></code> <code><a href="fragment_io.html">open_fragments_dir()</a></code> <code><a href="fragment_io.html">write_fragments_hdf5()</a></code> <code><a href="fragment_io.html">open_fragments_hdf5()</a></code>

        </dt>
        <dd>Read/write BPCells fragment objects</dd>
      </dl><dl><dt>

          <code><a href="fragment_R_conversion.html">convert_to_fragments()</a></code> <code><a href="fragment_R_conversion.html">as()</a></code> <code><a href="fragment_R_conversion.html">as.data.frame()</a></code>

        </dt>
        <dd>Convert between BPCells fragments and R objects.</dd>
      </dl></div><div class="section level2">

      <h3 id="atac-analysis">ATAC Analysis<a class="anchor" aria-label="anchor" href="#atac-analysis"></a></h3>



    </div><div class="section level2">




      <dl><dt>

          <code><a href="qc_scATAC.html">qc_scATAC()</a></code>

        </dt>
        <dd>Calculate ArchR-compatible per-cell QC statistics</dd>
      </dl><dl><dt>

          <code><a href="nucleosome_counts.html">nucleosome_counts()</a></code>

        </dt>
        <dd>Count fragments by nucleosomal size</dd>
      </dl><dl><dt>

          <code><a href="footprint.html">footprint()</a></code>

        </dt>
        <dd>Get footprints around a set of genomic coordinates</dd>
      </dl><dl><dt>

          <code><a href="peak_matrix.html">peak_matrix()</a></code>

        </dt>
        <dd>Calculate ranges x cells overlap matrix</dd>
      </dl><dl><dt>

          <code><a href="tile_matrix.html">tile_matrix()</a></code>

        </dt>
        <dd>Calculate ranges x cells tile overlap matrix</dd>
      </dl><dl><dt>

          <code><a href="gene_scores.html">gene_score_weights_archr()</a></code> <code><a href="gene_scores.html">gene_score_archr()</a></code>

        </dt>
        <dd>Calculate GeneActivityScores</dd>
      </dl><dl><dt>

          <code><a href="call_peaks_macs.html">call_peaks_macs()</a></code>

        </dt>
        <dd>Call peaks using MACS2/3</dd>
      </dl><dl><dt>

          <code><a href="call_peaks_tile.html">call_peaks_tile()</a></code>

        </dt>
        <dd>Call peaks from tiles</dd>
      </dl><dl><dt>

          <code><a href="merge_peaks_iterative.html">merge_peaks_iterative()</a></code>

        </dt>
        <dd>Merge peaks</dd>
      </dl><dl><dt>

          <code><a href="write_insertion_bedgraph.html">write_insertion_bedgraph()</a></code>

        </dt>
        <dd>Write insertion counts to bedgraph file</dd>
      </dl></div><div class="section level2">

      <h3 id="fragment-operations">Fragment Operations<a class="anchor" aria-label="anchor" href="#fragment-operations"></a></h3>



    </div><div class="section level2">




      <dl><dt>

          <code><a href="shift_fragments.html">shift_fragments()</a></code>

        </dt>
        <dd>Shift start or end coordinates</dd>
      </dl><dl><dt>

          <code><a href="select_chromosomes.html">select_chromosomes()</a></code>

        </dt>
        <dd>Subset, translate, or reorder chromosome IDs</dd>
      </dl><dl><dt>

          <code><a href="select_cells.html">select_cells()</a></code>

        </dt>
        <dd>Subset, translate, or reorder cell IDs</dd>
      </dl><dl><dt>

          <code><a href="merge_cells.html">merge_cells()</a></code>

        </dt>
        <dd>Merge cells into pseudobulks</dd>
      </dl><dl><dt>

          <code><a href="subset_lengths.html">subset_lengths()</a></code>

        </dt>
        <dd>Subset fragments by length</dd>
      </dl><dl><dt>

          <code><a href="select_regions.html">select_regions()</a></code>

        </dt>
        <dd>Subset fragments by genomic region</dd>
      </dl><dl><dt>

          <code><a href="prefix_cell_names.html">prefix_cell_names()</a></code>

        </dt>
        <dd>Add sample prefix to cell names</dd>
      </dl><dl><dt>

          <code><a href="IterableFragments-methods.html">show(<i>&lt;IterableFragments&gt;</i>)</a></code> <code><a href="IterableFragments-methods.html">cellNames()</a></code> <code><a href="IterableFragments-methods.html">`cellNames&lt;-`()</a></code> <code><a href="IterableFragments-methods.html">chrNames()</a></code> <code><a href="IterableFragments-methods.html">`chrNames&lt;-`()</a></code>

        </dt>
        <dd>IterableFragments methods</dd>
      </dl><dl><dt>

          <code><a href="fragments_identical.html">fragments_identical()</a></code>

        </dt>
        <dd>Check if two fragments objects are identical</dd>
      </dl></div><div class="section level2">

      <h3 id="genomic-range-calculations">Genomic Range Calculations<a class="anchor" aria-label="anchor" href="#genomic-range-calculations"></a></h3>



    </div><div class="section level2">




      <dl><dt>

          <code><a href="genomic-ranges-like.html">genomic-ranges-like</a></code>

        </dt>
        <dd>Genomic range formats</dd>
      </dl><dl><dt>

          <code><a href="order_ranges.html">order_ranges()</a></code>

        </dt>
        <dd>Get end-sorted ordering for genome ranges</dd>
      </dl><dl><dt>

          <code><a href="range_distance_to_nearest.html">range_distance_to_nearest()</a></code>

        </dt>
        <dd>Find signed distance to nearest genomic ranges</dd>
      </dl><dl><dt>

          <code><a href="extend_ranges.html">extend_ranges()</a></code>

        </dt>
        <dd>Extend genome ranges in a strand-aware fashion.</dd>
      </dl><dl><dt>

          <code><a href="gene_score_tiles_archr.html">gene_score_tiles_archr()</a></code>

        </dt>
        <dd>Calculate gene-tile distances for ArchR gene activities</dd>
      </dl><dl><dt>

          <code><a href="normalize_ranges.html">normalize_ranges()</a></code>

        </dt>
        <dd>Normalize an object representing genomic ranges</dd>
      </dl></div><div class="section level2">
      <h2 id="matrix-operations-rna-atac-">Matrix Operations (RNA + ATAC)<a class="anchor" aria-label="anchor" href="#matrix-operations-rna-atac-"></a></h2>




    </div><div class="section level2">

      <h3 id="matrix-i-o">Matrix I/O<a class="anchor" aria-label="anchor" href="#matrix-i-o"></a></h3>



    </div><div class="section level2">




      <dl><dt>

          <code><a href="open_matrix_10x_hdf5.html">open_matrix_10x_hdf5()</a></code> <code><a href="open_matrix_10x_hdf5.html">write_matrix_10x_hdf5()</a></code>

        </dt>
        <dd>Read/write a 10x feature matrix</dd>
      </dl><dl><dt>

          <code><a href="open_matrix_anndata_hdf5.html">open_matrix_anndata_hdf5()</a></code> <code><a href="open_matrix_anndata_hdf5.html">write_matrix_anndata_hdf5()</a></code>

        </dt>
        <dd>Read/write AnnData matrix</dd>
      </dl><dl><dt>

          <code><a href="matrix_io.html">write_matrix_memory()</a></code> <code><a href="matrix_io.html">write_matrix_dir()</a></code> <code><a href="matrix_io.html">open_matrix_dir()</a></code> <code><a href="matrix_io.html">write_matrix_hdf5()</a></code> <code><a href="matrix_io.html">open_matrix_hdf5()</a></code>

        </dt>
        <dd>Read/write sparse matrices</dd>
      </dl><dl><dt>

          <code><a href="import_matrix_market.html">import_matrix_market()</a></code> <code><a href="import_matrix_market.html">import_matrix_market_10x()</a></code>

        </dt>
        <dd>Import MatrixMarket files</dd>
      </dl><dl><dt>

          <code><a href="matrix_R_conversion.html">as()</a></code> <code><a href="matrix_R_conversion.html">as.matrix()</a></code>

        </dt>
        <dd>Convert between BPCells matrix and R objects.</dd>
      </dl></div><div class="section level2">

      <h3 id="matrix-operations">Matrix Operations<a class="anchor" aria-label="anchor" href="#matrix-operations"></a></h3>



    </div><div class="section level2">




      <dl><dt>

          <code><a href="marker_features.html">marker_features()</a></code>

        </dt>
        <dd>Test for marker features</dd>
      </dl><dl><dt>

          <code><a href="matrix_stats.html">matrix_stats()</a></code>

        </dt>
        <dd>Calculate matrix stats</dd>
      </dl><dl><dt>

          <code><a href="svds.html">svds()</a></code>

        </dt>
        <dd>Calculate svds</dd>
      </dl><dl><dt>

          <code><a href="convert_matrix_type.html">convert_matrix_type()</a></code>

        </dt>
        <dd>Convert the type of a matrix</dd>
      </dl><dl><dt>

          <code><a href="transpose_storage_order.html">transpose_storage_order()</a></code>

        </dt>
        <dd>Transpose the storage order for a matrix</dd>
      </dl><dl><dt>

          <code><a href="sctransform_pearson.html">sctransform_pearson()</a></code>

        </dt>
        <dd>SCTransform Pearson Residuals</dd>
      </dl><dl><dt>

          <code><a href="min_elementwise.html">min_scalar()</a></code> <code><a href="min_elementwise.html">min_by_row()</a></code> <code><a href="min_elementwise.html">min_by_col()</a></code>

        </dt>
        <dd>Elementwise minimum</dd>
      </dl><dl><dt>

          <code><a href="mat_norm.html">add_rows()</a></code> <code><a href="mat_norm.html">add_cols()</a></code> <code><a href="mat_norm.html">multiply_rows()</a></code> <code><a href="mat_norm.html">multiply_cols()</a></code>

        </dt>
        <dd>Broadcasting vector arithmetic</dd>
      </dl><dl><dt>

          <code><a href="binarize.html">binarize()</a></code>

        </dt>
        <dd>Convert matrix elements to zeros and ones</dd>
      </dl><dl><dt>

          <code><a href="all_matrix_inputs.html">all_matrix_inputs()</a></code> <code><a href="all_matrix_inputs.html">`all_matrix_inputs&lt;-`()</a></code>

        </dt>
        <dd>Get/set inputs to a matrix transform</dd>
      </dl><dl><dt>

          <code><a href="checksum.html">checksum()</a></code>

        </dt>
        <dd>Calculate the MD5 checksum of an IterableMatrix</dd>
      </dl><dl><dt>

          <code><a href="apply_by_row.html">apply_by_row()</a></code> <code><a href="apply_by_row.html">apply_by_col()</a></code>

        </dt>
        <dd>Apply a function to summarize rows/cols</dd>
      </dl><dl><dt>

          <code><a href="regress_out.html">regress_out()</a></code>

        </dt>
        <dd>Regress out unwanted variation</dd>
      </dl><dl><dt>

          <code><a href="IterableMatrix-methods.html">matrix_type()</a></code> <code><a href="IterableMatrix-methods.html">storage_order()</a></code> <code><a href="IterableMatrix-methods.html">show(<i>&lt;IterableMatrix&gt;</i>)</a></code> <code><a href="IterableMatrix-methods.html">t(<i>&lt;IterableMatrix&gt;</i>)</a></code> <code><a href="IterableMatrix-methods.html">`%*%`(<i>&lt;IterableMatrix&gt;</i>,<i>&lt;matrix&gt;</i>)</a></code> <code><a href="IterableMatrix-methods.html">rowSums(<i>&lt;IterableMatrix&gt;</i>)</a></code> <code><a href="IterableMatrix-methods.html">colSums(<i>&lt;IterableMatrix&gt;</i>)</a></code> <code><a href="IterableMatrix-methods.html">rowMeans(<i>&lt;IterableMatrix&gt;</i>)</a></code> <code><a href="IterableMatrix-methods.html">colMeans(<i>&lt;IterableMatrix&gt;</i>)</a></code> <code><a href="IterableMatrix-methods.html">colVars()</a></code> <code><a href="IterableMatrix-methods.html">rowVars()</a></code> <code><a href="IterableMatrix-methods.html">rowMaxs()</a></code> <code><a href="IterableMatrix-methods.html">colMaxs()</a></code> <code><a href="IterableMatrix-methods.html">rowQuantiles()</a></code> <code><a href="IterableMatrix-methods.html">colQuantiles()</a></code> <code><a href="IterableMatrix-methods.html">log1p(<i>&lt;IterableMatrix&gt;</i>)</a></code> <code><a href="IterableMatrix-methods.html">log1p_slow()</a></code> <code><a href="IterableMatrix-methods.html">expm1(<i>&lt;IterableMatrix&gt;</i>)</a></code> <code><a href="IterableMatrix-methods.html">expm1_slow()</a></code> <code><a href="IterableMatrix-methods.html">`^`(<i>&lt;IterableMatrix&gt;</i>,<i>&lt;numeric&gt;</i>)</a></code> <code><a href="IterableMatrix-methods.html">`&lt;`(<i>&lt;numeric&gt;</i>,<i>&lt;IterableMatrix&gt;</i>)</a></code> <code><a href="IterableMatrix-methods.html">`&gt;`(<i>&lt;IterableMatrix&gt;</i>,<i>&lt;numeric&gt;</i>)</a></code> <code><a href="IterableMatrix-methods.html">`&lt;=`(<i>&lt;numeric&gt;</i>,<i>&lt;IterableMatrix&gt;</i>)</a></code> <code><a href="IterableMatrix-methods.html">`&gt;=`(<i>&lt;IterableMatrix&gt;</i>,<i>&lt;numeric&gt;</i>)</a></code> <code><a href="IterableMatrix-methods.html">round(<i>&lt;IterableMatrix&gt;</i>)</a></code> <code><a href="IterableMatrix-methods.html">`*`(<i>&lt;IterableMatrix&gt;</i>,<i>&lt;numeric&gt;</i>)</a></code> <code><a href="IterableMatrix-methods.html">`+`(<i>&lt;IterableMatrix&gt;</i>,<i>&lt;numeric&gt;</i>)</a></code> <code><a href="IterableMatrix-methods.html">`/`(<i>&lt;IterableMatrix&gt;</i>,<i>&lt;numeric&gt;</i>)</a></code> <code><a href="IterableMatrix-methods.html">`-`(<i>&lt;IterableMatrix&gt;</i>,<i>&lt;numeric&gt;</i>)</a></code>

        </dt>
        <dd>IterableMatrix methods</dd>
      </dl><dl><dt>

          <code><a href="pseudobulk_matrix.html">pseudobulk_matrix()</a></code>

        </dt>
        <dd>Aggregate counts matrices by cell group or feature.</dd>
      </dl></div><div class="section level2">
      <h2 id="reference-annotations">Reference Annotations<a class="anchor" aria-label="anchor" href="#reference-annotations"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="gene_mapping.html">human_gene_mapping</a></code> <code><a href="gene_mapping.html">mouse_gene_mapping</a></code>

        </dt>
        <dd>Gene Symbol Mapping data</dd>
      </dl><dl><dt>

          <code><a href="gene_matching.html">match_gene_symbol()</a></code> <code><a href="gene_matching.html">canonical_gene_symbol()</a></code>

        </dt>
        <dd>Gene symbol matching</dd>
      </dl><dl><dt>

          <code><a href="read_gtf.html">read_gtf()</a></code> <code><a href="read_gtf.html">read_gencode_genes()</a></code> <code><a href="read_gtf.html">read_gencode_transcripts()</a></code>

        </dt>
        <dd>Read GTF gene annotations</dd>
      </dl><dl><dt>

          <code><a href="read_bed.html">read_bed()</a></code> <code><a href="read_bed.html">read_encode_blacklist()</a></code>

        </dt>
        <dd>Read a bed file into a data frame</dd>
      </dl><dl><dt>

          <code><a href="read_ucsc_chrom_sizes.html">read_ucsc_chrom_sizes()</a></code>

        </dt>
        <dd>Read UCSC chromosome sizes</dd>
      </dl></div><div class="section level2">
      <h2 id="clustering">Clustering<a class="anchor" aria-label="anchor" href="#clustering"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="knn.html">knn_hnsw()</a></code> <code><a href="knn.html">knn_annoy()</a></code>

        </dt>
        <dd>Get a knn matrix from reduced dimensions</dd>
      </dl><dl><dt>

          <code><a href="cluster.html">cluster_graph_leiden()</a></code> <code><a href="cluster.html">cluster_graph_louvain()</a></code> <code><a href="cluster.html">cluster_graph_seurat()</a></code>

        </dt>
        <dd>Cluster an adjacency matrix</dd>
      </dl><dl><dt>

          <code><a href="knn_graph.html">knn_to_graph()</a></code> <code><a href="knn_graph.html">knn_to_snn_graph()</a></code> <code><a href="knn_graph.html">knn_to_geodesic_graph()</a></code>

        </dt>
        <dd>K Nearest Neighbor (KNN) Graph</dd>
      </dl><dl><dt>

          <code><a href="cluster_membership_matrix.html">cluster_membership_matrix()</a></code>

        </dt>
        <dd>Convert grouping vector to sparse matrix</dd>
      </dl></div><div class="section level2">
      <h2 id="plots">Plots<a class="anchor" aria-label="anchor" href="#plots"></a></h2>




    </div><div class="section level2">

      <h3 id="single-cell-plots">Single cell plots<a class="anchor" aria-label="anchor" href="#single-cell-plots"></a></h3>



    </div><div class="section level2">




      <dl><dt>

          <code><a href="plot_read_count_knee.html">plot_read_count_knee()</a></code>

        </dt>
        <dd>Knee plot of single cell read counts</dd>
      </dl><dl><dt>

          <code><a href="plot_embedding.html">plot_embedding()</a></code>

        </dt>
        <dd>Plot UMAP or embeddings</dd>
      </dl><dl><dt>

          <code><a href="plot_dot.html">plot_dot()</a></code>

        </dt>
        <dd>Dotplot</dd>
      </dl><dl><dt>

          <code><a href="plot_fragment_length.html">plot_fragment_length()</a></code>

        </dt>
        <dd>Fragment size distribution</dd>
      </dl><dl><dt>

          <code><a href="plot_tf_footprint.html">plot_tf_footprint()</a></code>

        </dt>
        <dd>Plot TF footprint</dd>
      </dl><dl><dt>

          <code><a href="plot_tss_profile.html">plot_tss_profile()</a></code>

        </dt>
        <dd>Plot TSS profile</dd>
      </dl><dl><dt>

          <code><a href="plot_tss_scatter.html">plot_tss_scatter()</a></code>

        </dt>
        <dd>TSS Enrichment vs. Fragment Counts plot</dd>
      </dl></div><div class="section level2">

      <h3 id="genomic-track-plots">Genomic track plots<a class="anchor" aria-label="anchor" href="#genomic-track-plots"></a></h3>



    </div><div class="section level2">




      <dl><dt>

          <code><a href="trackplot_combine.html">trackplot_combine()</a></code>

        </dt>
        <dd>Combine track plots</dd>
      </dl><dl><dt>

          <code><a href="trackplot_coverage.html">trackplot_coverage()</a></code>

        </dt>
        <dd>Pseudobulk coverage trackplot</dd>
      </dl><dl><dt>

          <code><a href="trackplot_gene.html">trackplot_gene()</a></code>

        </dt>
        <dd>Plot transcript models</dd>
      </dl><dl><dt>

          <code><a href="trackplot_loop.html">trackplot_loop()</a></code>

        </dt>
        <dd>Plot loops</dd>
      </dl><dl><dt>

          <code><a href="trackplot_genome_annotation.html">trackplot_genome_annotation()</a></code>

        </dt>
        <dd>Plot range-based annotation tracks (e.g. peaks)</dd>
      </dl><dl><dt>

          <code><a href="trackplot_scalebar.html">trackplot_scalebar()</a></code>

        </dt>
        <dd>Plot scale bar</dd>
      </dl><dl><dt>

          <code><a href="gene_region.html">gene_region()</a></code>

        </dt>
        <dd>Find gene region</dd>
      </dl><dl><dt>

          <code><a href="trackplot_utils.html">set_trackplot_label()</a></code> <code><a href="trackplot_utils.html">set_trackplot_height()</a></code> <code><a href="trackplot_utils.html">get_trackplot_height()</a></code>

        </dt>
        <dd>Adjust trackplot properties</dd>
      </dl></div><div class="section level2">

      <h3 id="plotting-utilities">Plotting utilities<a class="anchor" aria-label="anchor" href="#plotting-utilities"></a></h3>



    </div><div class="section level2">




      <dl><dt>

          <code><a href="palettes.html">discrete_palette()</a></code> <code><a href="palettes.html">continuous_palette()</a></code>

        </dt>
        <dd>Color palettes</dd>
      </dl><dl><dt>

          <code><a href="collect_features.html">collect_features()</a></code>

        </dt>
        <dd>Collect features for plotting</dd>
      </dl><dl><dt>

          <code><a href="rotate_x_labels.html">rotate_x_labels()</a></code>

        </dt>
        <dd>Rotate ggplot x axis labels</dd>
      </dl></div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Created by Benjamin Parks.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>



  <img src="https://plausible.benparks.net/flask-plausible/bpcells-docs.png" style="position:absolute;"></body></html>

