<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Calculate ArchR-compatible per-cell QC statistics — qc_scATAC • BPCells</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Calculate ArchR-compatible per-cell QC statistics — qc_scATAC"><meta name="description" content="Calculate ArchR-compatible per-cell QC statistics"><meta property="og:description" content="Calculate ArchR-compatible per-cell QC statistics"><script defer data-domain="benparks.net" src="https://plausible.benparks.net/js/visit-counts.js"></script><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.16.15/dist/katex.min.css" integrity="sha384-Htz9HMhiwV8GuQ28Xr9pEs1B4qJiYu/nYLLwlDklR53QibDfmQzi7rYxXhMH/5/u" crossorigin="anonymous"><!-- The loading of KaTeX is deferred to speed up page rendering --><script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.15/dist/katex.min.js" integrity="sha384-bxmi2jLGCvnsEqMuYLKE/KsVCxV3PqmKeK6Y6+lmNXBry6+luFkEOsmp5vD9I/7+" crossorigin="anonymous"></script><!-- To automatically render math in text elements, include the auto-render extension: --><script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.15/dist/contrib/auto-render.min.js" integrity="sha384-hCXGrW6PitJEwbkoStFjeJxv+fSOOQKOPbJxSfM6G5sWZjAyWhXiTIIAmQqnlLlh" crossorigin="anonymous" onload="renderMathInElement(document.body);"></script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">BPCells</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.3.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/pbmc3k.html">Get Started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/pbmc3k.html">Basic tutorial</a></li>
    <li><a class="dropdown-item" href="../articles/web-only/benchmarks.html">Performance Benchmarks</a></li>
    <li><a class="dropdown-item" href="../articles/web-only/how-it-works.html">How BPCells works</a></li>
    <li><a class="dropdown-item" href="../articles/web-only/programming-efficiency.html">Efficiency tips</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/index.html">More articles...</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">News</a></li>
<li class="nav-item"><a class="nav-link" href="../python/index.html"><span class="fa fa-arrow-up-right-from-square"></span> Python Docs</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/bnprks/BPCells" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Calculate ArchR-compatible per-cell QC statistics</h1>
      <small class="dont-index">Source: <a href="https://github.com/bnprks/BPCells/tree/main/r/R/atac_utils.R" class="external-link"><code>R/atac_utils.R</code></a></small>
      <div class="d-none name"><code>qc_scATAC.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Calculate ArchR-compatible per-cell QC statistics</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">qc_scATAC</span><span class="op">(</span><span class="va">fragments</span>, <span class="va">genes</span>, <span class="va">blacklist</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-fragments">fragments<a class="anchor" aria-label="anchor" href="#arg-fragments"></a></dt>
<dd><p>IterableFragments object</p></dd>


<dt id="arg-genes">genes<a class="anchor" aria-label="anchor" href="#arg-genes"></a></dt>
<dd><p>Gene coordinates given as GRanges, data.frame, or list. See <code><a href="genomic-ranges-like.html">help("genomic-ranges-like")</a></code> for details on format and coordinate systems. Required attributes:</p><ul><li><p><code>chr</code>, <code>start</code>, <code>end</code>: genomic position</p></li>
</ul></dd>


<dt id="arg-blacklist">blacklist<a class="anchor" aria-label="anchor" href="#arg-blacklist"></a></dt>
<dd><p>Blacklisted regions given as GRanges, data.frame, or list. See <code><a href="genomic-ranges-like.html">help("genomic-ranges-like")</a></code> for details on format and coordinate systems. Required attributes:</p><ul><li><p><code>chr</code>, <code>start</code>, <code>end</code>: genomic position</p></li>
</ul></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>data.frame with QC data</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>This implementation mimics ArchR's default parameters. For uses requiring more flexibility to tweak default parameters,
the best option is to re-implement this function with required changes.
Output columns of data.frame:</p><ul><li><p><code>cellName</code>: cell name for each cell</p></li>
<li><p><code>nFrags</code>: number of fragments per cell</p></li>
<li><p><code>subNucleosomal</code>, <code>monoNucleosomal</code>, <code>multiNucleosomal</code>: number of fragments of size 1-146bp, 147-254bp, and 255bp + respectively.
equivalent to ArchR's nMonoFrags, nDiFrags, nMultiFrags respectively</p></li>
<li><p><code>TSSEnrichment</code>: <code>AvgInsertInTSS / max(AvgInsertFlankingTSS, 0.1)</code>, where <code>AvgInsertInTSS</code> is <code>ReadsInTSS / 101</code> (window size),
and <code>AvgInsertFlankingTSS</code> is <code>ReadsFlankingTSS / (100*2)</code> (window size). The <code>max(0.1)</code> ensures that very low-read cells
do not get assigned spuriously high TSSEnrichment.</p></li>
<li><p><code>ReadsInPromoter</code>: Number of reads from 2000bp upstream of TSS to 101bp downstream of TSS</p></li>
<li><p><code>ReadsInBlacklist</code>: Number of reads in the provided blacklist region</p></li>
<li><p><code>ReadsInTSS</code>: Number of reads overlapping the 101bp centered around each TSS</p></li>
<li><p><code>ReadsFlankingTSS</code>: Number of reads overlapping 1901-2000bp +/- each TSS</p></li>
</ul><p>Differences from ArchR:
Note that ArchR by default uses a different set of annotations to derive TSS sites and promoter sites.
This function uses just one annotation for gene start+end sites, so must be called twice to exactly
re-calculate the ArchR QC stats.</p>
<p>ArchR's <code>PromoterRatio</code> and <code>BlacklistRatio</code> are not included in the output, as they can be easily calculated
from <code>ReadsInPromoter / nFrags</code> and  <code>ReadsInBlacklist / nFrags</code>. Similarly, ArchR's <code>NucleosomeRatio</code> can be calculated
as <code>(monoNucleosomal + multiNucleosomal) / subNucleosomal</code>.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co">## Prep data</span></span></span>
<span class="r-in"><span><span class="va">frags</span> <span class="op">&lt;-</span> <span class="fu"><a href="demo_data.html">get_demo_frags</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">reference_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"references"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="read_gtf.html">read_gencode_transcripts</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="va">reference_dir</span>, </span></span>
<span class="r-in"><span>  release<span class="op">=</span><span class="st">"42"</span>, </span></span>
<span class="r-in"><span>  transcript_choice<span class="op">=</span><span class="st">"MANE_Select"</span>,</span></span>
<span class="r-in"><span>  annotation_set <span class="op">=</span> <span class="st">"basic"</span>,</span></span>
<span class="r-in"><span>  features<span class="op">=</span><span class="st">"transcript"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">blacklist</span> <span class="op">&lt;-</span> <span class="fu"><a href="read_bed.html">read_encode_blacklist</a></span><span class="op">(</span><span class="va">reference_dir</span>, genome <span class="op">=</span> <span class="st">"hg38"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Run qc</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu">qc_scATAC</span><span class="op">(</span><span class="va">frags</span>, <span class="va">genes</span>, <span class="va">blacklist</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 6 × 10</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   cellName  TSSEnrichment nFrags subNucleosomal monoNucleosomal multiNucleosomal</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">1</span> TTTAGCAA…          25.8   <span style="text-decoration: underline;">1</span>387            704             476              207</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">2</span> AGCCGGTT…          24.5   <span style="text-decoration: underline;">2</span>983           <span style="text-decoration: underline;">1</span>430            <span style="text-decoration: underline;">1</span>040              513</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">3</span> TGATTAGT…          18.5   <span style="text-decoration: underline;">1</span>093            542             367              184</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">4</span> ATTGACTC…          23.0   <span style="text-decoration: underline;">1</span>204            654             355              195</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">5</span> CGTTAGGT…          21.5   <span style="text-decoration: underline;">1</span>387            610             520              257</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">6</span> AAACCGCG…          42.9   <span style="text-decoration: underline;">2</span>650           <span style="text-decoration: underline;">1</span>279             925              446</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># ℹ 4 more variables: ReadsInTSS &lt;dbl&gt;, ReadsFlankingTSS &lt;dbl&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   ReadsInPromoter &lt;dbl&gt;, ReadsInBlacklist &lt;dbl&gt;</span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Created by Benjamin Parks.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>



  <img src="https://plausible.benparks.net/flask-plausible/bpcells-docs.png" style="position:absolute;"></body></html>

