<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Read GTF gene annotations — read_gtf • BPCells</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Read GTF gene annotations — read_gtf"><meta name="description" content="Read gene annotations from gtf format into a data frame. The source can be
a URL, a gtf file on disk, or a gencode release version."><meta property="og:description" content="Read gene annotations from gtf format into a data frame. The source can be
a URL, a gtf file on disk, or a gencode release version."><script defer data-domain="benparks.net" src="https://plausible.benparks.net/js/visit-counts.js"></script><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.16.15/dist/katex.min.css" integrity="sha384-Htz9HMhiwV8GuQ28Xr9pEs1B4qJiYu/nYLLwlDklR53QibDfmQzi7rYxXhMH/5/u" crossorigin="anonymous"><!-- The loading of KaTeX is deferred to speed up page rendering --><script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.15/dist/katex.min.js" integrity="sha384-bxmi2jLGCvnsEqMuYLKE/KsVCxV3PqmKeK6Y6+lmNXBry6+luFkEOsmp5vD9I/7+" crossorigin="anonymous"></script><!-- To automatically render math in text elements, include the auto-render extension: --><script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.15/dist/contrib/auto-render.min.js" integrity="sha384-hCXGrW6PitJEwbkoStFjeJxv+fSOOQKOPbJxSfM6G5sWZjAyWhXiTIIAmQqnlLlh" crossorigin="anonymous" onload="renderMathInElement(document.body);"></script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">BPCells</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.3.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/pbmc3k.html">Get Started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/pbmc3k.html">Basic tutorial</a></li>
    <li><a class="dropdown-item" href="../articles/web-only/benchmarks.html">Performance Benchmarks</a></li>
    <li><a class="dropdown-item" href="../articles/web-only/how-it-works.html">How BPCells works</a></li>
    <li><a class="dropdown-item" href="../articles/web-only/programming-efficiency.html">Efficiency tips</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/index.html">More articles...</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">News</a></li>
<li class="nav-item"><a class="nav-link" href="../python/index.html"><span class="fa fa-arrow-up-right-from-square"></span> Python Docs</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/bnprks/BPCells" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Read GTF gene annotations</h1>
      <small class="dont-index">Source: <a href="https://github.com/bnprks/BPCells/tree/main/r/R/genomeAnnotations.R" class="external-link"><code>R/genomeAnnotations.R</code></a></small>
      <div class="d-none name"><code>read_gtf.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Read gene annotations from gtf format into a data frame. The source can be
a URL, a gtf file on disk, or a gencode release version.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">read_gtf</span><span class="op">(</span></span>
<span>  <span class="va">path</span>,</span>
<span>  attributes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gene_id"</span><span class="op">)</span>,</span>
<span>  tags <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">character</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span>,</span>
<span>  features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gene"</span><span class="op">)</span>,</span>
<span>  keep_attribute_column <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  backup_url <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  timeout <span class="op">=</span> <span class="fl">300</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">read_gencode_genes</span><span class="op">(</span></span>
<span>  <span class="va">dir</span>,</span>
<span>  release <span class="op">=</span> <span class="st">"latest"</span>,</span>
<span>  annotation_set <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"basic"</span>, <span class="st">"comprehensive"</span><span class="op">)</span>,</span>
<span>  gene_type <span class="op">=</span> <span class="st">"lncRNA|protein_coding|IG_.*_gene|TR_.*_gene"</span>,</span>
<span>  attributes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gene_id"</span>, <span class="st">"gene_type"</span>, <span class="st">"gene_name"</span><span class="op">)</span>,</span>
<span>  tags <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">character</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span>,</span>
<span>  features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gene"</span><span class="op">)</span>,</span>
<span>  timeout <span class="op">=</span> <span class="fl">300</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">read_gencode_transcripts</span><span class="op">(</span></span>
<span>  <span class="va">dir</span>,</span>
<span>  release <span class="op">=</span> <span class="st">"latest"</span>,</span>
<span>  transcript_choice <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"MANE_Select"</span>, <span class="st">"Ensembl_Canonical"</span>, <span class="st">"all"</span><span class="op">)</span>,</span>
<span>  annotation_set <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"basic"</span>, <span class="st">"comprehensive"</span><span class="op">)</span>,</span>
<span>  gene_type <span class="op">=</span> <span class="st">"lncRNA|protein_coding|IG_.*_gene|TR_.*_gene"</span>,</span>
<span>  attributes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gene_id"</span>, <span class="st">"gene_type"</span>, <span class="st">"gene_name"</span>, <span class="st">"transcript_id"</span><span class="op">)</span>,</span>
<span>  features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"transcript"</span>, <span class="st">"exon"</span><span class="op">)</span>,</span>
<span>  timeout <span class="op">=</span> <span class="fl">300</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-path">path<a class="anchor" aria-label="anchor" href="#arg-path"></a></dt>
<dd><p>Path to file (or desired save location if backup_url is used)</p></dd>


<dt id="arg-attributes">attributes<a class="anchor" aria-label="anchor" href="#arg-attributes"></a></dt>
<dd><p>Vector of GTF attribute names to parse out as columns</p></dd>


<dt id="arg-tags">tags<a class="anchor" aria-label="anchor" href="#arg-tags"></a></dt>
<dd><p>Vector of tags to parse out as boolean presence/absence</p></dd>


<dt id="arg-features">features<a class="anchor" aria-label="anchor" href="#arg-features"></a></dt>
<dd><p>List of features types to keep from the GTF (e.g. gene, transcript, exon, intron)</p></dd>


<dt id="arg-keep-attribute-column">keep_attribute_column<a class="anchor" aria-label="anchor" href="#arg-keep-attribute-column"></a></dt>
<dd><p>Boolean for whether to preserve the raw attribute text column</p></dd>


<dt id="arg-backup-url">backup_url<a class="anchor" aria-label="anchor" href="#arg-backup-url"></a></dt>
<dd><p>If path does not exist, provides a URL to download the gtf from</p></dd>


<dt id="arg-timeout">timeout<a class="anchor" aria-label="anchor" href="#arg-timeout"></a></dt>
<dd><p>Maximum time in seconds to wait for download from backup_url</p></dd>


<dt id="arg-dir">dir<a class="anchor" aria-label="anchor" href="#arg-dir"></a></dt>
<dd><p>Output directory to cache the downloaded gtf file</p></dd>


<dt id="arg-release">release<a class="anchor" aria-label="anchor" href="#arg-release"></a></dt>
<dd><p>release version (prefix with M for mouse versions). For
most recent version, use "latest" or "latest_mouse"</p></dd>


<dt id="arg-annotation-set">annotation_set<a class="anchor" aria-label="anchor" href="#arg-annotation-set"></a></dt>
<dd><p>Either "basic" or "comprehensive" annotation sets (see details section).</p></dd>


<dt id="arg-gene-type">gene_type<a class="anchor" aria-label="anchor" href="#arg-gene-type"></a></dt>
<dd><p>Regular expression with which gene types to keep.
Defaults to protein_coding, lncRNA, and IG/TR genes</p></dd>


<dt id="arg-transcript-choice">transcript_choice<a class="anchor" aria-label="anchor" href="#arg-transcript-choice"></a></dt>
<dd><p>Method for selecting representative transcripts. Choices are:</p><ul><li><p>MANE_Select: human-only, most conservative</p></li>
<li><p>Ensembl_Canonical: human+mouse, superset of MANE_Select for human</p></li>
<li><p>all: Preserve all transcript models (not recommended for plotting)</p></li>
</ul></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>Data frame with coordinates using the 0-based convention. Columns are:</p><ul><li><p>chr</p></li>
<li><p>source</p></li>
<li><p>feature</p></li>
<li><p>start</p></li>
<li><p>end</p></li>
<li><p>score</p></li>
<li><p>strand</p></li>
<li><p>frame</p></li>
<li><p>attributes (optional; named according to listed attributes)</p></li>
<li><p>tags (named according to listed tags)</p></li>
</ul></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p><strong>read_gtf</strong></p>
<p>Read gtf from a file or URL</p>
<p><strong>read_gencode_genes</strong></p>
<p>Read gene annotations directly from GENCODE. The file name will vary depending
on the release and annotation set requested, but will be of the format
<code>gencode.v42.annotation.gtf.gz</code>. GENCODE currently recommends the basic set:
<a href="https://www.gencodegenes.org/human/" class="external-link">https://www.gencodegenes.org/human/</a>. In release 42, both the comprehensive and
basic sets had identical gene-level annotations, but the comprehensive set had
additional transcript variants annotated.</p>
<p><strong>read_gencode_transcripts</strong></p>
<p>Read transcript models from GENCODE, for use with trackplot_gene()</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="read_bed.html">read_bed()</a></code>, <code><a href="read_bed.html">read_encode_blacklist()</a></code></p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co">#######################################################################</span></span></span>
<span class="r-in"><span><span class="co">## read_gtf() example</span></span></span>
<span class="r-in"><span><span class="co">#######################################################################</span></span></span>
<span class="r-in"><span><span class="va">species</span> <span class="op">&lt;-</span> <span class="st">"Saccharomyces_cerevisiae"</span></span></span>
<span class="r-in"><span><span class="va">version</span> <span class="op">&lt;-</span> <span class="st">"GCF_000146045.2_R64"</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu">read_gtf</span><span class="op">(</span></span></span>
<span class="r-in"><span> path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"./reference/%s_genomic.gtf.gz"</span>, <span class="va">version</span><span class="op">)</span>,</span></span>
<span class="r-in"><span> backup_url <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>   <span class="st">"https://ftp.ncbi.nlm.nih.gov/genomes/refseq/fungi/%s/reference/%s/%s_genomic.gtf.gz"</span>,</span></span>
<span class="r-in"><span>   <span class="va">species</span>, <span class="va">version</span>, <span class="va">version</span></span></span>
<span class="r-in"><span> <span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 6 × 9</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   chr         source feature start   end score strand frame gene_id  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">1</span> NC_001133.9 RefSeq gene     <span style="text-decoration: underline;">1</span>806  <span style="text-decoration: underline;">2</span>169 .     -      .     YAL068C  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">2</span> NC_001133.9 RefSeq gene     <span style="text-decoration: underline;">2</span>479  <span style="text-decoration: underline;">2</span>707 .     +      .     YAL067W-A</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">3</span> NC_001133.9 RefSeq gene     <span style="text-decoration: underline;">7</span>234  <span style="text-decoration: underline;">9</span>016 .     -      .     YAL067C  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">4</span> NC_001133.9 RefSeq gene    <span style="text-decoration: underline;">11</span>564 <span style="text-decoration: underline;">11</span>951 .     -      .     YAL065C  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">5</span> NC_001133.9 RefSeq gene    <span style="text-decoration: underline;">12</span>045 <span style="text-decoration: underline;">12</span>426 .     +      .     YAL064W-B</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">6</span> NC_001133.9 RefSeq gene    <span style="text-decoration: underline;">13</span>362 <span style="text-decoration: underline;">13</span>743 .     -      .     YAL064C-A</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#######################################################################</span></span></span>
<span class="r-in"><span><span class="co">## read_gencode_genes() example</span></span></span>
<span class="r-in"><span><span class="co">#######################################################################</span></span></span>
<span class="r-in"><span><span class="fu">read_gencode_genes</span><span class="op">(</span><span class="st">"./references"</span>, release <span class="op">=</span> <span class="st">"42"</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 39,319 × 11</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    chr   source feature  start    end score strand frame gene_id       gene_type</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 1</span> chr1  HAVANA gene     <span style="text-decoration: underline;">11</span>868  <span style="text-decoration: underline;">14</span>409 .     +      .     ENSG00000290… lncRNA   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 2</span> chr1  HAVANA gene     <span style="text-decoration: underline;">29</span>553  <span style="text-decoration: underline;">31</span>109 .     +      .     ENSG00000243… lncRNA   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 3</span> chr1  HAVANA gene     <span style="text-decoration: underline;">34</span>553  <span style="text-decoration: underline;">36</span>081 .     -      .     ENSG00000237… lncRNA   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 4</span> chr1  HAVANA gene     <span style="text-decoration: underline;">57</span>597  <span style="text-decoration: underline;">64</span>116 .     +      .     ENSG00000290… lncRNA   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 5</span> chr1  HAVANA gene     <span style="text-decoration: underline;">65</span>418  <span style="text-decoration: underline;">71</span>585 .     +      .     ENSG00000186… protein_…</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 6</span> chr1  HAVANA gene     <span style="text-decoration: underline;">89</span>294 <span style="text-decoration: underline;">133</span>723 .     -      .     ENSG00000238… lncRNA   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 7</span> chr1  HAVANA gene     <span style="text-decoration: underline;">89</span>550  <span style="text-decoration: underline;">91</span>105 .     -      .     ENSG00000239… lncRNA   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 8</span> chr1  HAVANA gene    <span style="text-decoration: underline;">139</span>789 <span style="text-decoration: underline;">140</span>339 .     -      .     ENSG00000239… lncRNA   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 9</span> chr1  HAVANA gene    <span style="text-decoration: underline;">141</span>473 <span style="text-decoration: underline;">173</span>862 .     -      .     ENSG00000241… lncRNA   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">10</span> chr1  HAVANA gene    <span style="text-decoration: underline;">160</span>445 <span style="text-decoration: underline;">161</span>525 .     +      .     ENSG00000241… lncRNA   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># ℹ 39,309 more rows</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># ℹ 1 more variable: gene_name &lt;chr&gt;</span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#######################################################################</span></span></span>
<span class="r-in"><span><span class="co">## read_gencode_transcripts() example</span></span></span>
<span class="r-in"><span><span class="co">#######################################################################</span></span></span>
<span class="r-in"><span><span class="co">## If read_gencode_genes() was already ran on the same release, </span></span></span>
<span class="r-in"><span><span class="co">## will reuse previously downloaded annotations</span></span></span>
<span class="r-in"><span><span class="fu">read_gencode_transcripts</span><span class="op">(</span><span class="st">"./references"</span>, release <span class="op">=</span> <span class="st">"42"</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 220,296 × 13</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    chr   source feature     start    end score strand frame gene_id    gene_type</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 1</span> chr1  HAVANA transcript  <span style="text-decoration: underline;">65</span>418  <span style="text-decoration: underline;">71</span>585 .     +      .     ENSG00000… protein_…</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 2</span> chr1  HAVANA exon        <span style="text-decoration: underline;">65</span>418  <span style="text-decoration: underline;">65</span>433 .     +      .     ENSG00000… protein_…</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 3</span> chr1  HAVANA exon        <span style="text-decoration: underline;">65</span>519  <span style="text-decoration: underline;">65</span>573 .     +      .     ENSG00000… protein_…</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 4</span> chr1  HAVANA exon        <span style="text-decoration: underline;">69</span>036  <span style="text-decoration: underline;">71</span>585 .     +      .     ENSG00000… protein_…</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 5</span> chr1  HAVANA transcript <span style="text-decoration: underline;">450</span>739 <span style="text-decoration: underline;">451</span>678 .     -      .     ENSG00000… protein_…</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 6</span> chr1  HAVANA exon       <span style="text-decoration: underline;">450</span>739 <span style="text-decoration: underline;">451</span>678 .     -      .     ENSG00000… protein_…</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 7</span> chr1  HAVANA transcript <span style="text-decoration: underline;">685</span>715 <span style="text-decoration: underline;">686</span>654 .     -      .     ENSG00000… protein_…</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 8</span> chr1  HAVANA exon       <span style="text-decoration: underline;">685</span>715 <span style="text-decoration: underline;">686</span>654 .     -      .     ENSG00000… protein_…</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 9</span> chr1  HAVANA transcript <span style="text-decoration: underline;">923</span>922 <span style="text-decoration: underline;">944</span>574 .     +      .     ENSG00000… protein_…</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">10</span> chr1  HAVANA exon       <span style="text-decoration: underline;">923</span>922 <span style="text-decoration: underline;">924</span>948 .     +      .     ENSG00000… protein_…</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># ℹ 220,286 more rows</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># ℹ 3 more variables: gene_name &lt;chr&gt;, transcript_id &lt;chr&gt;, MANE_Select &lt;lgl&gt;</span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Created by Benjamin Parks.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>



  <img src="https://plausible.benparks.net/flask-plausible/bpcells-docs.png" style="position:absolute;"></body></html>

