<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Pseudobulk coverage trackplot — trackplot_coverage • BPCells</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Pseudobulk coverage trackplot — trackplot_coverage"><meta name="description" content="Plot a pseudobulk genome track, showing the number of fragment insertions
across a region for each cell type or group."><meta property="og:description" content="Plot a pseudobulk genome track, showing the number of fragment insertions
across a region for each cell type or group."><script defer data-domain="benparks.net" src="https://plausible.benparks.net/js/visit-counts.js"></script><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.16.15/dist/katex.min.css" integrity="sha384-Htz9HMhiwV8GuQ28Xr9pEs1B4qJiYu/nYLLwlDklR53QibDfmQzi7rYxXhMH/5/u" crossorigin="anonymous"><!-- The loading of KaTeX is deferred to speed up page rendering --><script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.15/dist/katex.min.js" integrity="sha384-bxmi2jLGCvnsEqMuYLKE/KsVCxV3PqmKeK6Y6+lmNXBry6+luFkEOsmp5vD9I/7+" crossorigin="anonymous"></script><!-- To automatically render math in text elements, include the auto-render extension: --><script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.15/dist/contrib/auto-render.min.js" integrity="sha384-hCXGrW6PitJEwbkoStFjeJxv+fSOOQKOPbJxSfM6G5sWZjAyWhXiTIIAmQqnlLlh" crossorigin="anonymous" onload="renderMathInElement(document.body);"></script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">BPCells</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.3.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/pbmc3k.html">Get Started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/pbmc3k.html">Basic tutorial</a></li>
    <li><a class="dropdown-item" href="../articles/web-only/benchmarks.html">Performance Benchmarks</a></li>
    <li><a class="dropdown-item" href="../articles/web-only/how-it-works.html">How BPCells works</a></li>
    <li><a class="dropdown-item" href="../articles/web-only/programming-efficiency.html">Efficiency tips</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/index.html">More articles...</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">News</a></li>
<li class="nav-item"><a class="nav-link" href="../python/index.html"><span class="fa fa-arrow-up-right-from-square"></span> Python Docs</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/bnprks/BPCells" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Pseudobulk coverage trackplot</h1>
      <small class="dont-index">Source: <a href="https://github.com/bnprks/BPCells/tree/main/r/R/trackplots.R" class="external-link"><code>R/trackplots.R</code></a></small>
      <div class="d-none name"><code>trackplot_coverage.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Plot a pseudobulk genome track, showing the number of fragment insertions
across a region for each cell type or group.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">trackplot_coverage</span><span class="op">(</span></span>
<span>  <span class="va">fragments</span>,</span>
<span>  <span class="va">region</span>,</span>
<span>  <span class="va">groups</span>,</span>
<span>  <span class="va">cell_read_counts</span>,</span>
<span>  group_order <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  bins <span class="op">=</span> <span class="fl">500</span>,</span>
<span>  clip_quantile <span class="op">=</span> <span class="fl">0.999</span>,</span>
<span>  colors <span class="op">=</span> <span class="fu"><a href="palettes.html">discrete_palette</a></span><span class="op">(</span><span class="st">"stallion"</span><span class="op">)</span>,</span>
<span>  legend_label <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  zero_based_coords <span class="op">=</span> <span class="op">!</span><span class="fu">is</span><span class="op">(</span><span class="va">region</span>, <span class="st">"GRanges"</span><span class="op">)</span>,</span>
<span>  return_data <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-fragments">fragments<a class="anchor" aria-label="anchor" href="#arg-fragments"></a></dt>
<dd><p>Fragments object</p></dd>


<dt id="arg-region">region<a class="anchor" aria-label="anchor" href="#arg-region"></a></dt>
<dd><p>Region to plot, e.g. output from <code><a href="gene_region.html">gene_region()</a></code>. String of format "chr1:100-200",
or list/data.frame/GRanges of length 1 specifying chr, start, end. See <code><a href="genomic-ranges-like.html">help("genomic-ranges-like")</a></code> for details</p></dd>


<dt id="arg-groups">groups<a class="anchor" aria-label="anchor" href="#arg-groups"></a></dt>
<dd><p>Vector with one entry per cell, specifying the cell's group</p></dd>


<dt id="arg-cell-read-counts">cell_read_counts<a class="anchor" aria-label="anchor" href="#arg-cell-read-counts"></a></dt>
<dd><p>Numeric vector of read counts for each cell (used for normalization)</p></dd>


<dt id="arg-group-order">group_order<a class="anchor" aria-label="anchor" href="#arg-group-order"></a></dt>
<dd><p>Optional vector listing ordering of groups</p></dd>


<dt id="arg-bins">bins<a class="anchor" aria-label="anchor" href="#arg-bins"></a></dt>
<dd><p>Number of bins to plot across the region</p></dd>


<dt id="arg-clip-quantile">clip_quantile<a class="anchor" aria-label="anchor" href="#arg-clip-quantile"></a></dt>
<dd><p>(optional) Quantile of values for clipping y-axis limits. Default of 0.999 will crop out
just the most extreme outliers across the region. NULL to disable clipping</p></dd>


<dt id="arg-colors">colors<a class="anchor" aria-label="anchor" href="#arg-colors"></a></dt>
<dd><p>Character vector of color values (optionally named by group)</p></dd>


<dt id="arg-legend-label">legend_label<a class="anchor" aria-label="anchor" href="#arg-legend-label"></a></dt>
<dd><p><a href="https://lifecycle.r-lib.org/articles/stages.html#deprecated" class="external-link"><img src="figures/lifecycle-deprecated.svg" alt="[Deprecated]"></a> Custom label to put on the legend (no longer used as color legend is not shown anymore)</p></dd>


<dt id="arg-zero-based-coords">zero_based_coords<a class="anchor" aria-label="anchor" href="#arg-zero-based-coords"></a></dt>
<dd><p>Whether to convert the ranges from a 1-based end-inclusive
coordinate system to a 0-based end-exclusive coordinate system. Defaults to true
for GRanges and false for other formats
(see this <a href="https://web.archive.org/web/20210920203703/http://genome.ucsc.edu/blog/the-ucsc-genome-browser-coordinate-counting-systems/" class="external-link">archived UCSC blogpost</a>)</p></dd>


<dt id="arg-return-data">return_data<a class="anchor" aria-label="anchor" href="#arg-return-data"></a></dt>
<dd><p>If true, return data from just before plotting rather than a plot.</p></dd>


<dt id="arg-scale-bar">scale_bar<a class="anchor" aria-label="anchor" href="#arg-scale-bar"></a></dt>
<dd><p>Whether to include a scale bar in the top track (<code>TRUE</code> or <code>FALSE</code>)</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>Returns a combined plot of pseudobulk genome tracks. For compatability with
<code><a href="draw_trackplot_grid.html">draw_trackplot_grid()</a></code>, the extra attribute <code>$patches$labels</code> will be added to
specify the labels for each track. If <code>return_data</code> or <code>return_plot_list</code> is
<code>TRUE</code>, the return value will be modified accordingly.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="trackplot_combine.html">trackplot_combine()</a></code>, <code><a href="trackplot_gene.html">trackplot_gene()</a></code>, <code><a href="trackplot_loop.html">trackplot_loop()</a></code>, <code><a href="trackplot_scalebar.html">trackplot_scalebar()</a></code></p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">frags</span> <span class="op">&lt;-</span> <span class="fu"><a href="demo_data.html">get_demo_frags</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Use genes and blacklist to determine proper number of reads per cell</span></span></span>
<span class="r-in"><span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="read_gtf.html">read_gencode_transcripts</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"references"</span><span class="op">)</span>, release <span class="op">=</span> <span class="st">"42"</span>,</span></span>
<span class="r-in"><span>  annotation_set <span class="op">=</span> <span class="st">"basic"</span>,</span></span>
<span class="r-in"><span>  features <span class="op">=</span> <span class="st">"transcript"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">blacklist</span> <span class="op">&lt;-</span> <span class="fu"><a href="read_bed.html">read_encode_blacklist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"references"</span><span class="op">)</span>, genome<span class="op">=</span><span class="st">"hg38"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">read_counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="qc_scATAC.html">qc_scATAC</a></span><span class="op">(</span><span class="va">frags</span>, <span class="va">genes</span>, <span class="va">blacklist</span><span class="op">)</span><span class="op">$</span><span class="va">nFrags</span></span></span>
<span class="r-in"><span><span class="va">region</span> <span class="op">&lt;-</span> <span class="st">"chr4:3034877-4034877"</span></span></span>
<span class="r-in"><span><span class="va">cell_types</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"Group"</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, length.out <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="IterableFragments-methods.html">cellNames</a></span><span class="op">(</span><span class="va">frags</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu">BPCells</span><span class="fu">:::</span><span class="fu"><a href="render_plot_from_storage.html">render_plot_from_storage</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="fu">trackplot_coverage</span><span class="op">(</span><span class="va">frags</span>, <span class="va">region</span>, groups <span class="op">=</span> <span class="va">cell_types</span>, cell_read_counts <span class="op">=</span> <span class="va">read_counts</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  width <span class="op">=</span> <span class="fl">6</span>, height <span class="op">=</span> <span class="fl">3</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="trackplot_coverage-1.png" alt="" width="700" height="433"></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Created by Benjamin Parks.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>



  <img src="https://plausible.benparks.net/flask-plausible/bpcells-docs.png" style="position:absolute;"></body></html>

