<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="BPCells">
<title>Performance Benchmarks • BPCells</title>
<script src="../../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../../pkgdown.js"></script><link href="../../extra.css" rel="stylesheet">
<meta property="og:title" content="Performance Benchmarks">
<meta property="og:description" content="BPCells">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><script defer data-domain="benparks.net" src="https://plausible.benparks.net/js/visit-counts.js"></script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../../index.html">BPCells</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../../articles/pbmc3k.html">Get Started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../../articles/pbmc3k.html">Basic tutorial</a>
    <a class="dropdown-item" href="../../articles/web-only/benchmarks.html">Performance Benchmarks</a>
    <a class="dropdown-item" href="../../articles/web-only/how-it-works.html">How BPCells works</a>
    <div class="dropdown-divider"></div>
    <a class="dropdown-item" href="../../articles/index.html">More articles...</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../../news/index.html">News</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/bnprks/BPCells/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Performance Benchmarks</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/bnprks/BPCells/blob/HEAD/vignettes/web-only/benchmarks.Rmd" class="external-link"><code>vignettes/web-only/benchmarks.Rmd</code></a></small>
      <div class="d-none name"><code>benchmarks.Rmd</code></div>
    </div>

    
    
<p>Note: these performance benchmarks are preliminary while our
manuscript is in preparation, though every effort has been made to
present a fair comparison with other tools. Fortunately, it is
straightforward to install and test BPCells on your own dataset to
replicate our claims.</p>
<div class="section level2">
<h2 id="rna-seq-normalization-pca">RNA-seq normalization + PCA<a class="anchor" aria-label="anchor" href="#rna-seq-normalization-pca"></a>
</h2>
<p>Because BPCells can perform all operations streaming from disk, it is
able to use dramatically less memory for operations that read all matrix
values. This chunking strategy is similar to tools like DelayedArray. We
have extensively optimized the C++ code to avoid slowdowns commonly
caused by using disk-backed computations.</p>
<p>In the benchmark below, we show the time and memory to perform a
basic workflow of data import, normalization, variable gene selection,
and PCA. We reproduce the PCA results of existing tools to within
numerical precision, starting from raw counts<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content="&lt;p&gt;In this case, our PCA embeddings have a mean relative
difference of about 0.02% compared to the original tool. In this
benchmark we are matching the results of
&lt;code&gt;scanpy.pp.recipe_zheng17&lt;/code&gt;, though we can also match the
results of Seurat’s analysis to similar precision.&lt;/p&gt;"><sup>1</sup></a>.</p>
<p><img src="benchmark-plots/rna_memory.png"></p>
<p>Note that Seurat in this plot crashes on the 1.3M cell dataset due to
R sparse matrices not supporting over roughly 2 billion non-zero entries
in a matrix. By utilizing Seurat v5’s BPCells integration, it is
possible to perform analyses on million cell datasets without
downsampling.</p>
</div>
<div class="section level2">
<h2 id="bitpacking-compression">Bitpacking compression<a class="anchor" aria-label="anchor" href="#bitpacking-compression"></a>
</h2>
<div class="section level3">
<h3 id="counts-matrices-rna-or-atac">Counts matrices (RNA or ATAC)<a class="anchor" aria-label="anchor" href="#counts-matrices-rna-or-atac"></a>
</h3>
<p>Single cell counts matrices are stored as sparse matrices. With
traditional formats like AnnData or 10x hdf5 feature matrices, there is
an inherent tradeoff between storage size and speed of loading.
Gzip-compressed files are small, but slow to read.</p>
<p>BPCells uses bitpacking compression to provide 4-6x smaller storage
for counts matrices compared to an uncompressed binary format
(e.g. AnnData’s default h5ad settings<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content="&lt;p&gt;Note that AnnData supports a wide variety compression
algorithms, though for counts matrices BPCells bitpacking is also faster
than lz4 and zstd read/write.&lt;/p&gt;"><sup>2</sup></a>). A single thread can load the matrix data
faster than reading uncompressed data in many cases, due to the smaller
data size and extremely fast decompression<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content="&lt;p&gt;This holds true for file storage that loads at under
~2GB/s, which is ~8x faster than a hard drive, and 16x faster than a
gigabit ethernet connection.&lt;/p&gt;"><sup>3</sup></a>.</p>
<p>Benchmark storing + loading a 1.3M cell RNA-seq experiment from 10x
Genomics. <img src="benchmark-plots/compression-1M-example.png"></p>
</div>
<div class="section level3">
<h3 id="fragment-alignments-atac">Fragment alignments (ATAC)<a class="anchor" aria-label="anchor" href="#fragment-alignments-atac"></a>
</h3>
<p>Single cell ATAC-seq data is stored as lists of fragment alignments,
including chromosome, start+end coordinates, and cell ID. The storage
space for the BPCells bitpacked format is about 2x smaller than a 10x
fragments.tsv.gz file. The storage format includes a lightweight
genome-position index to enable fast loading of data from a selected
genome location.</p>
<p>BPCells can decompress about 70 million fragments per second with a
single core, compared to about 5 million fragments per second with a 10x
fragments.tsv.gz file (14x faster). Compared to ArchR arrow files,
BPCells is also about 2x smaller with ~10x faster import times from
fragments.tsv.gz files.</p>
<p><img src="benchmark-plots/fragment-storage.png"></p>
</div>
</div>
<div class="section level2">
<h2 id="atac-seq-overlap-calculations">ATAC-seq overlap calculations<a class="anchor" aria-label="anchor" href="#atac-seq-overlap-calculations"></a>
</h2>
<p>BPCells stores fragments in a genome-indexed format that also enables
fast seeking by genome location. This helps for visualization of
selected genomic loci, as well as fast calculations of cell x peak
matrices. It is often necessary to re-quantify the cell x peak matrices
from fragments when comparing datasets with different original peak
calls. Our implementation is about 40x faster than ArchR when
calculating overlaps at 100k peaks across the genome, and over 500x
faster when calculating overlaps at just 10 peaks. This is due to the
fact that BPCells has fragments sorted by genome coordinate, whereas
arrow files sort by cell then genome coordinate.</p>
<p><img src="benchmark-plots/peak-matrix-wide.png"></p>
<p>Our peak matrix overlap calculation runs at a rate of about 25
million overlapping fragments per second, which is fast enough that it
would bottleneck on I/O speed rather than computation when reading from
a 250MB/s hard drive if bitpacking compression were not available to
reduce the I/O bandwidth.</p>
</div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Benjamin Parks.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  <img src="https://plausible.benparks.net/flask-plausible/bpcells-docs.png" style="position:absolute;">
</body>
</html>
