<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Basic tutorial • BPCells</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Basic tutorial">
<script defer data-domain="benparks.net" src="https://plausible.benparks.net/js/visit-counts.js"></script><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.16.15/dist/katex.min.css" integrity="sha384-Htz9HMhiwV8GuQ28Xr9pEs1B4qJiYu/nYLLwlDklR53QibDfmQzi7rYxXhMH/5/u" crossorigin="anonymous">
<!-- The loading of KaTeX is deferred to speed up page rendering --><script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.15/dist/katex.min.js" integrity="sha384-bxmi2jLGCvnsEqMuYLKE/KsVCxV3PqmKeK6Y6+lmNXBry6+luFkEOsmp5vD9I/7+" crossorigin="anonymous"></script><!-- To automatically render math in text elements, include the auto-render extension: --><script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.15/dist/contrib/auto-render.min.js" integrity="sha384-hCXGrW6PitJEwbkoStFjeJxv+fSOOQKOPbJxSfM6G5sWZjAyWhXiTIIAmQqnlLlh" crossorigin="anonymous" onload="renderMathInElement(document.body);"></script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">BPCells</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.3.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="../articles/pbmc3k.html">Get Started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/pbmc3k.html">Basic tutorial</a></li>
    <li><a class="dropdown-item" href="../articles/web-only/benchmarks.html">Performance Benchmarks</a></li>
    <li><a class="dropdown-item" href="../articles/web-only/how-it-works.html">How BPCells works</a></li>
    <li><a class="dropdown-item" href="../articles/web-only/programming-efficiency.html">Efficiency tips</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/index.html">More articles...</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">News</a></li>
<li class="nav-item"><a class="nav-link" href="../python/index.html"><span class="fa fa-arrow-up-right-from-square"></span> Python Docs</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/bnprks/BPCells" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Basic tutorial</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/bnprks/BPCells/tree/main/r/vignettes/pbmc3k.Rmd" class="external-link"><code>vignettes/pbmc3k.Rmd</code></a></small>
      <div class="d-none name"><code>pbmc3k.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>In this tutorial, we will will:</p>
<ul>
<li><a href="#data-loading">Load RNA and ATAC-seq data from a 10x
multiome experiment</a></li>
<li><a href="#filter-for-high-quality-cells">Filter for high-quality
cells</a></li>
<li><a href="#rna-normalization-pca-and-umap">RNA PCA + UMAP
dimensionality reduction</a></li>
<li><a href="#clustering">Unbiased clustering</a></li>
<li><a href="#visualize-marker-genes">Visualize marker genes to annotate
clusters</a></li>
<li><a href="#atac-seq-peaks">Call ATAC-seq peaks</a></li>
<li><a href="#atac-normalization-pca-and-umap">ATAC PCA + UMAP
dimensionality reduction</a></li>
<li><a href="#motif-footprinting">Visualize transcription factor
footprints</a></li>
<li><a href="#genome-accessibility-tracks">Plot accessibility genome
tracks</a></li>
</ul>
<p>This tutorial is a work-in-progress, inspired by Seurat’s <a href="https://satijalab.org/seurat/articles/pbmc3k_tutorial.html" class="external-link">PBMC
3k</a> clustering tutorial.</p>
</div>
<div class="section level2">
<h2 id="setup">Setup<a class="anchor" aria-label="anchor" href="#setup"></a>
</h2>
<div class="section level3">
<h3 id="install-packages">Install packages<a class="anchor" aria-label="anchor" href="#install-packages"></a>
</h3>
<details><summary><strong>Install analysis-specific packages</strong>
</summary><p>Install cran dependencies:</p>
<ul>
<li>irlba (PCA)</li>
<li>uwot (UMAP)</li>
<li>RcppHNSW (clustering)</li>
<li>igraph (clustering)</li>
<li>BiocManager (access bioconductor packages)</li>
<li>ggplot2 version &lt;=3.3.5 or &gt;=3.4.1 (hexbin was broken for
versions 3.3.6-3.4.0)</li>
</ul>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"irlba"</span>, <span class="st">"uwot"</span>, <span class="st">"RcppHNSW"</span>, <span class="st">"igraph"</span>, <span class="st">"BiocManager"</span>, <span class="st">"remotes"</span>, <span class="st">"ggplot2"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Bioconductor dependencies:</p>
<ul>
<li>BSgenome.Hsapiens.UCSC.hg38 (TF motif scanning)</li>
</ul>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"BSgenome.Hsapiens.UCSC.hg38"</span><span class="op">)</span></span></code></pre></div>
<p>Github:</p>
<ul>
<li>
<a href="https://github.com/GreenleafLab/motifmatchr" class="external-link">motifmatchr</a> (TF
motif scanning)</li>
<li>
<a href="https://github.com/GreenleafLab/chromVARmotifs" class="external-link">chromVARmotifs</a>
(TF motif database)</li>
</ul>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GreenleafLab/motifmatchr"</span>, <span class="st">"GreenleafLab/chromVARmotifs"</span><span class="op">)</span>, repos<span class="op">=</span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/repositories.html" class="external-link">repositories</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</details><details><summary><strong>Install BPCells</strong>
</summary><div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"bnprks/BPCells/r"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</details>
</div>
<div class="section level3">
<h3 id="set-up-analysis-folder">Set up analysis folder<a class="anchor" aria-label="anchor" href="#set-up-analysis-folder"></a>
</h3>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bnprks.github.io/BPCells">BPCells</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="op">{</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Substitute your preferred working directory for data_dir</span></span>
<span><span class="va">data_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"pbmc-3k"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">data_dir</span>, recursive <span class="op">=</span> <span class="cn">TRUE</span>, showWarnings <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">setwd</a></span><span class="op">(</span><span class="va">data_dir</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="download-data">Download data<a class="anchor" aria-label="anchor" href="#download-data"></a>
</h3>
<p>Next, we download this <a href="https://www.10xgenomics.com/resources/datasets/pbmc-from-a-healthy-donor-granulocytes-removed-through-cell-sorting-3-k-1-standard-2-0-0" class="external-link">3k
PBMC dataset</a> from 10x Genomics into a temporary directory.</p>
<p>The files are about 500MB large when combined</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">url_base</span> <span class="op">&lt;-</span> <span class="st">"https://cf.10xgenomics.com/samples/cell-arc/2.0.0/pbmc_granulocyte_sorted_3k/"</span></span>
<span><span class="va">rna_raw_url</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">url_base</span>, <span class="st">"pbmc_granulocyte_sorted_3k_raw_feature_bc_matrix.h5"</span><span class="op">)</span></span>
<span><span class="va">atac_raw_url</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">url_base</span>, <span class="st">"pbmc_granulocyte_sorted_3k_atac_fragments.tsv.gz"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Increase download timeout from 60 seconds to 5 minutes</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>timeout<span class="op">=</span><span class="fl">300</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Only download files if we haven't downloaded already</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="st">"pbmc_3k_10x.h5"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="va">rna_raw_url</span>, <span class="st">"pbmc_3k_10x.h5"</span>, mode<span class="op">=</span><span class="st">"wb"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="st">"pbmc_3k_10x.fragments.tsv.gz"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="va">atac_raw_url</span>, <span class="st">"pbmc_3k_10x.fragments.tsv.gz"</span>, mode<span class="op">=</span><span class="st">"wb"</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="data-loading">Data Loading<a class="anchor" aria-label="anchor" href="#data-loading"></a>
</h2>
<p>First, we convert the raw data inputs from a 10x format into a
bitpacked compressed format stored as binary files on disk.</p>
<p>BPCells can still read the data if we don’t convert the format, but
certain ATAC-seq functionality will run much faster on the converted
data.</p>
<p>Convert the RNA matrix:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Check if we already ran import</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="st">"pbmc_3k_rna_raw"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">mat_raw</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/open_matrix_10x_hdf5.html">open_matrix_10x_hdf5</a></span><span class="op">(</span><span class="st">"pbmc_3k_10x.h5"</span>, feature_type<span class="op">=</span><span class="st">"Gene Expression"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="../reference/matrix_io.html">write_matrix_dir</a></span><span class="op">(</span><span class="st">"pbmc_3k_rna_raw"</span><span class="op">)</span></span>
<span><span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>  <span class="va">mat_raw</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/matrix_io.html">open_matrix_dir</a></span><span class="op">(</span><span class="st">"pbmc_3k_rna_raw"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="va">mat_raw</span></span></code></pre></div>
<pre><code><span><span class="co">## 36601 x 650165 IterableMatrix object with class MatrixDir</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Row names: ENSG00000243485, ENSG00000237613 ... ENSG00000277196</span></span>
<span><span class="co">## Col names: AAACAGCCAAACAACA-1, AAACAGCCAAACATAG-1 ... TTTGTTGGTTTGTTGC-1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Data type: uint32_t</span></span>
<span><span class="co">## Storage order: column major</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Queued Operations:</span></span>
<span><span class="co">## 1. Load compressed matrix from directory /home/runner/work/BPCells/BPCells/r/vignettes/pbmc-3k-data/pbmc_3k_rna_raw</span></span></code></pre>
<p>Convert the ATAC-seq fragments</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Check if we already ran import</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="st">"pbmc_3k_frags"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">frags_raw</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/open_fragments_10x.html">open_fragments_10x</a></span><span class="op">(</span><span class="st">"pbmc_3k_10x.fragments.tsv.gz"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>      <span class="fu"><a href="../reference/fragment_io.html">write_fragments_dir</a></span><span class="op">(</span><span class="st">"pbmc_3k_frags"</span><span class="op">)</span></span>
<span><span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>  <span class="va">frags_raw</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fragment_io.html">open_fragments_dir</a></span><span class="op">(</span><span class="st">"pbmc_3k_frags"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="va">frags_raw</span></span></code></pre></div>
<pre><code><span><span class="co">## IterableFragments object of class "FragmentsDir"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Cells: 462264 cells with names TTTAGCAAGGTAGCTT-1, GCCTTTGGTTGGTTCT-1 ... ATCACCCTCCATAATG-1</span></span>
<span><span class="co">## Chromosomes: 39 chromosomes with names chr1, chr10 ... KI270713.1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Queued Operations:</span></span>
<span><span class="co">## 1. Read compressed fragments from directory /home/runner/work/BPCells/BPCells/r/vignettes/pbmc-3k-data/pbmc_3k_frags</span></span></code></pre>
<details><summary><strong>How much storage space did we save with bitpacking
compression?</strong>
</summary><p>ATAC storage space dropped from 468 MB for the gzipped 10x file to
209 MB for the bitpacked storage.</p>
<p>RNA storage space dropped from 51.2 MB for the 10x hdf5 file with
gzip compression to 33.5 MB using bitpacking compression. In this case,
the storage space is a little misleading since 39% of the bitpacked
storage is spent on the gene + cell names. In this case both 10x
compressed hdf5 and bitpacking compression are 4-6x smaller than an
uncompressed sparse matrix format such as AnnData h5ad’s default.</p>
Aside from compact storage, bitpacking compression also enables very
fast read speeds which is useful for downstream calculations.
</details>
</div>
<div class="section level2">
<h2 id="filter-for-high-quality-cells">Filter for high-quality cells<a class="anchor" aria-label="anchor" href="#filter-for-high-quality-cells"></a>
</h2>
<div class="section level3">
<h3 id="rna-seq-filtering">RNA-seq filtering<a class="anchor" aria-label="anchor" href="#rna-seq-filtering"></a>
</h3>
<p>We use a simple minimum read threshold for RNA-seq quality. The
cutoff we choose is just below the first knee of our log-log plot of
reads vs. barcode rank, which separates cells from empty droplets.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">reads_per_cell</span> <span class="op">&lt;-</span> <span class="fu">Matrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums-methods.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">mat_raw</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plot_read_count_knee.html">plot_read_count_knee</a></span><span class="op">(</span><span class="va">reads_per_cell</span>, cutoff <span class="op">=</span> <span class="fl">1e3</span><span class="op">)</span></span></code></pre></div>
<p><img src="pbmc3k_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="atac-seq-filtering">ATAC-seq filtering<a class="anchor" aria-label="anchor" href="#atac-seq-filtering"></a>
</h3>
<div class="section level4">
<h4 id="download-reference-annotations">Download reference annotations<a class="anchor" aria-label="anchor" href="#download-reference-annotations"></a>
</h4>
<p>We fetch reference information necessary to calculate quality-control
statistics. By default, this fetches the latest annotations for
hg38.</p>
<p>Since fetching the references involves downloading gtf and bed files,
we provide the name of a directory to save the files in. This also
allows us to skip re-downloading the same files next time.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_gtf.html">read_gencode_transcripts</a></span><span class="op">(</span></span>
<span>  <span class="st">"./references"</span>, </span>
<span>  release<span class="op">=</span><span class="st">"42"</span>, </span>
<span>  transcript_choice<span class="op">=</span><span class="st">"MANE_Select"</span>,</span>
<span>  annotation_set <span class="op">=</span> <span class="st">"basic"</span>, </span>
<span>  features<span class="op">=</span><span class="st">"transcript"</span> <span class="co"># Make sure to set this so we don't get exons as well</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">genes</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 6 × 13</span></span></span>
<span><span class="co">##   chr   source feature     start    end score strand frame gene_id     gene_type</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> chr1  HAVANA transcript  <span style="text-decoration: underline;">65</span>418  <span style="text-decoration: underline;">71</span>585 .     +      .     ENSG000001… protein_…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> chr1  HAVANA transcript <span style="text-decoration: underline;">450</span>739 <span style="text-decoration: underline;">451</span>678 .     -      .     ENSG000002… protein_…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> chr1  HAVANA transcript <span style="text-decoration: underline;">685</span>715 <span style="text-decoration: underline;">686</span>654 .     -      .     ENSG000002… protein_…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">4</span> chr1  HAVANA transcript <span style="text-decoration: underline;">923</span>922 <span style="text-decoration: underline;">944</span>574 .     +      .     ENSG000001… protein_…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">5</span> chr1  HAVANA transcript <span style="text-decoration: underline;">944</span>202 <span style="text-decoration: underline;">959</span>256 .     -      .     ENSG000001… protein_…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">6</span> chr1  HAVANA transcript <span style="text-decoration: underline;">960</span>583 <span style="text-decoration: underline;">965</span>719 .     +      .     ENSG000001… protein_…</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 3 more variables: gene_name &lt;chr&gt;, transcript_id &lt;chr&gt;, MANE_Select &lt;lgl&gt;</span></span></span></code></pre>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">blacklist</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_bed.html">read_encode_blacklist</a></span><span class="op">(</span><span class="st">"./references"</span>, genome<span class="op">=</span><span class="st">"hg38"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">blacklist</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 6 × 4</span></span></span>
<span><span class="co">##   chr      start      end reason            </span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> chr10        0    <span style="text-decoration: underline;">45</span>700 Low Mappability   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> chr10 38<span style="text-decoration: underline;">481</span>300 38<span style="text-decoration: underline;">596</span>500 High Signal Region</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> chr10 38<span style="text-decoration: underline;">782</span>600 38<span style="text-decoration: underline;">967</span>900 High Signal Region</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">4</span> chr10 39<span style="text-decoration: underline;">901</span>300 41<span style="text-decoration: underline;">712</span>900 High Signal Region</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">5</span> chr10 41<span style="text-decoration: underline;">838</span>900 42<span style="text-decoration: underline;">107</span>300 High Signal Region</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">6</span> chr10 42<span style="text-decoration: underline;">279</span>400 42<span style="text-decoration: underline;">322</span>500 High Signal Region</span></span></code></pre>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">chrom_sizes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_ucsc_chrom_sizes.html">read_ucsc_chrom_sizes</a></span><span class="op">(</span><span class="st">"./references"</span>, genome<span class="op">=</span><span class="st">"hg38"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">chrom_sizes</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 6 × 3</span></span></span>
<span><span class="co">##   chr   start       end</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> chr1      0 248<span style="text-decoration: underline;">956</span>422</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> chr2      0 242<span style="text-decoration: underline;">193</span>529</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> chr3      0 198<span style="text-decoration: underline;">295</span>559</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">4</span> chr4      0 190<span style="text-decoration: underline;">214</span>555</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">5</span> chr5      0 181<span style="text-decoration: underline;">538</span>259</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">6</span> chr6      0 170<span style="text-decoration: underline;">805</span>979</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="calculate-atac-seq-quality-control-metrics">Calculate ATAC-seq quality-control metrics<a class="anchor" aria-label="anchor" href="#calculate-atac-seq-quality-control-metrics"></a>
</h4>
<p>We can calculate several built-in quality control metrics for each
barcode, including number of fragments and TSS enrichment. These
calculations are fully compatible with ArchR’s methodology for quality
control statistics.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">atac_qc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qc_scATAC.html">qc_scATAC</a></span><span class="op">(</span><span class="va">frags_raw</span>, <span class="va">genes</span>, <span class="va">blacklist</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">atac_qc</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 6 × 10</span></span></span>
<span><span class="co">##   cellName  TSSEnrichment nFrags subNucleosomal monoNucleosomal multiNucleosomal</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> TTTAGCAA…        45.1    <span style="text-decoration: underline;">16</span>363           <span style="text-decoration: underline;">8</span>069            <span style="text-decoration: underline;">5</span>588             <span style="text-decoration: underline;">2</span>706</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> GCCTTTGG…         0.198      3              1               2                0</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> AGCCGGTT…        30.9    <span style="text-decoration: underline;">33</span>313          <span style="text-decoration: underline;">15</span>855           <span style="text-decoration: underline;">11</span>868             <span style="text-decoration: underline;">5</span>590</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">4</span> TGATTAGT…        41.9    <span style="text-decoration: underline;">11</span>908           <span style="text-decoration: underline;">6</span>103            <span style="text-decoration: underline;">3</span>817             <span style="text-decoration: underline;">1</span>988</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">5</span> ATTGACTC…        43.9    <span style="text-decoration: underline;">13</span>075           <span style="text-decoration: underline;">6</span>932            <span style="text-decoration: underline;">4</span>141             <span style="text-decoration: underline;">2</span>002</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">6</span> CGTTAGGT…        31.5    <span style="text-decoration: underline;">14</span>874           <span style="text-decoration: underline;">6</span>833            <span style="text-decoration: underline;">5</span>405             <span style="text-decoration: underline;">2</span>636</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 4 more variables: ReadsInTSS &lt;dbl&gt;, ReadsFlankingTSS &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   ReadsInPromoter &lt;dbl&gt;, ReadsInBlacklist &lt;dbl&gt;</span></span></span></code></pre>
<p>One of the key ways to identify high-quality cells in ATAC-seq data
is to plot the number of fragments vs. TSS Enrichment. This plot puts
empty droplets in the bottom-left quadrant, low-quality/dead cells in
the bottom-right quadrant, and high-quality cells in the top-right
quadrant. From a flow-cytometry perspective, we use the bottom-left
group of empty droplets as a negative control to help set our
cutoffs.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_tss_scatter.html">plot_tss_scatter</a></span><span class="op">(</span><span class="va">atac_qc</span>, min_frags<span class="op">=</span><span class="fl">1000</span>, min_tss<span class="op">=</span><span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<p><img src="pbmc3k_files/figure-html/unnamed-chunk-15-1.png" width="700"></p>
<details><summary><strong>Why is this TSS Enrichment number different from ArchR?</strong>
</summary>
The TSS Enrichment number varies depending on the TSS annotation set
used, as a stricter definition of what counts as a TSS will tend to give
higher TSS enrichment. BPCells uses the exact same formula as ArchR to
calculate TSS enrichment, so if you pass the ArchR TSS set to BPCells
you’ll get identical results. However, in this case we’ve used a
somewhat more strict definition of TSS by taking just the MANE select
transcripts from Gencode.
</details><details><summary><strong>Why are there no cells in the top-left quadrant?</strong>
</summary><p>Due to a thresholding that ArchR’s formula applies in the denominator
of the TSS Enrichment calculation, low-read cells can’t be assigned a
very high TSS Enrichment value.</p>
<p>To plot the TSS enrichment without this thresholding, do the
following:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">atac_qc</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>TSSEnrichment<span class="op">=</span><span class="va">ReadsInTSS</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">pmax</a></span><span class="op">(</span><span class="fl">1</span>,<span class="va">ReadsFlankingTSS</span><span class="op">)</span> <span class="op">*</span> <span class="fl">200</span><span class="op">/</span><span class="fl">101</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/plot_tss_scatter.html">plot_tss_scatter</a></span><span class="op">(</span>min_frags<span class="op">=</span><span class="fl">2000</span>, min_tss<span class="op">=</span><span class="fl">20</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title<span class="op">=</span><span class="st">"Raw TSS Enrichment"</span><span class="op">)</span></span></code></pre></div>
<p><img src="pbmc3k_files/figure-html/unnamed-chunk-16-1.png" width="700"></p>
<p>Note that the <code>200/101</code> fraction accounts for
<code>ReadsInTSS</code> drawing from 101-bp windows, and
ReadsFlankingTSS drawing from 2x100-bp windows. This results in more
low-read droplets measuring high TSS Enrichment, so we use slightly
adjusted cutoffs.</p>
ArchR’s formula is equivalent to replacing
<code>pmax(1,ReadsFlankingTSS)</code> with
<code>pmax(20,ReadsFlankingTSS)</code> in the above code.
</details><p>We can also plot some sample-level quality control plots.</p>
<p>On the left, the fragment length distribution shows three broad bumps
corresponding to nucleosome spacing (147bp), and smaller wiggles
corresponding to DNA winding (11.5bp).</p>
<p>On the right, the TSS enrichment profile shows a strong enrichment of
signal at transcription start sites, as well as a small asymmetrical
bump downstream of the TSS after the +1 nucleosome.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_fragment_length.html">plot_fragment_length</a></span><span class="op">(</span><span class="va">frags_raw</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="../reference/plot_tss_profile.html">plot_tss_profile</a></span><span class="op">(</span><span class="va">frags_raw</span>, <span class="va">genes</span><span class="op">)</span></span></code></pre></div>
<p><img src="pbmc3k_files/figure-html/unnamed-chunk-17-1.png" width="700"></p>
</div>
</div>
<div class="section level3">
<h3 id="select-high-quality-cells">Select high-quality cells<a class="anchor" aria-label="anchor" href="#select-high-quality-cells"></a>
</h3>
<p>We take cells that pass our minimum RNA reads, minimum ATAC reads,
and minimum TSS Enrichment cutoffs.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pass_atac</span> <span class="op">&lt;-</span> <span class="va">atac_qc</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">nFrags</span> <span class="op">&gt;</span> <span class="fl">1000</span>, <span class="va">TSSEnrichment</span> <span class="op">&gt;</span> <span class="fl">10</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">cellName</span><span class="op">)</span></span>
<span><span class="va">pass_rna</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">mat_raw</span><span class="op">)</span><span class="op">[</span><span class="fu">Matrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums-methods.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">mat_raw</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">1e3</span><span class="op">]</span></span>
<span><span class="va">keeper_cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://generics.r-lib.org/reference/setops.html" class="external-link">intersect</a></span><span class="op">(</span><span class="va">pass_atac</span>, <span class="va">pass_rna</span><span class="op">)</span></span></code></pre></div>
<p>We subset the RNA and ATAC input data to just the cells passing
filter. And for RNA, we subset to genes with at least 3 reads. This
subset operation also puts the cells in a matching order which
simplifies cross-modality calculations later on.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">frags</span> <span class="op">&lt;-</span> <span class="va">frags_raw</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="../reference/select_cells.html">select_cells</a></span><span class="op">(</span><span class="va">keeper_cells</span><span class="op">)</span></span>
<span></span>
<span><span class="va">keeper_genes</span> <span class="op">&lt;-</span> <span class="fu">Matrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums-methods.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">mat_raw</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">3</span></span>
<span><span class="va">mat</span> <span class="op">&lt;-</span> <span class="va">mat_raw</span><span class="op">[</span><span class="va">keeper_genes</span>,<span class="va">keeper_cells</span><span class="op">]</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="rna-normalization-pca-and-umap">RNA Normalization, PCA and UMAP<a class="anchor" aria-label="anchor" href="#rna-normalization-pca-and-umap"></a>
</h2>
<div class="section level3">
<h3 id="matrix-normalization">Matrix normalization<a class="anchor" aria-label="anchor" href="#matrix-normalization"></a>
</h3>
<p>Here, we walk through the Seurat-style matrix normalization
calculations manually, though soon there will be helper functions to
simplify the process.</p>
<p>First we log-normalize, roughly equivalent to
<code>Seurat::NormalizeData</code></p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Normalize by reads-per-cell</span></span>
<span><span class="va">mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mat_norm.html">multiply_cols</a></span><span class="op">(</span><span class="va">mat</span>, <span class="fl">1</span><span class="op">/</span><span class="fu">Matrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums-methods.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Log normalization</span></span>
<span><span class="va">mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log1p</a></span><span class="op">(</span><span class="va">mat</span> <span class="op">*</span> <span class="fl">10000</span><span class="op">)</span> <span class="co"># Log normalization</span></span></code></pre></div>
<p>Next we pick out variable genes:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">stats</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/matrix_stats.html">matrix_stats</a></span><span class="op">(</span><span class="va">mat</span>, row_stats<span class="op">=</span><span class="st">"variance"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># To keep the example small, we'll do a very naive variable gene selection</span></span>
<span><span class="va">variable_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">stats</span><span class="op">$</span><span class="va">row_stats</span><span class="op">[</span><span class="st">"variance"</span>,<span class="op">]</span>, decreasing<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">1000</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">mat_norm</span> <span class="op">&lt;-</span> <span class="va">mat</span><span class="op">[</span><span class="va">variable_genes</span>,<span class="op">]</span></span></code></pre></div>
<p>If we look at our normalized matrix object, we can see we have quite
a few math operations queued up which are performed on-the-fly as
needed.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mat_norm</span></span></code></pre></div>
<pre><code><span><span class="co">## 1000 x 2600 IterableMatrix object with class TransformLog1p</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Row names: ENSG00000078369, ENSG00000116251 ... ENSG00000212907</span></span>
<span><span class="co">## Col names: TTTAGCAAGGTAGCTT-1, AGCCGGTTCCGGAACC-1 ... TACTAAGTCCAATAGC-1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Data type: double</span></span>
<span><span class="co">## Storage order: column major</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Queued Operations:</span></span>
<span><span class="co">## 1. Load compressed matrix from directory /home/runner/work/BPCells/BPCells/r/vignettes/pbmc-3k-data/pbmc_3k_rna_raw</span></span>
<span><span class="co">## 2. Select rows: 87, 171 ... 36568 and cols: 640783, 89020 ... 504383</span></span>
<span><span class="co">## 3. Convert type from uint32_t to double</span></span>
<span><span class="co">## 4. Scale by 1e+04</span></span>
<span><span class="co">## 5. Scale columns by 0.000221, 0.000118 ... 0.000177</span></span>
<span><span class="co">## 6. Transform log1p</span></span></code></pre>
<p>To improve performance of the downstream PCA, we save this sparse
normalized matrix to a temporary file just prior to normalizations that
would make the matrix dense. This saves storage space while preventing
us from having to re-calculate the queued operations several-hundred
times during the PCA optimization iterations.</p>
<p>In this case, the matrix is quite small so we’ll just store it in
memory. In a larger example we could swap this for
<code>write_matrix_dir(tempfile("mat"))</code></p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mat_norm</span> <span class="op">&lt;-</span> <span class="va">mat_norm</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="../reference/matrix_io.html">write_matrix_memory</a></span><span class="op">(</span>compress<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>Finally, we perform z-score normalization which makes the matrix
dense.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gene_means</span> <span class="op">&lt;-</span> <span class="va">stats</span><span class="op">$</span><span class="va">row_stats</span><span class="op">[</span><span class="st">"mean"</span>,<span class="va">variable_genes</span><span class="op">]</span></span>
<span><span class="va">gene_vars</span> <span class="op">&lt;-</span> <span class="va">stats</span><span class="op">$</span><span class="va">row_stats</span><span class="op">[</span><span class="st">"variance"</span>, <span class="va">variable_genes</span><span class="op">]</span></span>
<span><span class="va">mat_norm</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="va">mat_norm</span> <span class="op">-</span> <span class="va">gene_means</span><span class="op">)</span> <span class="op">/</span> <span class="va">gene_vars</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="pca-and-umap">PCA and UMAP<a class="anchor" aria-label="anchor" href="#pca-and-umap"></a>
</h3>
<p>PCA can be performed with a standard solver like that in
<code>irlba</code>, though BPCells also provides a C++-level solver
based on the <code>Spectra</code> package which has built-in
parallelization support.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">svd</span> <span class="op">&lt;-</span> <span class="fu">BPCells</span><span class="fu">::</span><span class="fu"><a href="../reference/svds.html">svds</a></span><span class="op">(</span><span class="va">mat_norm</span>, k<span class="op">=</span><span class="fl">50</span><span class="op">)</span></span>
<span><span class="co"># Alternate option: irlba::irlba(mat_norm, nv=50)</span></span>
<span><span class="va">pca</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mat_norm.html">multiply_cols</a></span><span class="op">(</span><span class="va">svd</span><span class="op">$</span><span class="va">v</span>, <span class="va">svd</span><span class="op">$</span><span class="va">d</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"PCA dimensions: %s\n"</span>, <span class="fu"><a href="https://rdrr.io/r/base/toString.html" class="external-link">toString</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">pca</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">pca</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## PCA dimensions: 2600, 50</span></span>
<span><span class="co">##           [,1]       [,2]       [,3]</span></span>
<span><span class="co">## [1,] 15.167732  0.8951489 -2.3650024</span></span>
<span><span class="co">## [2,]  6.599775  7.2484737  4.4369185</span></span>
<span><span class="co">## [3,] 14.621697 -1.1929478 -0.6439663</span></span>
<span><span class="co">## [4,]  8.142875  1.0977223 -2.5066235</span></span></code></pre>
<p>Next we calculate UMAP coordinates</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">12341512</span><span class="op">)</span></span>
<span><span class="va">umap</span> <span class="op">&lt;-</span> <span class="fu">uwot</span><span class="fu">::</span><span class="fu"><a href="https://jlmelville.github.io/uwot/reference/umap.html" class="external-link">umap</a></span><span class="op">(</span><span class="va">pca</span><span class="op">)</span></span>
<span><span class="va">umap</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span>,<span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##            [,1]        [,2]</span></span>
<span><span class="co">## [1,]  9.9284977   2.8806480</span></span>
<span><span class="co">## [2,] -0.7640393 -11.7900724</span></span>
<span><span class="co">## [3,]  9.8782410   3.1049877</span></span>
<span><span class="co">## [4,]  7.7622828   0.5004428</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="clustering">Clustering<a class="anchor" aria-label="anchor" href="#clustering"></a>
</h3>
<p>We perform a quick clustering as follows, based on the PCA
coordinates.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">clusts</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/knn.html">knn_hnsw</a></span><span class="op">(</span><span class="va">pca</span>, ef<span class="op">=</span><span class="fl">500</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="co"># Find approximate nearest neighbors</span></span>
<span>  <span class="fu"><a href="../reference/knn_graph.html">knn_to_snn_graph</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="co"># Convert to a SNN graph</span></span>
<span>  <span class="fu"><a href="../reference/cluster.html">cluster_graph_louvain</a></span><span class="op">(</span><span class="op">)</span> <span class="co"># Perform graph-based clustering</span></span></code></pre></div>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"Clusts length: %s\n"</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">clusts</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">clusts</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## Clusts length: 2600</span></span>
<span><span class="co">##  [1] 1 2 1 2 2 3 2 2 4 5</span></span>
<span><span class="co">## Levels: 1 2 3 4 5 6 7 8 9 10 11 12</span></span></code></pre>
<p>And now we can visualize the clusters on a UMAP:</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_embedding.html">plot_embedding</a></span><span class="op">(</span><span class="va">clusts</span>, <span class="va">umap</span><span class="op">)</span></span></code></pre></div>
<p><img src="pbmc3k_files/figure-html/unnamed-chunk-29-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="visualize-marker-genes">Visualize marker genes<a class="anchor" aria-label="anchor" href="#visualize-marker-genes"></a>
</h3>
<p>To annotate our clusters with cell types, we can plot several marker
genes overlaid onto the UMAP.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_embedding.html">plot_embedding</a></span><span class="op">(</span></span>
<span>  source <span class="op">=</span> <span class="va">mat</span>,</span>
<span>  <span class="va">umap</span>,</span>
<span>  features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"MS4A1"</span>, <span class="st">"GNLY"</span>, <span class="st">"CD3E"</span>, </span>
<span>               <span class="st">"CD14"</span>, <span class="st">"FCER1A"</span>, <span class="st">"FCGR3A"</span>, </span>
<span>               <span class="st">"LYZ"</span>, <span class="st">"CD4"</span>,<span class="st">"CD8"</span><span class="op">)</span>,</span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="pbmc3k_files/figure-html/unnamed-chunk-30-1.png" width="960"></p>
<p>We observe cluster-specific enrichment of B-cell marker MS4A1, T-cell
marker CD3E, and Monocyte marker LYZ. This allows us to make some broad
cell type groupings as follows:</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cluster_annotations</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="st">"1"</span> <span class="op">=</span> <span class="st">"T"</span>,</span>
<span>  <span class="st">"2"</span> <span class="op">=</span> <span class="st">"CD8 T"</span>,</span>
<span>  <span class="st">"3"</span> <span class="op">=</span> <span class="st">"B"</span>,</span>
<span>  <span class="st">"4"</span> <span class="op">=</span> <span class="st">"T"</span>,</span>
<span>  <span class="st">"5"</span> <span class="op">=</span> <span class="st">"NK"</span>,</span>
<span>  <span class="st">"6"</span> <span class="op">=</span> <span class="st">"Mono"</span>,</span>
<span>  <span class="st">"7"</span> <span class="op">=</span> <span class="st">"Mono"</span>,</span>
<span>  <span class="st">"8"</span> <span class="op">=</span> <span class="st">"Mono"</span>,</span>
<span>  <span class="st">"9"</span> <span class="op">=</span> <span class="st">"T"</span>,</span>
<span>  <span class="st">"10"</span> <span class="op">=</span> <span class="st">"DC"</span>,</span>
<span>  <span class="st">"11"</span> <span class="op">=</span> <span class="st">"Mono"</span>,</span>
<span>  <span class="st">"12"</span> <span class="op">=</span> <span class="st">"DC"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">cell_types</span> <span class="op">&lt;-</span> <span class="va">cluster_annotations</span><span class="op">[</span><span class="va">clusts</span><span class="op">]</span></span>
<span><span class="fu"><a href="../reference/plot_embedding.html">plot_embedding</a></span><span class="op">(</span><span class="va">cell_types</span>, <span class="va">umap</span><span class="op">)</span></span></code></pre></div>
<p><img src="pbmc3k_files/figure-html/unnamed-chunk-31-1.png" width="700"></p>
<p>We can further visualize the marker genes for each cluster using a
dot plot. As is typical for these situations, some known marker genes
are very clear, while others are less specific.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_dot.html">plot_dot</a></span><span class="op">(</span></span>
<span>  <span class="va">mat</span>, </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"MS4A1"</span>, <span class="st">"GNLY"</span>, <span class="st">"CD3E"</span>, </span>
<span>    <span class="st">"CD14"</span>, <span class="st">"FCER1A"</span>, <span class="st">"FCGR3A"</span>, </span>
<span>    <span class="st">"LYZ"</span>, <span class="st">"CD4"</span>, <span class="st">"CD8"</span><span class="op">)</span>, </span>
<span>  <span class="va">cell_types</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="pbmc3k_files/figure-html/unnamed-chunk-32-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="atac-normalization-pca-and-umap">ATAC Normalization, PCA and UMAP<a class="anchor" aria-label="anchor" href="#atac-normalization-pca-and-umap"></a>
</h2>
<p>We start with tile-based peak calling, which tests pre-determined
overlapping tile positions for significant enrichment of ATAC-seq signal
over the genome-wide background in each cell type independently. This is
faster than using a traditional peak-caller like MACS, though with the
default parameters the peaks will always be 200bp wide and the
positioning resolution is approximately +/- 30bp.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">frags_filter_blacklist</span> <span class="op">&lt;-</span> <span class="va">frags</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/select_regions.html">select_regions</a></span><span class="op">(</span><span class="va">blacklist</span>, invert_selection <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">peaks</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/call_peaks_tile.html">call_peaks_tile</a></span><span class="op">(</span><span class="va">frags_filter_blacklist</span>, <span class="va">chrom_sizes</span>, cell_groups<span class="op">=</span><span class="va">cell_types</span>,</span>
<span>                           effective_genome_size <span class="op">=</span> <span class="fl">2.8e9</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">peaks</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 6 × 7</span></span></span>
<span><span class="co">##   chr       start       end group     p_val     q_val enrichment</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> chr1   16<span style="text-decoration: underline;">644</span>600  16<span style="text-decoration: underline;">644</span>800 T     0   <span style="color: #949494;"> </span>     0   <span style="color: #949494;"> </span>          <span style="text-decoration: underline;">1</span>017.</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> chr19  18<span style="text-decoration: underline;">281</span>733  18<span style="text-decoration: underline;">281</span>933 Mono  0   <span style="color: #949494;"> </span>     0   <span style="color: #949494;"> </span>           518.</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> chr17  81<span style="text-decoration: underline;">860</span>866  81<span style="text-decoration: underline;">861</span>066 DC    7.87<span style="color: #949494;">e</span><span style="color: #BB0000;">- 63</span> 1.21<span style="color: #949494;">e</span><span style="color: #BB0000;">- 55</span>       552.</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">4</span> chr1    1<span style="text-decoration: underline;">724</span>333   1<span style="text-decoration: underline;">724</span>533 Mono  0   <span style="color: #949494;"> </span>     0   <span style="color: #949494;"> </span>           512.</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">5</span> chr1  228<span style="text-decoration: underline;">140</span>000 228<span style="text-decoration: underline;">140</span>200 NK    2.77<span style="color: #949494;">e</span><span style="color: #BB0000;">-162</span> 2.14<span style="color: #949494;">e</span><span style="color: #BB0000;">-155</span>       842.</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">6</span> chr8   30<span style="text-decoration: underline;">083</span>133  30<span style="text-decoration: underline;">083</span>333 CD8 T 4.80<span style="color: #949494;">e</span><span style="color: #BB0000;">-220</span> 7.41<span style="color: #949494;">e</span><span style="color: #BB0000;">-213</span>       744.</span></span></code></pre>
<p>Next we compute a peak matrix counting how many ATAC-seq insertions
overlap each peak. We again save it in memory rather than saving it to
disk since our dataset is quite small.</p>
<details><summary><strong>Why do we calculate <code>ordered_peaks</code>?</strong>
</summary>
For peak matrix calculation to run efficiently, peaks must be sorted in
order by chromosome and end coordinate. Although
<code>peak_matrix</code> can automatically reorder the peaks
appropriately, we show how to perform the sorting manually for this
example.
</details><div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">top_peaks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">peaks</span>, <span class="fl">50000</span><span class="op">)</span></span>
<span><span class="va">top_peaks</span> <span class="op">&lt;-</span> <span class="va">top_peaks</span><span class="op">[</span><span class="fu"><a href="../reference/order_ranges.html">order_ranges</a></span><span class="op">(</span><span class="va">top_peaks</span>, <span class="fu"><a href="../reference/IterableFragments-methods.html">chrNames</a></span><span class="op">(</span><span class="va">frags</span><span class="op">)</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="va">peak_mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/peak_matrix.html">peak_matrix</a></span><span class="op">(</span><span class="va">frags</span>, <span class="va">top_peaks</span>, mode<span class="op">=</span><span class="st">"insertions"</span><span class="op">)</span></span></code></pre></div>
<p>Next we calculate the TF-IDF normalization. This formula is the
TF-IDF variant from Stuart et al.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mat_lsi</span> <span class="op">&lt;-</span> <span class="va">peak_mat</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/mat_norm.html">multiply_cols</a></span><span class="op">(</span><span class="fl">1</span> <span class="op">/</span> <span class="fu">Matrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums-methods.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">peak_mat</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/mat_norm.html">multiply_rows</a></span><span class="op">(</span><span class="fl">1</span> <span class="op">/</span> <span class="fu">Matrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums-methods.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="va">peak_mat</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">mat_lsi</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log1p</a></span><span class="op">(</span><span class="fl">10000</span> <span class="op">*</span> <span class="va">mat_lsi</span><span class="op">)</span></span></code></pre></div>
<p>Looking at this LSI matrix, we can see the power of BPCells
performing matrix operations on-the-fly: the LSI normalization is in
fact calculated at the same time as the fragment overlap calculations
when we read this matrix. We don’t need to store any intermediate
matrices during these calculations, and even the peak matrix can be
re-calculated on-the-fly based on the fragments object saved to
disk.</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mat_lsi</span></span></code></pre></div>
<pre><code><span><span class="co">## 50000 x 2600 IterableMatrix object with class TransformLog1p</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Row names: chr1:817200-817400, chr1:827466-827666 ... chrX:155881200-155881400</span></span>
<span><span class="co">## Col names: TTTAGCAAGGTAGCTT-1, AGCCGGTTCCGGAACC-1 ... TACTAAGTCCAATAGC-1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Data type: double</span></span>
<span><span class="co">## Storage order: row major</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Queued Operations:</span></span>
<span><span class="co">## 1. Read compressed fragments from directory /home/runner/work/BPCells/BPCells/r/vignettes/pbmc-3k-data/pbmc_3k_frags</span></span>
<span><span class="co">## 2. Select 2600 cells by name: TTTAGCAAGGTAGCTT-1, AGCCGGTTCCGGAACC-1 ... TACTAAGTCCAATAGC-1</span></span>
<span><span class="co">## 3. Calculate 2600 peaks over 50000 ranges: chr1:817201-817400 ... chrX:155881201-155881400</span></span>
<span><span class="co">## 4. Convert type from uint32_t to double</span></span>
<span><span class="co">## 5. Scale by 1e+04</span></span>
<span><span class="co">## 6. Scale columns by 6.71e-05, 5.17e-05 ... 0.000801</span></span>
<span><span class="co">## 7. Scale rows by 11.1, 2.78 ... 3.77</span></span>
<span><span class="co">## 8. Transform log1p</span></span></code></pre>
<p>Just like for the RNA, we save our matrix before running PCA. On a
larger dataset, we would save to disk rather than to memory.</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mat_lsi</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/matrix_io.html">write_matrix_memory</a></span><span class="op">(</span><span class="va">mat_lsi</span>, compress<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>Finally, we will do a z-score normalization on the LSI matrix then
run PCA. This is standard practice for running a PCA, but is not
commonly done on ATAC-seq datasets due to the fact that it greatly
increases memory usage. In other methods, the 1st PC is highly
correlated to the number of reads per cell, and is thrown out as an
empirical correction.</p>
<p>Luckily, BPCells can avoid this memory usage so we can just normalize
our data and run PCA as usual</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Compute colMean and colVariance in one pass</span></span>
<span><span class="va">cell_peak_stats</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/matrix_stats.html">matrix_stats</a></span><span class="op">(</span><span class="va">mat_lsi</span>, col_stats<span class="op">=</span><span class="st">"variance"</span><span class="op">)</span><span class="op">$</span><span class="va">col_stats</span></span>
<span><span class="va">cell_means</span> <span class="op">&lt;-</span> <span class="va">cell_peak_stats</span><span class="op">[</span><span class="st">"mean"</span>,<span class="op">]</span></span>
<span><span class="va">cell_vars</span> <span class="op">&lt;-</span> <span class="va">cell_peak_stats</span><span class="op">[</span><span class="st">"variance"</span>,<span class="op">]</span></span>
<span><span class="va">mat_lsi_norm</span> <span class="op">&lt;-</span> <span class="va">mat_lsi</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/mat_norm.html">add_cols</a></span><span class="op">(</span><span class="op">-</span><span class="va">cell_means</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/mat_norm.html">multiply_cols</a></span><span class="op">(</span><span class="fl">1</span> <span class="op">/</span> <span class="va">cell_vars</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">svd_atac</span> <span class="op">&lt;-</span> <span class="fu">BPCells</span><span class="fu">::</span><span class="fu"><a href="../reference/svds.html">svds</a></span><span class="op">(</span><span class="va">mat_lsi_norm</span>, k<span class="op">=</span><span class="fl">10</span><span class="op">)</span></span>
<span><span class="va">pca_atac</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mat_norm.html">multiply_cols</a></span><span class="op">(</span><span class="va">svd_atac</span><span class="op">$</span><span class="va">v</span>, <span class="va">svd_atac</span><span class="op">$</span><span class="va">d</span><span class="op">)</span></span>
<span><span class="va">pca_atac</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##            [,1]       [,2]       [,3]      [,4]</span></span>
<span><span class="co">## [1,] -103.64071   1.553515   2.436148  21.22977</span></span>
<span><span class="co">## [2,]  -44.75342 -28.737622 -12.681591 -10.01745</span></span>
<span><span class="co">## [3,]  -90.74857   3.266168   3.627660  12.95109</span></span>
<span><span class="co">## [4,]  -90.74640  -6.447103   6.853840 -15.76629</span></span></code></pre>
<p>Next we calculate our UMAP</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">12341512</span><span class="op">)</span></span>
<span><span class="va">umap_atac</span> <span class="op">&lt;-</span> <span class="fu">uwot</span><span class="fu">::</span><span class="fu"><a href="https://jlmelville.github.io/uwot/reference/umap.html" class="external-link">umap</a></span><span class="op">(</span><span class="va">pca_atac</span><span class="op">)</span></span>
<span><span class="va">umap_atac</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span>,<span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##           [,1]      [,2]</span></span>
<span><span class="co">## [1,]  6.089028  4.207542</span></span>
<span><span class="co">## [2,] -7.985246 -1.984984</span></span>
<span><span class="co">## [3,] 11.562254  5.900722</span></span>
<span><span class="co">## [4,]  3.042405  2.246846</span></span></code></pre>
<p>Then we cluster, just like for the RNA</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">clusts_atac</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/knn.html">knn_hnsw</a></span><span class="op">(</span><span class="va">pca_atac</span>, ef<span class="op">=</span><span class="fl">500</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="co"># Find approximate nearest neighbors</span></span>
<span>  <span class="fu"><a href="../reference/knn_graph.html">knn_to_snn_graph</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="co"># Convert to a SNN graph</span></span>
<span>  <span class="fu"><a href="../reference/cluster.html">cluster_graph_louvain</a></span><span class="op">(</span><span class="op">)</span> <span class="co"># Perform graph-based clustering</span></span></code></pre></div>
<p>And we can plot our ATAC-seq embedding with ATAC-derived clusters,
and easily compare that to the RNA-derived clusters from earlier.</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_embedding.html">plot_embedding</a></span><span class="op">(</span><span class="va">clusts_atac</span>, <span class="va">umap_atac</span>, colors_discrete <span class="op">=</span> <span class="fu"><a href="../reference/palettes.html">discrete_palette</a></span><span class="op">(</span><span class="st">"ironMan"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>color<span class="op">=</span><span class="st">"none"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="../reference/plot_embedding.html">plot_embedding</a></span><span class="op">(</span><span class="va">cell_types</span>, <span class="va">umap_atac</span><span class="op">)</span></span></code></pre></div>
<p><img src="pbmc3k_files/figure-html/unnamed-chunk-42-1.png" width="700"></p>
<details><summary><strong>How did we combine the RNA clusters with the ATAC data?</strong>
</summary><p>BPCells works based on the order of cells in a matrix or fragment
object. Since our ATAC PCA rows are in the same cell order as our RNA
clusters, the datasets combine with no additional work.</p>
It does take some effort at the beginning of your analysis to make sure
the cell orders are consistent, which is what we did in our <a href="#select-high-quality-cells">cell filtering</a> step. But after
ordering the input RNA matrix and ATAC fragments, all downstream
calculations should maintain the same cell ordering by default.
</details><details><summary><strong>What if we skip z-score normalization on the LSI
matrix?</strong>
</summary><div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">svd_atac_no_norm</span> <span class="op">&lt;-</span> <span class="fu">BPCells</span><span class="fu">::</span><span class="fu"><a href="../reference/svds.html">svds</a></span><span class="op">(</span><span class="va">mat_lsi</span>, k<span class="op">=</span><span class="fl">10</span><span class="op">)</span></span>
<span><span class="va">pca_atac_no_norm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mat_norm.html">multiply_cols</a></span><span class="op">(</span><span class="va">svd_atac_no_norm</span><span class="op">$</span><span class="va">v</span>, <span class="va">svd_atac</span><span class="op">$</span><span class="va">d</span><span class="op">)</span></span></code></pre></div>
<p>If we skip normalization, we first observe that we get very high
correlation of our first PC to the reads-per-cell in our peak matrix</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cor_to_depth</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_rows.html" class="external-link">bind_rows</a></span><span class="op">(</span></span>
<span>  <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span></span>
<span>    method<span class="op">=</span><span class="st">"z-score normalize"</span>,</span>
<span>    abs_cor_to_depth <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cor</a></span><span class="op">(</span><span class="fu">Matrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums-methods.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">mat_lsi</span><span class="op">)</span>, <span class="va">pca_atac</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    PC<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span><span class="va">abs_cor_to_depth</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span></span>
<span>    method<span class="op">=</span><span class="st">"raw TF-IDF"</span>,</span>
<span>    abs_cor_to_depth <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cor</a></span><span class="op">(</span><span class="fu">Matrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums-methods.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">mat_lsi</span><span class="op">)</span>, <span class="va">pca_atac_no_norm</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    PC<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span><span class="va">abs_cor_to_depth</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">cor_to_depth</span>, <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">PC</span>, <span class="va">abs_cor_to_depth</span>, color<span class="op">=</span><span class="va">method</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title<span class="op">=</span><span class="st">"Correlation to of PCs to read depth"</span><span class="op">)</span></span></code></pre></div>
<p><img src="pbmc3k_files/figure-html/unnamed-chunk-44-1.png" width="700"></p>
<p>In terms of the actual PCA results, we can see that the cell
embeddings have mostly 1-to-1 correspondence across the first 6 PCs,
though later PCs start to diverge. The first PC of the raw TF-IDF
corresponds mostly to read depth, and the signal is spread out across 2
PCs in the z-score normalized variant.</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cor_between_embeddings</span> <span class="op">&lt;-</span> <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/expand_grid.html" class="external-link">expand_grid</a></span><span class="op">(</span></span>
<span>  pca_atac_no_norm <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">pca_atac_no_norm</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  pca_atac<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">pca_atac</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span></span>
<span>  cor <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cor</a></span><span class="op">(</span><span class="va">.env</span><span class="op">$</span><span class="va">pca_atac</span>, <span class="va">.env</span><span class="op">$</span><span class="va">pca_atac_no_norm</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">cor_between_embeddings</span>, <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">pca_atac</span>, <span class="va">pca_atac_no_norm</span>, fill<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">cor</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html" class="external-link">geom_tile</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html" class="external-link">geom_text</a></span><span class="op">(</span>mapping<span class="op">=</span><span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>label<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"%.2f"</span>, <span class="va">cor</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_continuous</a></span><span class="op">(</span>breaks<span class="op">=</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span>breaks<span class="op">=</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title<span class="op">=</span><span class="st">"Correlation between cell embeddings"</span>, </span>
<span>                x<span class="op">=</span><span class="st">"z-score normalize PCs"</span>,</span>
<span>                y <span class="op">=</span><span class="st">"raw TF-IDF PCs"</span><span class="op">)</span></span></code></pre></div>
<p><img src="pbmc3k_files/figure-html/unnamed-chunk-45-1.png" width="700"></p>
<p>And if we look at the loading on each peak from the PCA, we see a
similar result.</p>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cor_between_loadings</span> <span class="op">&lt;-</span> <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/expand_grid.html" class="external-link">expand_grid</a></span><span class="op">(</span></span>
<span>  pca_atac_no_norm <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">svd_atac_no_norm</span><span class="op">$</span><span class="va">u</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  pca_atac<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">svd_atac</span><span class="op">$</span><span class="va">u</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span></span>
<span>  cor <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cor</a></span><span class="op">(</span><span class="va">.env</span><span class="op">$</span><span class="va">svd_atac</span><span class="op">$</span><span class="va">u</span>, <span class="va">.env</span><span class="op">$</span><span class="va">svd_atac_no_norm</span><span class="op">$</span><span class="va">u</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">cor_between_loadings</span>, <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">pca_atac</span>, <span class="va">pca_atac_no_norm</span>, fill<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">cor</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html" class="external-link">geom_tile</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html" class="external-link">geom_text</a></span><span class="op">(</span>mapping<span class="op">=</span><span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>label<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"%.2f"</span>, <span class="va">cor</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_continuous</a></span><span class="op">(</span>breaks<span class="op">=</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span>breaks<span class="op">=</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title<span class="op">=</span><span class="st">"Correlation between peak loadings"</span>,</span>
<span>                x<span class="op">=</span><span class="st">"z-score normalize PCs"</span>,</span>
<span>                y <span class="op">=</span><span class="st">"raw TF-IDF PCs"</span><span class="op">)</span></span></code></pre></div>
<p><img src="pbmc3k_files/figure-html/unnamed-chunk-46-1.png" width="700"></p>
<p>Finally, the UMAP generated once we exclude the first PC is fairly
similar, though with a notable difference in positioning for some of the
dendritic cells</p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">12341512</span><span class="op">)</span></span>
<span><span class="va">umap_atac_no_norm</span> <span class="op">&lt;-</span> <span class="fu">uwot</span><span class="fu">::</span><span class="fu"><a href="https://jlmelville.github.io/uwot/reference/umap.html" class="external-link">umap</a></span><span class="op">(</span><span class="va">pca_atac_no_norm</span><span class="op">[</span>,<span class="op">-</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plot_embedding.html">plot_embedding</a></span><span class="op">(</span><span class="va">clusts_atac</span>, <span class="va">umap_atac_no_norm</span>, colors_discrete <span class="op">=</span> <span class="fu"><a href="../reference/palettes.html">discrete_palette</a></span><span class="op">(</span><span class="st">"ironMan"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>color<span class="op">=</span><span class="st">"none"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="../reference/plot_embedding.html">plot_embedding</a></span><span class="op">(</span><span class="va">cell_types</span>, <span class="va">umap_atac_no_norm</span><span class="op">)</span></span></code></pre></div>
<p><img src="pbmc3k_files/figure-html/unnamed-chunk-47-1.png" width="700"></p>
</details>
</div>
<div class="section level2">
<h2 id="motif-footprinting">Motif footprinting<a class="anchor" aria-label="anchor" href="#motif-footprinting"></a>
</h2>
<p>For motif footprinting, first we need to find all instances of our
motifs-of-interest in peaks</p>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="op">{</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicRanges" class="external-link">GenomicRanges</a></span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/Biostrings" class="external-link">Biostrings</a></span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="va">peaks_sorted</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">peaks</span>, <span class="va">chr</span>, <span class="va">start</span><span class="op">)</span></span>
<span><span class="va">peaks_gr</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span><span class="va">peaks_sorted</span>, start <span class="op">=</span> <span class="va">start</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/methods/as.html" class="external-link">as</a></span><span class="op">(</span><span class="st">"GenomicRanges"</span><span class="op">)</span></span>
<span><span class="va">selected_motifs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="st">"CEBPA"</span> <span class="op">=</span> <span class="st">"ENSG00000245848_LINE568_CEBPA_D_N4"</span>,</span>
<span>  <span class="st">"EOMES"</span> <span class="op">=</span> <span class="st">"ENSG00000163508_LINE3544_EOMES_D_N1"</span>, </span>
<span>  <span class="st">"SPI1"</span> <span class="op">=</span> <span class="st">"ENSG00000066336_LINE1813_SPI1_D_N5"</span>,</span>
<span>  <span class="st">"CTCF"</span> <span class="op">=</span> <span class="st">"ENSG00000102974_LINE747_CTCF_D_N67"</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span><span class="op">{</span></span>
<span>  <span class="va">motif_positions</span> <span class="op">&lt;-</span> <span class="fu">motifmatchr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/motifmatchr/man/matchMotifs.html" class="external-link">matchMotifs</a></span><span class="op">(</span></span>
<span>      <span class="fu">chromVARmotifs</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/chromVARmotifs/man/human_pwms_v2.html" class="external-link">human_pwms_v2</a></span><span class="op">[</span><span class="va">selected_motifs</span><span class="op">]</span>, </span>
<span>      <span class="va">peaks_gr</span>, genome<span class="op">=</span><span class="st">"hg38"</span>, out<span class="op">=</span><span class="st">"positions"</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">motif_positions</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">selected_motifs</span><span class="op">)</span></span>
<span><span class="va">motif_positions</span></span></code></pre></div>
<pre><code><span><span class="co">## GRangesList object of length 4:</span></span>
<span><span class="co">## $CEBPA</span></span>
<span><span class="co">## GRanges object with 13983 ranges and 1 metadata column:</span></span>
<span><span class="co">##           seqnames              ranges strand |     score</span></span>
<span><span class="co">##              &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt;</span></span>
<span><span class="co">##       [1]     chr1     1060191-1060200      + |   7.24878</span></span>
<span><span class="co">##       [2]     chr1     1398356-1398365      - |   7.31950</span></span>
<span><span class="co">##       [3]     chr1     1408228-1408237      - |   7.91954</span></span>
<span><span class="co">##       [4]     chr1     1470604-1470613      + |   7.26055</span></span>
<span><span class="co">##       [5]     chr1     1614370-1614379      - |   7.33072</span></span>
<span><span class="co">##       ...      ...                 ...    ... .       ...</span></span>
<span><span class="co">##   [13979]     chrX 154247973-154247982      + |   7.91954</span></span>
<span><span class="co">##   [13980]     chrX 154377819-154377828      - |   7.91954</span></span>
<span><span class="co">##   [13981]     chrX 154497506-154497515      + |   8.62478</span></span>
<span><span class="co">##   [13982]     chrX 154734157-154734166      + |   7.33771</span></span>
<span><span class="co">##   [13983]     chrX 155242494-155242503      - |   7.24396</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 39 sequences from an unspecified genome; no seqlengths</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## ...</span></span>
<span><span class="co">## &lt;3 more elements&gt;</span></span></code></pre>
<p>Next, we can use these motif positions to plot aggregate
accessibility surrounding TF binding sites across our cell types as a
proxy for TF activity. Here we’re able to see enrichment of
accessibility neighboring the sites of a myeloid transcription factor in
our DC and Monocyte cells. Transcription factor binding is (generally)
mutually-exclusive with nucleosome occupancy, so the transcription
factor is bound it creates accessibility in its flanking regions. The
squiggly bit in the center is due to Tn5 insertion bias of the motif
itself.</p>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_tf_footprint.html">plot_tf_footprint</a></span><span class="op">(</span></span>
<span>  <span class="va">frags</span>,</span>
<span>  <span class="va">motif_positions</span><span class="op">$</span><span class="va">CEBPA</span>,</span>
<span>  cell_groups <span class="op">=</span> <span class="va">cell_types</span>,</span>
<span>  flank <span class="op">=</span> <span class="fl">250</span>,</span>
<span>  smooth <span class="op">=</span> <span class="fl">2</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title<span class="op">=</span><span class="st">"CEBPA"</span><span class="op">)</span></span></code></pre></div>
<p><img src="pbmc3k_files/figure-html/unnamed-chunk-49-1.png" width="700"></p>
<p>We can also use the <code>patchwork</code> library to show multiple
plots in a grid, highlighting cell-type-specific factors as well as
general factors like CTCF.</p>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">footprinting_plots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">motif</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">selected_motifs</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">footprinting_plots</span><span class="op">[[</span><span class="va">motif</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_tf_footprint.html">plot_tf_footprint</a></span><span class="op">(</span></span>
<span>    <span class="va">frags</span>, </span>
<span>    <span class="va">motif_positions</span><span class="op">[[</span><span class="va">motif</span><span class="op">]</span><span class="op">]</span>, </span>
<span>    cell_groups <span class="op">=</span> <span class="va">cell_types</span>, </span>
<span>    flank<span class="op">=</span><span class="fl">250</span>,</span>
<span>    smooth<span class="op">=</span><span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title<span class="op">=</span><span class="va">motif</span>, color<span class="op">=</span><span class="st">"Cluster"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu">patchwork</span><span class="fu">::</span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="va">footprinting_plots</span>, guides<span class="op">=</span><span class="st">"collect"</span><span class="op">)</span></span></code></pre></div>
<p><img src="pbmc3k_files/figure-html/unnamed-chunk-50-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="genome-accessibility-tracks">Genome accessibility tracks<a class="anchor" aria-label="anchor" href="#genome-accessibility-tracks"></a>
</h2>
<p>To plot genome accessibility tracks, we need to select a genome
region to view. BPCells provides a helper function to find genome
regions centered around a gene.</p>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">region</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gene_region.html">gene_region</a></span><span class="op">(</span><span class="va">genes</span>, <span class="st">"CD19"</span>, extend_bp <span class="op">=</span> <span class="fl">1e5</span><span class="op">)</span></span>
<span><span class="va">region</span></span></code></pre></div>
<pre><code><span><span class="co">## $chr</span></span>
<span><span class="co">## [1] "chr16"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $start</span></span>
<span><span class="co">## [1] 28831970</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $end</span></span>
<span><span class="co">## [1] 29039342</span></span></code></pre>
<p>For normalizing the tracks, we need to provide the total number of
reads for each cell type. This can be substituted for total reads in
peaks or other metrics if desired.</p>
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">read_counts</span> <span class="op">&lt;-</span> <span class="va">atac_qc</span><span class="op">$</span><span class="va">nFrags</span><span class="op">[</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">match</a></span><span class="op">(</span><span class="fu"><a href="../reference/IterableFragments-methods.html">cellNames</a></span><span class="op">(</span><span class="va">frags</span><span class="op">)</span>, <span class="va">atac_qc</span><span class="op">$</span><span class="va">cellName</span><span class="op">)</span></span>
<span><span class="op">]</span></span></code></pre></div>
<p>We can create the first component of our track plot by plotting the
genome tracks themselves. We can see a small peak in the center that is
mainly present in B cells (top row), but it is unclear where this sits
relative to the B-cell marker CD19.</p>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">coverage_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/trackplot_coverage.html">trackplot_coverage</a></span><span class="op">(</span></span>
<span>  <span class="va">frags</span>,</span>
<span>  region <span class="op">=</span> <span class="va">region</span>, </span>
<span>  groups<span class="op">=</span><span class="va">cell_types</span>,</span>
<span>  <span class="va">read_counts</span>,</span>
<span>  bins<span class="op">=</span><span class="fl">500</span></span>
<span><span class="op">)</span></span>
<span><span class="va">coverage_plot</span></span></code></pre></div>
<p><img src="pbmc3k_files/figure-html/unnamed-chunk-53-1.png" width="700"></p>
<p>This is much more useful with a gene annotation track added in. For
this we’ll get a set of canonical transcripts (one per gene) from
Gencode</p>
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">transcripts</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_gtf.html">read_gencode_transcripts</a></span><span class="op">(</span><span class="st">"./references"</span>, release<span class="op">=</span><span class="st">"42"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">transcripts</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 6 × 13</span></span></span>
<span><span class="co">##   chr   source feature     start    end score strand frame gene_id     gene_type</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> chr1  HAVANA transcript  <span style="text-decoration: underline;">65</span>418  <span style="text-decoration: underline;">71</span>585 .     +      .     ENSG000001… protein_…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> chr1  HAVANA exon        <span style="text-decoration: underline;">65</span>418  <span style="text-decoration: underline;">65</span>433 .     +      .     ENSG000001… protein_…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> chr1  HAVANA exon        <span style="text-decoration: underline;">65</span>519  <span style="text-decoration: underline;">65</span>573 .     +      .     ENSG000001… protein_…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">4</span> chr1  HAVANA exon        <span style="text-decoration: underline;">69</span>036  <span style="text-decoration: underline;">71</span>585 .     +      .     ENSG000001… protein_…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">5</span> chr1  HAVANA transcript <span style="text-decoration: underline;">450</span>739 <span style="text-decoration: underline;">451</span>678 .     -      .     ENSG000002… protein_…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">6</span> chr1  HAVANA exon       <span style="text-decoration: underline;">450</span>739 <span style="text-decoration: underline;">451</span>678 .     -      .     ENSG000002… protein_…</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 3 more variables: gene_name &lt;chr&gt;, transcript_id &lt;chr&gt;, MANE_Select &lt;lgl&gt;</span></span></span></code></pre>
<p>Then we can make an annotation track</p>
<div class="sourceCode" id="cb71"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gene_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/trackplot_gene.html">trackplot_gene</a></span><span class="op">(</span><span class="va">transcripts</span>, <span class="va">region</span><span class="op">)</span></span>
<span><span class="va">gene_plot</span></span></code></pre></div>
<p><img src="pbmc3k_files/figure-html/unnamed-chunk-55-1.png" width="576"></p>
<p>And optionally a scale bar.</p>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scalebar_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/trackplot_scalebar.html">trackplot_scalebar</a></span><span class="op">(</span><span class="va">region</span><span class="op">)</span></span>
<span><span class="va">scalebar_plot</span></span></code></pre></div>
<p><img src="pbmc3k_files/figure-html/unnamed-chunk-56-1.png" width="576"></p>
<p>Finally, we can stack these elements with
<code><a href="../reference/trackplot_combine.html">trackplot_combine()</a></code>. Now we see that the small peak is just
upstream of the CD19 gene.</p>
<div class="sourceCode" id="cb73"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># We list plots in order from top to bottom to combine.</span></span>
<span></span>
<span><span class="co"># Notice that our inputs are also just ggplot objects, so we can make modifications</span></span>
<span><span class="co"># like removing the color legend from our gene track.</span></span>
<span><span class="fu"><a href="../reference/trackplot_combine.html">trackplot_combine</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    <span class="va">scalebar_plot</span>, </span>
<span>    <span class="va">coverage_plot</span>, </span>
<span>    <span class="va">gene_plot</span> <span class="op">+</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>color<span class="op">=</span><span class="st">"none"</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="pbmc3k_files/figure-html/unnamed-chunk-57-1.png" width="700"></p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Created by Benjamin Parks.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>



  <img src="https://plausible.benparks.net/flask-plausible/bpcells-docs.png" style="position:absolute;">
</body>
</html>
